{
 "cells": [
  {
   "cell_type": "code",
   "execution_count": 1,
   "id": "09e6507a",
   "metadata": {},
   "outputs": [],
   "source": [
    "import pandas as pd\n",
    "import os\n",
    "import utils\n",
    "import importlib\n",
    "\n",
    "importlib.reload(utils)\n",
    "PATH = os.getcwd()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 2,
   "id": "9a1a137d",
   "metadata": {},
   "outputs": [
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "C:\\Users\\shtolz\\AppData\\Local\\Temp\\ipykernel_36464\\3020032990.py:4: FutureWarning: The default value of regex will change from True to False in a future version.\n",
      "  tsv['sample_title'] = tsv['sample_title'].str.replace('\\.', ' ')\n"
     ]
    },
    {
     "data": {
      "application/vnd.microsoft.datawrangler.viewer.v0+json": {
       "columns": [
        {
         "name": "index",
         "rawType": "int64",
         "type": "integer"
        },
        {
         "name": "study_accession",
         "rawType": "object",
         "type": "string"
        },
        {
         "name": "sample_accession",
         "rawType": "object",
         "type": "string"
        },
        {
         "name": "secondary_sample_accession",
         "rawType": "object",
         "type": "string"
        },
        {
         "name": "experiment_accession",
         "rawType": "object",
         "type": "string"
        },
        {
         "name": "run_accession",
         "rawType": "object",
         "type": "string"
        },
        {
         "name": "tax_id",
         "rawType": "int64",
         "type": "integer"
        },
        {
         "name": "scientific_name",
         "rawType": "object",
         "type": "string"
        },
        {
         "name": "library_name",
         "rawType": "float64",
         "type": "float"
        },
        {
         "name": "library_layout",
         "rawType": "object",
         "type": "string"
        },
        {
         "name": "library_strategy",
         "rawType": "object",
         "type": "string"
        },
        {
         "name": "library_source",
         "rawType": "object",
         "type": "string"
        },
        {
         "name": "library_selection",
         "rawType": "object",
         "type": "string"
        },
        {
         "name": "read_count",
         "rawType": "int64",
         "type": "integer"
        },
        {
         "name": "experiment_title",
         "rawType": "object",
         "type": "string"
        },
        {
         "name": "study_title",
         "rawType": "object",
         "type": "string"
        },
        {
         "name": "experiment_alias",
         "rawType": "object",
         "type": "string"
        },
        {
         "name": "fastq_ftp",
         "rawType": "object",
         "type": "string"
        },
        {
         "name": "submitted_ftp",
         "rawType": "float64",
         "type": "float"
        },
        {
         "name": "sample_alias",
         "rawType": "object",
         "type": "string"
        },
        {
         "name": "sample_title",
         "rawType": "object",
         "type": "string"
        },
        {
         "name": "nominal_sdev",
         "rawType": "float64",
         "type": "float"
        }
       ],
       "conversionMethod": "pd.DataFrame",
       "ref": "419b8a98-d5a1-495a-b03c-32d37b39d91e",
       "rows": [
        [
         "0",
         "PRJNA312377",
         "SAMN04504459",
         "SRS1302527",
         "SRX1589961",
         "SRR3175541",
         "9606",
         "Homo sapiens",
         null,
         "PAIRED",
         "RNA-Seq",
         "TRANSCRIPTOMIC",
         "cDNA",
         "25177108",
         "Illumina HiSeq 2000 sequencing; GSM2064645: HEB rep1; Homo sapiens; RNA-Seq",
         "Acute exposure to graded levels of hypoxia in U87-MG cells: the role of hypoxia in GBM",
         "GSM2064645",
         "ftp.sra.ebi.ac.uk/vol1/fastq/SRR317/001/SRR3175541/SRR3175541_1.fastq.gz;ftp.sra.ebi.ac.uk/vol1/fastq/SRR317/001/SRR3175541/SRR3175541_2.fastq.gz",
         null,
         "GSM2064645",
         "HEB rep1",
         null
        ],
        [
         "1",
         "PRJNA312377",
         "SAMN04504460",
         "SRS1302526",
         "SRX1589962",
         "SRR3175542",
         "9606",
         "Homo sapiens",
         null,
         "PAIRED",
         "RNA-Seq",
         "TRANSCRIPTOMIC",
         "cDNA",
         "26420999",
         "Illumina HiSeq 2000 sequencing; GSM2064646: HEB rep2; Homo sapiens; RNA-Seq",
         "Acute exposure to graded levels of hypoxia in U87-MG cells: the role of hypoxia in GBM",
         "GSM2064646",
         "ftp.sra.ebi.ac.uk/vol1/fastq/SRR317/002/SRR3175542/SRR3175542_1.fastq.gz;ftp.sra.ebi.ac.uk/vol1/fastq/SRR317/002/SRR3175542/SRR3175542_2.fastq.gz",
         null,
         "GSM2064646",
         "HEB rep2",
         null
        ],
        [
         "2",
         "PRJNA312377",
         "SAMN04504461",
         "SRS1302525",
         "SRX1589963",
         "SRR3175543",
         "9606",
         "Homo sapiens",
         null,
         "PAIRED",
         "RNA-Seq",
         "TRANSCRIPTOMIC",
         "cDNA",
         "26424104",
         "Illumina HiSeq 2000 sequencing; GSM2064647: HEB rep3; Homo sapiens; RNA-Seq",
         "Acute exposure to graded levels of hypoxia in U87-MG cells: the role of hypoxia in GBM",
         "GSM2064647",
         "ftp.sra.ebi.ac.uk/vol1/fastq/SRR317/003/SRR3175543/SRR3175543_1.fastq.gz;ftp.sra.ebi.ac.uk/vol1/fastq/SRR317/003/SRR3175543/SRR3175543_2.fastq.gz",
         null,
         "GSM2064647",
         "HEB rep3",
         null
        ],
        [
         "3",
         "PRJNA312377",
         "SAMN04504462",
         "SRS1302524",
         "SRX1589964",
         "SRR3175544",
         "9606",
         "Homo sapiens",
         null,
         "PAIRED",
         "RNA-Seq",
         "TRANSCRIPTOMIC",
         "cDNA",
         "25178516",
         "Illumina HiSeq 2000 sequencing; GSM2064648: HEB.1%hypoxia.rep1; Homo sapiens; RNA-Seq",
         "Acute exposure to graded levels of hypoxia in U87-MG cells: the role of hypoxia in GBM",
         "GSM2064648",
         "ftp.sra.ebi.ac.uk/vol1/fastq/SRR317/004/SRR3175544/SRR3175544_1.fastq.gz;ftp.sra.ebi.ac.uk/vol1/fastq/SRR317/004/SRR3175544/SRR3175544_2.fastq.gz",
         null,
         "GSM2064648",
         "HEB 1%hypoxia rep1",
         null
        ],
        [
         "4",
         "PRJNA312377",
         "SAMN04504463",
         "SRS1302523",
         "SRX1589965",
         "SRR3175545",
         "9606",
         "Homo sapiens",
         null,
         "PAIRED",
         "RNA-Seq",
         "TRANSCRIPTOMIC",
         "cDNA",
         "26416051",
         "Illumina HiSeq 2000 sequencing; GSM2064649: HEB.1%hypoxia.rep2; Homo sapiens; RNA-Seq",
         "Acute exposure to graded levels of hypoxia in U87-MG cells: the role of hypoxia in GBM",
         "GSM2064649",
         "ftp.sra.ebi.ac.uk/vol1/fastq/SRR317/005/SRR3175545/SRR3175545_1.fastq.gz;ftp.sra.ebi.ac.uk/vol1/fastq/SRR317/005/SRR3175545/SRR3175545_2.fastq.gz",
         null,
         "GSM2064649",
         "HEB 1%hypoxia rep2",
         null
        ]
       ],
       "shape": {
        "columns": 21,
        "rows": 5
       }
      },
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>study_accession</th>\n",
       "      <th>sample_accession</th>\n",
       "      <th>secondary_sample_accession</th>\n",
       "      <th>experiment_accession</th>\n",
       "      <th>run_accession</th>\n",
       "      <th>tax_id</th>\n",
       "      <th>scientific_name</th>\n",
       "      <th>library_name</th>\n",
       "      <th>library_layout</th>\n",
       "      <th>library_strategy</th>\n",
       "      <th>...</th>\n",
       "      <th>library_selection</th>\n",
       "      <th>read_count</th>\n",
       "      <th>experiment_title</th>\n",
       "      <th>study_title</th>\n",
       "      <th>experiment_alias</th>\n",
       "      <th>fastq_ftp</th>\n",
       "      <th>submitted_ftp</th>\n",
       "      <th>sample_alias</th>\n",
       "      <th>sample_title</th>\n",
       "      <th>nominal_sdev</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>PRJNA312377</td>\n",
       "      <td>SAMN04504459</td>\n",
       "      <td>SRS1302527</td>\n",
       "      <td>SRX1589961</td>\n",
       "      <td>SRR3175541</td>\n",
       "      <td>9606</td>\n",
       "      <td>Homo sapiens</td>\n",
       "      <td>NaN</td>\n",
       "      <td>PAIRED</td>\n",
       "      <td>RNA-Seq</td>\n",
       "      <td>...</td>\n",
       "      <td>cDNA</td>\n",
       "      <td>25177108</td>\n",
       "      <td>Illumina HiSeq 2000 sequencing; GSM2064645: HE...</td>\n",
       "      <td>Acute exposure to graded levels of hypoxia in ...</td>\n",
       "      <td>GSM2064645</td>\n",
       "      <td>ftp.sra.ebi.ac.uk/vol1/fastq/SRR317/001/SRR317...</td>\n",
       "      <td>NaN</td>\n",
       "      <td>GSM2064645</td>\n",
       "      <td>HEB rep1</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>PRJNA312377</td>\n",
       "      <td>SAMN04504460</td>\n",
       "      <td>SRS1302526</td>\n",
       "      <td>SRX1589962</td>\n",
       "      <td>SRR3175542</td>\n",
       "      <td>9606</td>\n",
       "      <td>Homo sapiens</td>\n",
       "      <td>NaN</td>\n",
       "      <td>PAIRED</td>\n",
       "      <td>RNA-Seq</td>\n",
       "      <td>...</td>\n",
       "      <td>cDNA</td>\n",
       "      <td>26420999</td>\n",
       "      <td>Illumina HiSeq 2000 sequencing; GSM2064646: HE...</td>\n",
       "      <td>Acute exposure to graded levels of hypoxia in ...</td>\n",
       "      <td>GSM2064646</td>\n",
       "      <td>ftp.sra.ebi.ac.uk/vol1/fastq/SRR317/002/SRR317...</td>\n",
       "      <td>NaN</td>\n",
       "      <td>GSM2064646</td>\n",
       "      <td>HEB rep2</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>PRJNA312377</td>\n",
       "      <td>SAMN04504461</td>\n",
       "      <td>SRS1302525</td>\n",
       "      <td>SRX1589963</td>\n",
       "      <td>SRR3175543</td>\n",
       "      <td>9606</td>\n",
       "      <td>Homo sapiens</td>\n",
       "      <td>NaN</td>\n",
       "      <td>PAIRED</td>\n",
       "      <td>RNA-Seq</td>\n",
       "      <td>...</td>\n",
       "      <td>cDNA</td>\n",
       "      <td>26424104</td>\n",
       "      <td>Illumina HiSeq 2000 sequencing; GSM2064647: HE...</td>\n",
       "      <td>Acute exposure to graded levels of hypoxia in ...</td>\n",
       "      <td>GSM2064647</td>\n",
       "      <td>ftp.sra.ebi.ac.uk/vol1/fastq/SRR317/003/SRR317...</td>\n",
       "      <td>NaN</td>\n",
       "      <td>GSM2064647</td>\n",
       "      <td>HEB rep3</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>PRJNA312377</td>\n",
       "      <td>SAMN04504462</td>\n",
       "      <td>SRS1302524</td>\n",
       "      <td>SRX1589964</td>\n",
       "      <td>SRR3175544</td>\n",
       "      <td>9606</td>\n",
       "      <td>Homo sapiens</td>\n",
       "      <td>NaN</td>\n",
       "      <td>PAIRED</td>\n",
       "      <td>RNA-Seq</td>\n",
       "      <td>...</td>\n",
       "      <td>cDNA</td>\n",
       "      <td>25178516</td>\n",
       "      <td>Illumina HiSeq 2000 sequencing; GSM2064648: HE...</td>\n",
       "      <td>Acute exposure to graded levels of hypoxia in ...</td>\n",
       "      <td>GSM2064648</td>\n",
       "      <td>ftp.sra.ebi.ac.uk/vol1/fastq/SRR317/004/SRR317...</td>\n",
       "      <td>NaN</td>\n",
       "      <td>GSM2064648</td>\n",
       "      <td>HEB 1%hypoxia rep1</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>PRJNA312377</td>\n",
       "      <td>SAMN04504463</td>\n",
       "      <td>SRS1302523</td>\n",
       "      <td>SRX1589965</td>\n",
       "      <td>SRR3175545</td>\n",
       "      <td>9606</td>\n",
       "      <td>Homo sapiens</td>\n",
       "      <td>NaN</td>\n",
       "      <td>PAIRED</td>\n",
       "      <td>RNA-Seq</td>\n",
       "      <td>...</td>\n",
       "      <td>cDNA</td>\n",
       "      <td>26416051</td>\n",
       "      <td>Illumina HiSeq 2000 sequencing; GSM2064649: HE...</td>\n",
       "      <td>Acute exposure to graded levels of hypoxia in ...</td>\n",
       "      <td>GSM2064649</td>\n",
       "      <td>ftp.sra.ebi.ac.uk/vol1/fastq/SRR317/005/SRR317...</td>\n",
       "      <td>NaN</td>\n",
       "      <td>GSM2064649</td>\n",
       "      <td>HEB 1%hypoxia rep2</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>5 rows × 21 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "  study_accession sample_accession secondary_sample_accession  \\\n",
       "0     PRJNA312377     SAMN04504459                 SRS1302527   \n",
       "1     PRJNA312377     SAMN04504460                 SRS1302526   \n",
       "2     PRJNA312377     SAMN04504461                 SRS1302525   \n",
       "3     PRJNA312377     SAMN04504462                 SRS1302524   \n",
       "4     PRJNA312377     SAMN04504463                 SRS1302523   \n",
       "\n",
       "  experiment_accession run_accession  tax_id scientific_name  library_name  \\\n",
       "0           SRX1589961    SRR3175541    9606    Homo sapiens           NaN   \n",
       "1           SRX1589962    SRR3175542    9606    Homo sapiens           NaN   \n",
       "2           SRX1589963    SRR3175543    9606    Homo sapiens           NaN   \n",
       "3           SRX1589964    SRR3175544    9606    Homo sapiens           NaN   \n",
       "4           SRX1589965    SRR3175545    9606    Homo sapiens           NaN   \n",
       "\n",
       "  library_layout library_strategy  ... library_selection read_count  \\\n",
       "0         PAIRED          RNA-Seq  ...              cDNA   25177108   \n",
       "1         PAIRED          RNA-Seq  ...              cDNA   26420999   \n",
       "2         PAIRED          RNA-Seq  ...              cDNA   26424104   \n",
       "3         PAIRED          RNA-Seq  ...              cDNA   25178516   \n",
       "4         PAIRED          RNA-Seq  ...              cDNA   26416051   \n",
       "\n",
       "                                    experiment_title  \\\n",
       "0  Illumina HiSeq 2000 sequencing; GSM2064645: HE...   \n",
       "1  Illumina HiSeq 2000 sequencing; GSM2064646: HE...   \n",
       "2  Illumina HiSeq 2000 sequencing; GSM2064647: HE...   \n",
       "3  Illumina HiSeq 2000 sequencing; GSM2064648: HE...   \n",
       "4  Illumina HiSeq 2000 sequencing; GSM2064649: HE...   \n",
       "\n",
       "                                         study_title experiment_alias  \\\n",
       "0  Acute exposure to graded levels of hypoxia in ...       GSM2064645   \n",
       "1  Acute exposure to graded levels of hypoxia in ...       GSM2064646   \n",
       "2  Acute exposure to graded levels of hypoxia in ...       GSM2064647   \n",
       "3  Acute exposure to graded levels of hypoxia in ...       GSM2064648   \n",
       "4  Acute exposure to graded levels of hypoxia in ...       GSM2064649   \n",
       "\n",
       "                                           fastq_ftp submitted_ftp  \\\n",
       "0  ftp.sra.ebi.ac.uk/vol1/fastq/SRR317/001/SRR317...           NaN   \n",
       "1  ftp.sra.ebi.ac.uk/vol1/fastq/SRR317/002/SRR317...           NaN   \n",
       "2  ftp.sra.ebi.ac.uk/vol1/fastq/SRR317/003/SRR317...           NaN   \n",
       "3  ftp.sra.ebi.ac.uk/vol1/fastq/SRR317/004/SRR317...           NaN   \n",
       "4  ftp.sra.ebi.ac.uk/vol1/fastq/SRR317/005/SRR317...           NaN   \n",
       "\n",
       "   sample_alias        sample_title nominal_sdev  \n",
       "0    GSM2064645            HEB rep1          NaN  \n",
       "1    GSM2064646            HEB rep2          NaN  \n",
       "2    GSM2064647            HEB rep3          NaN  \n",
       "3    GSM2064648  HEB 1%hypoxia rep1          NaN  \n",
       "4    GSM2064649  HEB 1%hypoxia rep2          NaN  \n",
       "\n",
       "[5 rows x 21 columns]"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    }
   ],
   "source": [
    "tsv_file = 'U87_hypx_2.tsv'\n",
    "\n",
    "tsv = pd.read_csv(os.path.join(PATH, 'data', 'metadata', tsv_file), sep = '\\t')\n",
    "tsv['sample_title'] = tsv['sample_title'].str.replace('\\.', ' ')\n",
    "display(tsv.head())"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 3,
   "id": "0e783b24",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "application/vnd.microsoft.datawrangler.viewer.v0+json": {
       "columns": [
        {
         "name": "index",
         "rawType": "int64",
         "type": "integer"
        },
        {
         "name": "run_accession",
         "rawType": "object",
         "type": "string"
        },
        {
         "name": "cell",
         "rawType": "object",
         "type": "string"
        },
        {
         "name": "condition",
         "rawType": "object",
         "type": "string"
        },
        {
         "name": "replicate",
         "rawType": "object",
         "type": "string"
        }
       ],
       "conversionMethod": "pd.DataFrame",
       "ref": "b67bd340-aa74-406c-87e4-c0a1c3f7f094",
       "rows": [
        [
         "0",
         "SRR3175541",
         "HEB",
         "normoxia",
         "rep1"
        ],
        [
         "1",
         "SRR3175542",
         "HEB",
         "normoxia",
         "rep2"
        ],
        [
         "2",
         "SRR3175543",
         "HEB",
         "normoxia",
         "rep3"
        ],
        [
         "3",
         "SRR3175544",
         "HEB",
         "1%hypoxia",
         "rep1"
        ],
        [
         "4",
         "SRR3175545",
         "HEB",
         "1%hypoxia",
         "rep2"
        ]
       ],
       "shape": {
        "columns": 4,
        "rows": 5
       }
      },
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>run_accession</th>\n",
       "      <th>cell</th>\n",
       "      <th>condition</th>\n",
       "      <th>replicate</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>SRR3175541</td>\n",
       "      <td>HEB</td>\n",
       "      <td>normoxia</td>\n",
       "      <td>rep1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>SRR3175542</td>\n",
       "      <td>HEB</td>\n",
       "      <td>normoxia</td>\n",
       "      <td>rep2</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>SRR3175543</td>\n",
       "      <td>HEB</td>\n",
       "      <td>normoxia</td>\n",
       "      <td>rep3</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>SRR3175544</td>\n",
       "      <td>HEB</td>\n",
       "      <td>1%hypoxia</td>\n",
       "      <td>rep1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>SRR3175545</td>\n",
       "      <td>HEB</td>\n",
       "      <td>1%hypoxia</td>\n",
       "      <td>rep2</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "  run_accession cell  condition replicate\n",
       "0    SRR3175541  HEB   normoxia      rep1\n",
       "1    SRR3175542  HEB   normoxia      rep2\n",
       "2    SRR3175543  HEB   normoxia      rep3\n",
       "3    SRR3175544  HEB  1%hypoxia      rep1\n",
       "4    SRR3175545  HEB  1%hypoxia      rep2"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    }
   ],
   "source": [
    "md = utils.prepare_metadata(df = tsv, id_col = 'run_accession', label_col = 'sample_title', sep = ' ', label_list = ['cell', 'condition', 'replicate'])\n",
    "md['replicate'] = md['replicate'].fillna(md['condition'])\n",
    "md.loc[md['condition'].str.contains('rep'), 'condition'] = 'normoxia'\n",
    "display(md.head(5))\n",
    "md.to_csv(os.path.join(PATH, 'data', 'metadata', 'hypx_u87_2.csv'), index = False)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 4,
   "id": "d520a1e8",
   "metadata": {},
   "outputs": [],
   "source": [
    "nanopore_folder = 'nanopore'\n",
    "folder = 'counts_hcc_hch'\n",
    "\n",
    "counts = utils.prepare_counts(os.path.join(PATH, 'data',nanopore_folder, folder), keep_symbol=True, keep_suffix=True)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 5,
   "id": "84ba51b9",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "application/vnd.microsoft.datawrangler.viewer.v0+json": {
       "columns": [
        {
         "name": "id",
         "rawType": "object",
         "type": "string"
        },
        {
         "name": "symbol",
         "rawType": "object",
         "type": "string"
        },
        {
         "name": "HAC",
         "rawType": "int64",
         "type": "integer"
        },
        {
         "name": "HAH",
         "rawType": "int64",
         "type": "integer"
        }
       ],
       "conversionMethod": "pd.DataFrame",
       "ref": "6a50a3f8-4fff-4764-a3bd-07ef63eaf5bf",
       "rows": [
        [
         "ENSG00000000003.15",
         "TSPAN6",
         "295",
         "359"
        ],
        [
         "ENSG00000000005.6",
         "TNMD",
         "0",
         "0"
        ],
        [
         "ENSG00000000419.14",
         "DPM1",
         "1414",
         "993"
        ],
        [
         "ENSG00000000457.14",
         "SCYL3",
         "7",
         "3"
        ],
        [
         "ENSG00000000460.17",
         "C1orf112",
         "26",
         "7"
        ]
       ],
       "shape": {
        "columns": 3,
        "rows": 5
       }
      },
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>symbol</th>\n",
       "      <th>HAC</th>\n",
       "      <th>HAH</th>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>id</th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>ENSG00000000003.15</th>\n",
       "      <td>TSPAN6</td>\n",
       "      <td>295</td>\n",
       "      <td>359</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>ENSG00000000005.6</th>\n",
       "      <td>TNMD</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>ENSG00000000419.14</th>\n",
       "      <td>DPM1</td>\n",
       "      <td>1414</td>\n",
       "      <td>993</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>ENSG00000000457.14</th>\n",
       "      <td>SCYL3</td>\n",
       "      <td>7</td>\n",
       "      <td>3</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>ENSG00000000460.17</th>\n",
       "      <td>C1orf112</td>\n",
       "      <td>26</td>\n",
       "      <td>7</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "                      symbol   HAC  HAH\n",
       "id                                     \n",
       "ENSG00000000003.15    TSPAN6   295  359\n",
       "ENSG00000000005.6       TNMD     0    0\n",
       "ENSG00000000419.14      DPM1  1414  993\n",
       "ENSG00000000457.14     SCYL3     7    3\n",
       "ENSG00000000460.17  C1orf112    26    7"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    }
   ],
   "source": [
    "nanopore_folder = 'nanopore'\n",
    "folder = 'counts_hac_hah'\n",
    "\n",
    "counts = utils.prepare_counts(os.path.join(PATH, 'data',nanopore_folder, folder), keep_symbol=True, keep_suffix=True)\n",
    "names = ['symbol', 'HAC', 'HAH']\n",
    "counts.columns = names\n",
    "display(counts.head(5))\n",
    "counts.to_csv(os.path.join(PATH, 'data', nanopore_folder,'count_tables', f'{folder}.csv'))"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 6,
   "id": "8cf9707d",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "application/vnd.microsoft.datawrangler.viewer.v0+json": {
       "columns": [
        {
         "name": "id",
         "rawType": "object",
         "type": "string"
        },
        {
         "name": "symbol",
         "rawType": "object",
         "type": "string"
        },
        {
         "name": "DM12",
         "rawType": "int64",
         "type": "integer"
        },
        {
         "name": "DM13",
         "rawType": "int64",
         "type": "integer"
        },
        {
         "name": "DM14",
         "rawType": "int64",
         "type": "integer"
        },
        {
         "name": "DM15",
         "rawType": "int64",
         "type": "integer"
        },
        {
         "name": "DM16",
         "rawType": "int64",
         "type": "integer"
        },
        {
         "name": "DM17",
         "rawType": "int64",
         "type": "integer"
        },
        {
         "name": "DM18",
         "rawType": "int64",
         "type": "integer"
        },
        {
         "name": "DM19",
         "rawType": "int64",
         "type": "integer"
        },
        {
         "name": "DM20",
         "rawType": "int64",
         "type": "integer"
        },
        {
         "name": "DM21",
         "rawType": "int64",
         "type": "integer"
        },
        {
         "name": "DM22",
         "rawType": "int64",
         "type": "integer"
        },
        {
         "name": "DM23",
         "rawType": "int64",
         "type": "integer"
        },
        {
         "name": "DM24",
         "rawType": "int64",
         "type": "integer"
        },
        {
         "name": "DM25",
         "rawType": "int64",
         "type": "integer"
        },
        {
         "name": "DM26",
         "rawType": "int64",
         "type": "integer"
        },
        {
         "name": "DM27",
         "rawType": "int64",
         "type": "integer"
        },
        {
         "name": "DM3",
         "rawType": "int64",
         "type": "integer"
        },
        {
         "name": "DM4",
         "rawType": "int64",
         "type": "integer"
        },
        {
         "name": "DM5",
         "rawType": "int64",
         "type": "integer"
        },
        {
         "name": "DM6",
         "rawType": "int64",
         "type": "integer"
        },
        {
         "name": "DM7",
         "rawType": "int64",
         "type": "integer"
        },
        {
         "name": "DM8",
         "rawType": "int64",
         "type": "integer"
        }
       ],
       "conversionMethod": "pd.DataFrame",
       "ref": "669746ea-c2a0-4115-9f62-12b3e8b66bd3",
       "rows": [
        [
         "ENSG00000000003.15",
         "TSPAN6",
         "638",
         "989",
         "652",
         "892",
         "395",
         "616",
         "563",
         "616",
         "927",
         "665",
         "1032",
         "1011",
         "1153",
         "1030",
         "384",
         "608",
         "961",
         "535",
         "1155",
         "1308",
         "1005",
         "1914"
        ],
        [
         "ENSG00000000005.6",
         "TNMD",
         "0",
         "0",
         "0",
         "0",
         "0",
         "0",
         "0",
         "0",
         "1",
         "0",
         "0",
         "2",
         "2",
         "2",
         "0",
         "2",
         "0",
         "0",
         "0",
         "0",
         "0",
         "0"
        ],
        [
         "ENSG00000000419.14",
         "DPM1",
         "2122",
         "2348",
         "2510",
         "2161",
         "2948",
         "2859",
         "3760",
         "3002",
         "2344",
         "1213",
         "2693",
         "1824",
         "2836",
         "1768",
         "863",
         "1050",
         "3194",
         "1547",
         "3991",
         "4129",
         "3425",
         "5922"
        ],
        [
         "ENSG00000000457.14",
         "SCYL3",
         "280",
         "250",
         "260",
         "273",
         "192",
         "233",
         "263",
         "242",
         "336",
         "153",
         "346",
         "305",
         "431",
         "204",
         "136",
         "131",
         "146",
         "85",
         "176",
         "245",
         "173",
         "270"
        ],
        [
         "ENSG00000000460.17",
         "C1orf112",
         "192",
         "137",
         "187",
         "181",
         "236",
         "278",
         "277",
         "276",
         "459",
         "278",
         "483",
         "425",
         "481",
         "368",
         "135",
         "189",
         "440",
         "209",
         "587",
         "764",
         "524",
         "847"
        ]
       ],
       "shape": {
        "columns": 23,
        "rows": 5
       }
      },
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>symbol</th>\n",
       "      <th>DM12</th>\n",
       "      <th>DM13</th>\n",
       "      <th>DM14</th>\n",
       "      <th>DM15</th>\n",
       "      <th>DM16</th>\n",
       "      <th>DM17</th>\n",
       "      <th>DM18</th>\n",
       "      <th>DM19</th>\n",
       "      <th>DM20</th>\n",
       "      <th>...</th>\n",
       "      <th>DM24</th>\n",
       "      <th>DM25</th>\n",
       "      <th>DM26</th>\n",
       "      <th>DM27</th>\n",
       "      <th>DM3</th>\n",
       "      <th>DM4</th>\n",
       "      <th>DM5</th>\n",
       "      <th>DM6</th>\n",
       "      <th>DM7</th>\n",
       "      <th>DM8</th>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>id</th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>ENSG00000000003.15</th>\n",
       "      <td>TSPAN6</td>\n",
       "      <td>638</td>\n",
       "      <td>989</td>\n",
       "      <td>652</td>\n",
       "      <td>892</td>\n",
       "      <td>395</td>\n",
       "      <td>616</td>\n",
       "      <td>563</td>\n",
       "      <td>616</td>\n",
       "      <td>927</td>\n",
       "      <td>...</td>\n",
       "      <td>1153</td>\n",
       "      <td>1030</td>\n",
       "      <td>384</td>\n",
       "      <td>608</td>\n",
       "      <td>961</td>\n",
       "      <td>535</td>\n",
       "      <td>1155</td>\n",
       "      <td>1308</td>\n",
       "      <td>1005</td>\n",
       "      <td>1914</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>ENSG00000000005.6</th>\n",
       "      <td>TNMD</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>1</td>\n",
       "      <td>...</td>\n",
       "      <td>2</td>\n",
       "      <td>2</td>\n",
       "      <td>0</td>\n",
       "      <td>2</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>ENSG00000000419.14</th>\n",
       "      <td>DPM1</td>\n",
       "      <td>2122</td>\n",
       "      <td>2348</td>\n",
       "      <td>2510</td>\n",
       "      <td>2161</td>\n",
       "      <td>2948</td>\n",
       "      <td>2859</td>\n",
       "      <td>3760</td>\n",
       "      <td>3002</td>\n",
       "      <td>2344</td>\n",
       "      <td>...</td>\n",
       "      <td>2836</td>\n",
       "      <td>1768</td>\n",
       "      <td>863</td>\n",
       "      <td>1050</td>\n",
       "      <td>3194</td>\n",
       "      <td>1547</td>\n",
       "      <td>3991</td>\n",
       "      <td>4129</td>\n",
       "      <td>3425</td>\n",
       "      <td>5922</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>ENSG00000000457.14</th>\n",
       "      <td>SCYL3</td>\n",
       "      <td>280</td>\n",
       "      <td>250</td>\n",
       "      <td>260</td>\n",
       "      <td>273</td>\n",
       "      <td>192</td>\n",
       "      <td>233</td>\n",
       "      <td>263</td>\n",
       "      <td>242</td>\n",
       "      <td>336</td>\n",
       "      <td>...</td>\n",
       "      <td>431</td>\n",
       "      <td>204</td>\n",
       "      <td>136</td>\n",
       "      <td>131</td>\n",
       "      <td>146</td>\n",
       "      <td>85</td>\n",
       "      <td>176</td>\n",
       "      <td>245</td>\n",
       "      <td>173</td>\n",
       "      <td>270</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>ENSG00000000460.17</th>\n",
       "      <td>C1orf112</td>\n",
       "      <td>192</td>\n",
       "      <td>137</td>\n",
       "      <td>187</td>\n",
       "      <td>181</td>\n",
       "      <td>236</td>\n",
       "      <td>278</td>\n",
       "      <td>277</td>\n",
       "      <td>276</td>\n",
       "      <td>459</td>\n",
       "      <td>...</td>\n",
       "      <td>481</td>\n",
       "      <td>368</td>\n",
       "      <td>135</td>\n",
       "      <td>189</td>\n",
       "      <td>440</td>\n",
       "      <td>209</td>\n",
       "      <td>587</td>\n",
       "      <td>764</td>\n",
       "      <td>524</td>\n",
       "      <td>847</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>5 rows × 23 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "                      symbol  DM12  DM13  DM14  DM15  DM16  DM17  DM18  DM19  \\\n",
       "id                                                                             \n",
       "ENSG00000000003.15    TSPAN6   638   989   652   892   395   616   563   616   \n",
       "ENSG00000000005.6       TNMD     0     0     0     0     0     0     0     0   \n",
       "ENSG00000000419.14      DPM1  2122  2348  2510  2161  2948  2859  3760  3002   \n",
       "ENSG00000000457.14     SCYL3   280   250   260   273   192   233   263   242   \n",
       "ENSG00000000460.17  C1orf112   192   137   187   181   236   278   277   276   \n",
       "\n",
       "                    DM20  ...  DM24  DM25  DM26  DM27   DM3   DM4   DM5   DM6  \\\n",
       "id                        ...                                                   \n",
       "ENSG00000000003.15   927  ...  1153  1030   384   608   961   535  1155  1308   \n",
       "ENSG00000000005.6      1  ...     2     2     0     2     0     0     0     0   \n",
       "ENSG00000000419.14  2344  ...  2836  1768   863  1050  3194  1547  3991  4129   \n",
       "ENSG00000000457.14   336  ...   431   204   136   131   146    85   176   245   \n",
       "ENSG00000000460.17   459  ...   481   368   135   189   440   209   587   764   \n",
       "\n",
       "                     DM7   DM8  \n",
       "id                              \n",
       "ENSG00000000003.15  1005  1914  \n",
       "ENSG00000000005.6      0     0  \n",
       "ENSG00000000419.14  3425  5922  \n",
       "ENSG00000000457.14   173   270  \n",
       "ENSG00000000460.17   524   847  \n",
       "\n",
       "[5 rows x 23 columns]"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    }
   ],
   "source": [
    "folder = 'counts_u87_u_a_b_mcf7_hct_116'\n",
    "\n",
    "counts = utils.prepare_counts(os.path.join(PATH, 'data', folder), keep_symbol=True, keep_suffix=False)\n",
    "cols = [i.split('_')[0] for i in counts.columns]\n",
    "counts.columns = cols\n",
    "display(counts.head(5))\n",
    "counts.to_csv(os.path.join(PATH, 'data', 'count_tables', f'{folder}.csv'))"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 16,
   "id": "287d75f8",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>symbol</th>\n",
       "      <th>SAMN13059448</th>\n",
       "      <th>SAMN13059449</th>\n",
       "      <th>SAMN13059450</th>\n",
       "      <th>SAMN13059451</th>\n",
       "      <th>SAMN13059452</th>\n",
       "      <th>SAMN13059453</th>\n",
       "      <th>SAMN13059454</th>\n",
       "      <th>SAMN13059455</th>\n",
       "      <th>SAMN13059456</th>\n",
       "      <th>SAMN13059457</th>\n",
       "      <th>SAMN13059478</th>\n",
       "      <th>SAMN13059479</th>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>id</th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>ENSG00000000003.16</th>\n",
       "      <td>TSPAN6</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>2</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>1</td>\n",
       "      <td>0</td>\n",
       "      <td>1</td>\n",
       "      <td>1</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>ENSG00000000005.6</th>\n",
       "      <td>TNMD</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>ENSG00000000419.14</th>\n",
       "      <td>DPM1</td>\n",
       "      <td>2867</td>\n",
       "      <td>1146</td>\n",
       "      <td>1280</td>\n",
       "      <td>1303</td>\n",
       "      <td>1005</td>\n",
       "      <td>1202</td>\n",
       "      <td>2024</td>\n",
       "      <td>1628</td>\n",
       "      <td>2212</td>\n",
       "      <td>1942</td>\n",
       "      <td>1142</td>\n",
       "      <td>1173</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>ENSG00000000457.14</th>\n",
       "      <td>SCYL3</td>\n",
       "      <td>1506</td>\n",
       "      <td>691</td>\n",
       "      <td>731</td>\n",
       "      <td>731</td>\n",
       "      <td>764</td>\n",
       "      <td>961</td>\n",
       "      <td>878</td>\n",
       "      <td>751</td>\n",
       "      <td>1137</td>\n",
       "      <td>1093</td>\n",
       "      <td>935</td>\n",
       "      <td>901</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>ENSG00000000460.17</th>\n",
       "      <td>C1orf112</td>\n",
       "      <td>925</td>\n",
       "      <td>594</td>\n",
       "      <td>901</td>\n",
       "      <td>898</td>\n",
       "      <td>543</td>\n",
       "      <td>648</td>\n",
       "      <td>908</td>\n",
       "      <td>597</td>\n",
       "      <td>696</td>\n",
       "      <td>653</td>\n",
       "      <td>867</td>\n",
       "      <td>950</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "                      symbol  SAMN13059448  SAMN13059449  SAMN13059450  \\\n",
       "id                                                                       \n",
       "ENSG00000000003.16    TSPAN6             0             0             2   \n",
       "ENSG00000000005.6       TNMD             0             0             0   \n",
       "ENSG00000000419.14      DPM1          2867          1146          1280   \n",
       "ENSG00000000457.14     SCYL3          1506           691           731   \n",
       "ENSG00000000460.17  C1orf112           925           594           901   \n",
       "\n",
       "                    SAMN13059451  SAMN13059452  SAMN13059453  SAMN13059454  \\\n",
       "id                                                                           \n",
       "ENSG00000000003.16             0             0             0             1   \n",
       "ENSG00000000005.6              0             0             0             0   \n",
       "ENSG00000000419.14          1303          1005          1202          2024   \n",
       "ENSG00000000457.14           731           764           961           878   \n",
       "ENSG00000000460.17           898           543           648           908   \n",
       "\n",
       "                    SAMN13059455  SAMN13059456  SAMN13059457  SAMN13059478  \\\n",
       "id                                                                           \n",
       "ENSG00000000003.16             0             1             1             0   \n",
       "ENSG00000000005.6              0             0             0             0   \n",
       "ENSG00000000419.14          1628          2212          1942          1142   \n",
       "ENSG00000000457.14           751          1137          1093           935   \n",
       "ENSG00000000460.17           597           696           653           867   \n",
       "\n",
       "                    SAMN13059479  \n",
       "id                                \n",
       "ENSG00000000003.16             0  \n",
       "ENSG00000000005.6              0  \n",
       "ENSG00000000419.14          1173  \n",
       "ENSG00000000457.14           901  \n",
       "ENSG00000000460.17           950  "
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    }
   ],
   "source": [
    "folder = 'counts_hs_jun_o_e_t'\n",
    "\n",
    "counts = utils.prepare_counts(os.path.join(PATH, 'data', folder), keep_symbol=True, keep_suffix=False)\n",
    "display(counts.head(5))\n",
    "counts.to_csv(os.path.join(PATH, 'data', 'count_tables', f'{folder}.csv'))"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 22,
   "id": "1ece0065",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>study_accession</th>\n",
       "      <th>sample_accession</th>\n",
       "      <th>secondary_sample_accession</th>\n",
       "      <th>experiment_accession</th>\n",
       "      <th>run_accession</th>\n",
       "      <th>tax_id</th>\n",
       "      <th>scientific_name</th>\n",
       "      <th>library_name</th>\n",
       "      <th>library_layout</th>\n",
       "      <th>library_strategy</th>\n",
       "      <th>library_source</th>\n",
       "      <th>library_selection</th>\n",
       "      <th>experiment_title</th>\n",
       "      <th>study_title</th>\n",
       "      <th>experiment_alias</th>\n",
       "      <th>fastq_ftp</th>\n",
       "      <th>submitted_ftp</th>\n",
       "      <th>sample_alias</th>\n",
       "      <th>sample_title</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>PRJNA823675</td>\n",
       "      <td>SAMN27351154</td>\n",
       "      <td>SRS12508444</td>\n",
       "      <td>SRX14746971</td>\n",
       "      <td>SRR18643781</td>\n",
       "      <td>9606</td>\n",
       "      <td>Homo sapiens</td>\n",
       "      <td>GSM6022032</td>\n",
       "      <td>PAIRED</td>\n",
       "      <td>RNA-Seq</td>\n",
       "      <td>TRANSCRIPTOMIC</td>\n",
       "      <td>cDNA</td>\n",
       "      <td>Illumina NovaSeq 6000 sequencing; GSM6022032: ...</td>\n",
       "      <td>RNA-seq analysis of gene expression in normoxi...</td>\n",
       "      <td>GSM6022032_r1</td>\n",
       "      <td>ftp.sra.ebi.ac.uk/vol1/fastq/SRR186/081/SRR186...</td>\n",
       "      <td>NaN</td>\n",
       "      <td>GSM6022032</td>\n",
       "      <td>T47D Normoxia rep3</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "  study_accession sample_accession secondary_sample_accession  \\\n",
       "0     PRJNA823675     SAMN27351154                SRS12508444   \n",
       "\n",
       "  experiment_accession run_accession  tax_id scientific_name library_name  \\\n",
       "0          SRX14746971   SRR18643781    9606    Homo sapiens   GSM6022032   \n",
       "\n",
       "  library_layout library_strategy  library_source library_selection  \\\n",
       "0         PAIRED          RNA-Seq  TRANSCRIPTOMIC              cDNA   \n",
       "\n",
       "                                    experiment_title  \\\n",
       "0  Illumina NovaSeq 6000 sequencing; GSM6022032: ...   \n",
       "\n",
       "                                         study_title experiment_alias  \\\n",
       "0  RNA-seq analysis of gene expression in normoxi...    GSM6022032_r1   \n",
       "\n",
       "                                           fastq_ftp  submitted_ftp  \\\n",
       "0  ftp.sra.ebi.ac.uk/vol1/fastq/SRR186/081/SRR186...            NaN   \n",
       "\n",
       "  sample_alias        sample_title  \n",
       "0   GSM6022032  T47D Normoxia rep3  "
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "0    [T47D, Normoxia, rep3]\n",
      "1    [T47D, Normoxia, rep2]\n",
      "Name: sample_title, dtype: object\n"
     ]
    },
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>run_accession</th>\n",
       "      <th>cell_type</th>\n",
       "      <th>treatment</th>\n",
       "      <th>replicate</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>SRR18643781</td>\n",
       "      <td>T47D</td>\n",
       "      <td>Normoxia</td>\n",
       "      <td>rep3</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>SRR18643782</td>\n",
       "      <td>T47D</td>\n",
       "      <td>Normoxia</td>\n",
       "      <td>rep2</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>SRR18643783</td>\n",
       "      <td>T47D</td>\n",
       "      <td>Normoxia</td>\n",
       "      <td>rep1</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "  run_accession cell_type treatment replicate\n",
       "0   SRR18643781      T47D  Normoxia      rep3\n",
       "1   SRR18643782      T47D  Normoxia      rep2\n",
       "2   SRR18643783      T47D  Normoxia      rep1"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    }
   ],
   "source": [
    "\n",
    "metadata = pd.read_csv(os.path.join(PATH, 'data', 'metadata', 'diff_cell_lines_hypx.tsv'), sep = '\\t')\n",
    "display(metadata.head(1))\n",
    "metadata = utils.prepare_metadata(id_col = 'run_accession', label_col = 'sample_title', sep = ' ', label_list = ['cell_type', 'treatment', 'replicate'], df = metadata)\n",
    "display(metadata.head(3))\n",
    "metadata.to_csv(os.path.join(PATH, 'data', 'metadata', 'diff_cell_lines_hypx.csv'), index = False)\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 11,
   "id": "50d69589",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>symbol</th>\n",
       "      <th>SAMN14600527</th>\n",
       "      <th>SAMN14600528</th>\n",
       "      <th>SAMN14600529</th>\n",
       "      <th>SAMN14600530</th>\n",
       "      <th>SAMN16205736</th>\n",
       "      <th>SAMN16205737</th>\n",
       "      <th>SAMN16205738</th>\n",
       "      <th>SAMN16205739</th>\n",
       "      <th>SAMN16205740</th>\n",
       "      <th>SAMN16205741</th>\n",
       "      <th>SAMN16205742</th>\n",
       "      <th>SAMN16205743</th>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>id</th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>ENSMUSG00000000001.5</th>\n",
       "      <td>Gnai3</td>\n",
       "      <td>3849</td>\n",
       "      <td>3682</td>\n",
       "      <td>3474</td>\n",
       "      <td>3233</td>\n",
       "      <td>2542</td>\n",
       "      <td>1954</td>\n",
       "      <td>2353</td>\n",
       "      <td>1553</td>\n",
       "      <td>2451</td>\n",
       "      <td>1761</td>\n",
       "      <td>3552</td>\n",
       "      <td>2064</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>ENSMUSG00000000003.16</th>\n",
       "      <td>Pbsn</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>ENSMUSG00000000028.16</th>\n",
       "      <td>Cdc45</td>\n",
       "      <td>3863</td>\n",
       "      <td>3485</td>\n",
       "      <td>4060</td>\n",
       "      <td>3552</td>\n",
       "      <td>3463</td>\n",
       "      <td>2755</td>\n",
       "      <td>2910</td>\n",
       "      <td>1945</td>\n",
       "      <td>3180</td>\n",
       "      <td>2203</td>\n",
       "      <td>1953</td>\n",
       "      <td>1115</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>ENSMUSG00000000031.18</th>\n",
       "      <td>H19</td>\n",
       "      <td>4094</td>\n",
       "      <td>3411</td>\n",
       "      <td>4206</td>\n",
       "      <td>3994</td>\n",
       "      <td>6032</td>\n",
       "      <td>4529</td>\n",
       "      <td>7559</td>\n",
       "      <td>3902</td>\n",
       "      <td>29552</td>\n",
       "      <td>18163</td>\n",
       "      <td>42277</td>\n",
       "      <td>26152</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>ENSMUSG00000000037.18</th>\n",
       "      <td>Scml2</td>\n",
       "      <td>10284</td>\n",
       "      <td>9464</td>\n",
       "      <td>11369</td>\n",
       "      <td>9929</td>\n",
       "      <td>8893</td>\n",
       "      <td>7217</td>\n",
       "      <td>6624</td>\n",
       "      <td>4724</td>\n",
       "      <td>8215</td>\n",
       "      <td>6443</td>\n",
       "      <td>5193</td>\n",
       "      <td>3031</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "                      symbol  SAMN14600527  SAMN14600528  SAMN14600529   \n",
       "id                                                                       \n",
       "ENSMUSG00000000001.5   Gnai3          3849          3682          3474  \\\n",
       "ENSMUSG00000000003.16   Pbsn             0             0             0   \n",
       "ENSMUSG00000000028.16  Cdc45          3863          3485          4060   \n",
       "ENSMUSG00000000031.18    H19          4094          3411          4206   \n",
       "ENSMUSG00000000037.18  Scml2         10284          9464         11369   \n",
       "\n",
       "                       SAMN14600530  SAMN16205736  SAMN16205737  SAMN16205738   \n",
       "id                                                                              \n",
       "ENSMUSG00000000001.5           3233          2542          1954          2353  \\\n",
       "ENSMUSG00000000003.16             0             0             0             0   \n",
       "ENSMUSG00000000028.16          3552          3463          2755          2910   \n",
       "ENSMUSG00000000031.18          3994          6032          4529          7559   \n",
       "ENSMUSG00000000037.18          9929          8893          7217          6624   \n",
       "\n",
       "                       SAMN16205739  SAMN16205740  SAMN16205741  SAMN16205742   \n",
       "id                                                                              \n",
       "ENSMUSG00000000001.5           1553          2451          1761          3552  \\\n",
       "ENSMUSG00000000003.16             0             0             0             0   \n",
       "ENSMUSG00000000028.16          1945          3180          2203          1953   \n",
       "ENSMUSG00000000031.18          3902         29552         18163         42277   \n",
       "ENSMUSG00000000037.18          4724          8215          6443          5193   \n",
       "\n",
       "                       SAMN16205743  \n",
       "id                                   \n",
       "ENSMUSG00000000001.5           2064  \n",
       "ENSMUSG00000000003.16             0  \n",
       "ENSMUSG00000000028.16          1115  \n",
       "ENSMUSG00000000031.18         26152  \n",
       "ENSMUSG00000000037.18          3031  "
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    }
   ],
   "source": [
    "BASE = os.path.join(PATH, 'Hypoxia_across_evolution')\n",
    "name = 'counts_mm_sperm_hyp'\n",
    "counts = utils.prepare_counts(os.path.join(BASE, 'data', name), keep_symbol=True)\n",
    "display(counts.head(5))\n",
    "counts.to_csv(os.path.join(BASE, 'data', 'count_tables', name + '.csv'))"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 14,
   "id": "ca304626",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>sample_accession</th>\n",
       "      <th>run_accession</th>\n",
       "      <th>fastq_ftp</th>\n",
       "      <th>experiment_accession</th>\n",
       "      <th>study_accession</th>\n",
       "      <th>tax_id</th>\n",
       "      <th>scientific_name</th>\n",
       "      <th>submitted_ftp</th>\n",
       "      <th>sra_ftp</th>\n",
       "      <th>sample_alias</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>SAMN07645764</td>\n",
       "      <td>['SRR6037990']</td>\n",
       "      <td>['ftp.sra.ebi.ac.uk/vol1/fastq/SRR603/000/SRR6...</td>\n",
       "      <td>SRX3186174</td>\n",
       "      <td>PRJNA407293</td>\n",
       "      <td>10090</td>\n",
       "      <td>Mus musculus</td>\n",
       "      <td>NaN</td>\n",
       "      <td>ftp.sra.ebi.ac.uk/vol1/srr/SRR603/000/SRR6037990</td>\n",
       "      <td>Hypo_R2</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "  sample_accession   run_accession   \n",
       "0     SAMN07645764  ['SRR6037990']  \\\n",
       "\n",
       "                                           fastq_ftp experiment_accession   \n",
       "0  ['ftp.sra.ebi.ac.uk/vol1/fastq/SRR603/000/SRR6...           SRX3186174  \\\n",
       "\n",
       "  study_accession  tax_id scientific_name  submitted_ftp   \n",
       "0     PRJNA407293   10090    Mus musculus            NaN  \\\n",
       "\n",
       "                                            sra_ftp sample_alias  \n",
       "0  ftp.sra.ebi.ac.uk/vol1/srr/SRR603/000/SRR6037990      Hypo_R2  "
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    },
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>sample_accession</th>\n",
       "      <th>treatment</th>\n",
       "      <th>replicate</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>SAMN07645764</td>\n",
       "      <td>Hypo</td>\n",
       "      <td>R2</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>SAMN07645757</td>\n",
       "      <td>Nor</td>\n",
       "      <td>R3</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>SAMN07645755</td>\n",
       "      <td>Nor</td>\n",
       "      <td>R1</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "  sample_accession treatment replicate\n",
       "0     SAMN07645764      Hypo        R2\n",
       "1     SAMN07645757       Nor        R3\n",
       "2     SAMN07645755       Nor        R1"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    }
   ],
   "source": [
    "file_name = 'MM_hepatocytes_hypx'\n",
    "label_list = ['treatment', 'replicate']\n",
    "\n",
    "metadata = pd.read_csv(os.path.join(BASE, 'data', 'metadata', file_name + '.tsv'), sep = '\\t')\n",
    "display(metadata.head(1))\n",
    "metadata = utils.prepare_metadata(id_col = 'sample_accession', label_col = 'sample_alias', sep = '_', label_list = label_list, df = metadata)\n",
    "display(metadata.head(3))\n",
    "metadata.to_csv(os.path.join(BASE, 'data', 'metadata', file_name + '.csv'), index = False)\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 4,
   "id": "9e24f1e9",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>sample_accession</th>\n",
       "      <th>run_accession</th>\n",
       "      <th>fastq_ftp</th>\n",
       "      <th>experiment_accession</th>\n",
       "      <th>study_accession</th>\n",
       "      <th>tax_id</th>\n",
       "      <th>scientific_name</th>\n",
       "      <th>submitted_ftp</th>\n",
       "      <th>sra_ftp</th>\n",
       "      <th>sample_title</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>SAMN27567837</td>\n",
       "      <td>['SRR18745695']</td>\n",
       "      <td>['ftp.sra.ebi.ac.uk/vol1/fastq/SRR187/095/SRR1...</td>\n",
       "      <td>SRX14845719</td>\n",
       "      <td>PRJNA826504</td>\n",
       "      <td>10090</td>\n",
       "      <td>Mus musculus</td>\n",
       "      <td>NaN</td>\n",
       "      <td>ftp.sra.ebi.ac.uk/vol1/srr/SRR187/095/SRR18745695</td>\n",
       "      <td>GL261-GFP_&lt;0.1%O2_rep2_ATACseq</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "  sample_accession    run_accession   \n",
       "0     SAMN27567837  ['SRR18745695']  \\\n",
       "\n",
       "                                           fastq_ftp experiment_accession   \n",
       "0  ['ftp.sra.ebi.ac.uk/vol1/fastq/SRR187/095/SRR1...          SRX14845719  \\\n",
       "\n",
       "  study_accession  tax_id scientific_name  submitted_ftp   \n",
       "0     PRJNA826504   10090    Mus musculus            NaN  \\\n",
       "\n",
       "                                             sra_ftp   \n",
       "0  ftp.sra.ebi.ac.uk/vol1/srr/SRR187/095/SRR18745695  \\\n",
       "\n",
       "                     sample_title  \n",
       "0  GL261-GFP_<0.1%O2_rep2_ATACseq  "
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    },
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>sample_accession</th>\n",
       "      <th>cell_type</th>\n",
       "      <th>ox_conc</th>\n",
       "      <th>replicate</th>\n",
       "      <th>method</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>SAMN27567837</td>\n",
       "      <td>GL261-GFP</td>\n",
       "      <td>&lt;0.1%O2</td>\n",
       "      <td>rep2</td>\n",
       "      <td>ATACseq</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>SAMN27567840</td>\n",
       "      <td>GL261-GFP</td>\n",
       "      <td>1%O2</td>\n",
       "      <td>rep1</td>\n",
       "      <td>ATACseq</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>SAMN27567841</td>\n",
       "      <td>GL261-GFP</td>\n",
       "      <td>21%O2</td>\n",
       "      <td>rep2</td>\n",
       "      <td>ATACseq</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "  sample_accession  cell_type  ox_conc replicate   method\n",
       "0     SAMN27567837  GL261-GFP  <0.1%O2      rep2  ATACseq\n",
       "1     SAMN27567840  GL261-GFP     1%O2      rep1  ATACseq\n",
       "2     SAMN27567841  GL261-GFP    21%O2      rep2  ATACseq"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    }
   ],
   "source": [
    "file_name = 'MM_atac_hypx_glioma'\n",
    "label_list = ['cell_type', 'ox_conc', 'replicate', 'method']\n",
    "\n",
    "metadata = pd.read_csv(os.path.join('data', 'metadata', file_name + '.tsv'), sep = '\\t')\n",
    "display(metadata.head(1))\n",
    "metadata = utils.prepare_metadata(id_col = 'sample_accession', label_col = 'sample_title', sep = '_', label_list = label_list, df = metadata)\n",
    "display(metadata.head(3))\n",
    "metadata.to_csv(os.path.join('data', 'metadata', file_name + '.csv'), index = False)\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 22,
   "id": "3595dd45",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>sample_accession</th>\n",
       "      <th>run_accession</th>\n",
       "      <th>fastq_ftp</th>\n",
       "      <th>experiment_accession</th>\n",
       "      <th>study_accession</th>\n",
       "      <th>tax_id</th>\n",
       "      <th>scientific_name</th>\n",
       "      <th>submitted_ftp</th>\n",
       "      <th>sra_ftp</th>\n",
       "      <th>sample_title</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>SAMN14600527</td>\n",
       "      <td>['SRR11548535']</td>\n",
       "      <td>['ftp.sra.ebi.ac.uk/vol1/fastq/SRR115/035/SRR1...</td>\n",
       "      <td>SRX8118557</td>\n",
       "      <td>PRJNA625466</td>\n",
       "      <td>10090</td>\n",
       "      <td>Mus musculus</td>\n",
       "      <td>NaN</td>\n",
       "      <td>ftp.sra.ebi.ac.uk/vol1/srr/SRR115/035/SRR11548535</td>\n",
       "      <td>1% oxygen rep2 [WT_1_O2_ex2]</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "  sample_accession    run_accession   \n",
       "0     SAMN14600527  ['SRR11548535']  \\\n",
       "\n",
       "                                           fastq_ftp experiment_accession   \n",
       "0  ['ftp.sra.ebi.ac.uk/vol1/fastq/SRR115/035/SRR1...           SRX8118557  \\\n",
       "\n",
       "  study_accession  tax_id scientific_name  submitted_ftp   \n",
       "0     PRJNA625466   10090    Mus musculus            NaN  \\\n",
       "\n",
       "                                             sra_ftp   \n",
       "0  ftp.sra.ebi.ac.uk/vol1/srr/SRR115/035/SRR11548535  \\\n",
       "\n",
       "                   sample_title  \n",
       "0  1% oxygen rep2 [WT_1_O2_ex2]  "
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    },
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>sample_accession</th>\n",
       "      <th>treatment</th>\n",
       "      <th>replicate</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>SAMN14600527</td>\n",
       "      <td>Hypoxia</td>\n",
       "      <td>rep2</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>SAMN16205747</td>\n",
       "      <td>Hypoxia</td>\n",
       "      <td>rep1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>SAMN16205743</td>\n",
       "      <td>Hypoxia</td>\n",
       "      <td>rep1</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "  sample_accession treatment replicate\n",
       "0     SAMN14600527   Hypoxia      rep2\n",
       "1     SAMN16205747   Hypoxia      rep1\n",
       "2     SAMN16205743   Hypoxia      rep1"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    }
   ],
   "source": [
    "file_name = 'MM_sperm_hypx'\n",
    "label_list = ['treatment', 't', 'replicate', 't2']\n",
    "\n",
    "metadata = pd.read_csv(os.path.join(BASE, 'data', 'metadata', file_name + '.tsv'), sep = '\\t')\n",
    "display(metadata.head(1))\n",
    "metadata = utils.prepare_metadata(id_col = 'sample_accession', label_col = 'sample_title', sep = ' ', label_list = label_list, df = metadata)\n",
    "# Drop columns t and t2\n",
    "metadata = metadata.drop(['t', 't2'], axis = 1)\n",
    "# Replace 20% with control and 1% with hypoxia\n",
    "metadata['treatment'] = metadata['treatment'].replace('20%', 'Control')\n",
    "metadata['treatment'] = metadata['treatment'].replace('1%', 'Hypoxia')\n",
    "display(metadata.head(3))\n",
    "metadata.to_csv(os.path.join(BASE, 'data', 'metadata', file_name + '.csv'), index = False)\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 44,
   "id": "a80cfa69",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>study_accession</th>\n",
       "      <th>sample_accession</th>\n",
       "      <th>experiment_accession</th>\n",
       "      <th>run_accession</th>\n",
       "      <th>tax_id</th>\n",
       "      <th>scientific_name</th>\n",
       "      <th>fastq_ftp</th>\n",
       "      <th>submitted_ftp</th>\n",
       "      <th>sra_ftp</th>\n",
       "      <th>sample_title</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>PRJNA533075</td>\n",
       "      <td>SAMN11433314</td>\n",
       "      <td>SRX5693618</td>\n",
       "      <td>SRR8907797</td>\n",
       "      <td>10116</td>\n",
       "      <td>Rattus norvegicus</td>\n",
       "      <td>ftp.sra.ebi.ac.uk/vol1/fastq/SRR890/007/SRR890...</td>\n",
       "      <td>NaN</td>\n",
       "      <td>ftp.sra.ebi.ac.uk/vol1/srr/SRR890/007/SRR8907797</td>\n",
       "      <td>Fetal fetal normoxia RNAseq 1</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "  study_accession sample_accession experiment_accession run_accession  tax_id  \\\n",
       "0     PRJNA533075     SAMN11433314           SRX5693618    SRR8907797   10116   \n",
       "\n",
       "     scientific_name                                          fastq_ftp  \\\n",
       "0  Rattus norvegicus  ftp.sra.ebi.ac.uk/vol1/fastq/SRR890/007/SRR890...   \n",
       "\n",
       "   submitted_ftp                                           sra_ftp  \\\n",
       "0            NaN  ftp.sra.ebi.ac.uk/vol1/srr/SRR890/007/SRR8907797   \n",
       "\n",
       "                    sample_title  \n",
       "0  Fetal fetal normoxia RNAseq 1  "
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    },
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>run_accession</th>\n",
       "      <th>age</th>\n",
       "      <th>sex</th>\n",
       "      <th>treatment</th>\n",
       "      <th>method</th>\n",
       "      <th>replicate</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>SRR8907797</td>\n",
       "      <td>Fetal</td>\n",
       "      <td>fetal</td>\n",
       "      <td>normoxia</td>\n",
       "      <td>RNAseq</td>\n",
       "      <td>1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>SRR8907798</td>\n",
       "      <td>Fetal</td>\n",
       "      <td>fetal</td>\n",
       "      <td>normoxia</td>\n",
       "      <td>RNAseq</td>\n",
       "      <td>2</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>SRR8907799</td>\n",
       "      <td>Fetal</td>\n",
       "      <td>fetal</td>\n",
       "      <td>normoxia</td>\n",
       "      <td>RNAseq</td>\n",
       "      <td>3</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "  run_accession    age    sex treatment  method replicate\n",
       "0    SRR8907797  Fetal  fetal  normoxia  RNAseq         1\n",
       "1    SRR8907798  Fetal  fetal  normoxia  RNAseq         2\n",
       "2    SRR8907799  Fetal  fetal  normoxia  RNAseq         3"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    }
   ],
   "source": [
    "file_name = 'RN_heart_hypx'\n",
    "label_list = ['age', 'sex', 'treatment', 'method', 'replicate']\n",
    "\n",
    "metadata = pd.read_csv(os.path.join(BASE, 'data', 'metadata', file_name + '.tsv'), sep = '\\t')\n",
    "display(metadata.head(1))\n",
    "metadata = utils.prepare_metadata(id_col = 'run_accession', label_col = 'sample_title', sep = ' ', label_list = label_list, df = metadata)\n",
    "display(metadata.head(3))\n",
    "metadata.to_csv(os.path.join(BASE, 'data', 'metadata', file_name + '.csv'), index = False)\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 15,
   "id": "a6a48101",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>sample_accession</th>\n",
       "      <th>run_accession</th>\n",
       "      <th>fastq_ftp</th>\n",
       "      <th>secondary_study_accession</th>\n",
       "      <th>experiment_accession</th>\n",
       "      <th>study_accession</th>\n",
       "      <th>tax_id</th>\n",
       "      <th>scientific_name</th>\n",
       "      <th>instrument_model</th>\n",
       "      <th>library_layout</th>\n",
       "      <th>library_source</th>\n",
       "      <th>study_alias</th>\n",
       "      <th>experiment_alias</th>\n",
       "      <th>run_alias</th>\n",
       "      <th>submitted_ftp</th>\n",
       "      <th>sra_ftp</th>\n",
       "      <th>sample_alias</th>\n",
       "      <th>sample_title</th>\n",
       "      <th>bam_ftp</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>SAMN35298889</td>\n",
       "      <td>['SRR24681499']</td>\n",
       "      <td>['ftp.sra.ebi.ac.uk/vol1/fastq/SRR246/099/SRR2...</td>\n",
       "      <td>SRP438824</td>\n",
       "      <td>SRX20461070</td>\n",
       "      <td>PRJNA974833</td>\n",
       "      <td>9606</td>\n",
       "      <td>Homo sapiens</td>\n",
       "      <td>Illumina HiSeq 2500</td>\n",
       "      <td>PAIRED</td>\n",
       "      <td>GENOMIC</td>\n",
       "      <td>GSE233048</td>\n",
       "      <td>GSM7400931_r1</td>\n",
       "      <td>GSM7400931_r1</td>\n",
       "      <td>NaN</td>\n",
       "      <td>ftp.sra.ebi.ac.uk/vol1/srr/SRR246/099/SRR24681...</td>\n",
       "      <td>GSM7400931</td>\n",
       "      <td>CD4_mem_JUN-OE_IgG_CnR_Rep3</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "  sample_accession    run_accession  \\\n",
       "0     SAMN35298889  ['SRR24681499']   \n",
       "\n",
       "                                           fastq_ftp  \\\n",
       "0  ['ftp.sra.ebi.ac.uk/vol1/fastq/SRR246/099/SRR2...   \n",
       "\n",
       "  secondary_study_accession experiment_accession study_accession  tax_id  \\\n",
       "0                 SRP438824          SRX20461070     PRJNA974833    9606   \n",
       "\n",
       "  scientific_name     instrument_model library_layout library_source  \\\n",
       "0    Homo sapiens  Illumina HiSeq 2500         PAIRED        GENOMIC   \n",
       "\n",
       "  study_alias experiment_alias      run_alias  submitted_ftp  \\\n",
       "0   GSE233048    GSM7400931_r1  GSM7400931_r1            NaN   \n",
       "\n",
       "                                             sra_ftp sample_alias  \\\n",
       "0  ftp.sra.ebi.ac.uk/vol1/srr/SRR246/099/SRR24681...   GSM7400931   \n",
       "\n",
       "                  sample_title  bam_ftp  \n",
       "0  CD4_mem_JUN-OE_IgG_CnR_Rep3      NaN  "
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    },
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>sample_accession</th>\n",
       "      <th>treatment</th>\n",
       "      <th>vector</th>\n",
       "      <th>replicate</th>\n",
       "      <th>cell_type</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>SAMN35298889</td>\n",
       "      <td>JUN-OE</td>\n",
       "      <td>IgG</td>\n",
       "      <td>Rep3</td>\n",
       "      <td>CD4_mem</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>SAMN35298893</td>\n",
       "      <td>control</td>\n",
       "      <td>IgG</td>\n",
       "      <td>Rep2</td>\n",
       "      <td>CD4_mem</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>SAMN35298894</td>\n",
       "      <td>control</td>\n",
       "      <td>IgG</td>\n",
       "      <td>Rep1</td>\n",
       "      <td>CD4_mem</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "  sample_accession treatment vector replicate cell_type\n",
       "0     SAMN35298889    JUN-OE    IgG      Rep3   CD4_mem\n",
       "1     SAMN35298893   control    IgG      Rep2   CD4_mem\n",
       "2     SAMN35298894   control    IgG      Rep1   CD4_mem"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    }
   ],
   "source": [
    "file_name = 'T_CD4_JUN_OE'\n",
    "label_list = [ 'cell_type1', 'cell_type2', 'treatment', 'vector', 'remove2', 'replicate']\n",
    "\n",
    "metadata = pd.read_csv(os.path.join('data', 'metadata', file_name + '.tsv'), sep = '\\t')\n",
    "display(metadata.head(1))\n",
    "metadata = utils.prepare_metadata(id_col = 'sample_accession', label_col = 'sample_title', sep = '_', label_list = label_list, df = metadata)\n",
    "metadata['cell_type'] = metadata['cell_type1'] + '_' + metadata['cell_type2']\n",
    "metadata = metadata.drop(['cell_type1', 'cell_type2', 'remove2'], axis = 1)\n",
    "display(metadata.head(3))\n",
    "metadata.to_csv(os.path.join('data', 'metadata', file_name + '.csv'), index = False)\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 3,
   "id": "7ac266bc",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>sample_accession</th>\n",
       "      <th>run_accession</th>\n",
       "      <th>fastq_ftp</th>\n",
       "      <th>secondary_study_accession</th>\n",
       "      <th>experiment_accession</th>\n",
       "      <th>study_accession</th>\n",
       "      <th>tax_id</th>\n",
       "      <th>scientific_name</th>\n",
       "      <th>library_layout</th>\n",
       "      <th>library_source</th>\n",
       "      <th>experiment_alias</th>\n",
       "      <th>run_alias</th>\n",
       "      <th>submitted_ftp</th>\n",
       "      <th>sra_ftp</th>\n",
       "      <th>sample_alias</th>\n",
       "      <th>sample_title</th>\n",
       "      <th>bam_ftp</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>SAMN25754583</td>\n",
       "      <td>['SRR17939984']</td>\n",
       "      <td>['ftp.sra.ebi.ac.uk/vol1/fastq/SRR179/084/SRR1...</td>\n",
       "      <td>SRP358908</td>\n",
       "      <td>SRX14097871</td>\n",
       "      <td>PRJNA804500</td>\n",
       "      <td>9606</td>\n",
       "      <td>Homo sapiens</td>\n",
       "      <td>PAIRED</td>\n",
       "      <td>TRANSCRIPTOMIC</td>\n",
       "      <td>GSM5876435_r1</td>\n",
       "      <td>GSM5876435_r1</td>\n",
       "      <td>NaN</td>\n",
       "      <td>ftp.sra.ebi.ac.uk/vol1/srr/SRR179/084/SRR17939984</td>\n",
       "      <td>GSM5876435</td>\n",
       "      <td>JUN_AA</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "  sample_accession    run_accession  \\\n",
       "0     SAMN25754583  ['SRR17939984']   \n",
       "\n",
       "                                           fastq_ftp  \\\n",
       "0  ['ftp.sra.ebi.ac.uk/vol1/fastq/SRR179/084/SRR1...   \n",
       "\n",
       "  secondary_study_accession experiment_accession study_accession  tax_id  \\\n",
       "0                 SRP358908          SRX14097871     PRJNA804500    9606   \n",
       "\n",
       "  scientific_name library_layout  library_source experiment_alias  \\\n",
       "0    Homo sapiens         PAIRED  TRANSCRIPTOMIC    GSM5876435_r1   \n",
       "\n",
       "       run_alias  submitted_ftp  \\\n",
       "0  GSM5876435_r1            NaN   \n",
       "\n",
       "                                             sra_ftp sample_alias  \\\n",
       "0  ftp.sra.ebi.ac.uk/vol1/srr/SRR179/084/SRR17939984   GSM5876435   \n",
       "\n",
       "  sample_title  bam_ftp  \n",
       "0       JUN_AA      NaN  "
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    },
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>sample_accession</th>\n",
       "      <th>treatment</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>SAMN25754583</td>\n",
       "      <td>JUN_AA</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>SAMN25754585</td>\n",
       "      <td>con</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>SAMN25754581</td>\n",
       "      <td>iCAFs</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "  sample_accession treatment\n",
       "0     SAMN25754583    JUN_AA\n",
       "1     SAMN25754585       con\n",
       "2     SAMN25754581     iCAFs"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    }
   ],
   "source": [
    "file_name = 'Fibroblast_JUN_OE'\n",
    "label_list = [ 'treatment']\n",
    "\n",
    "metadata = pd.read_csv(os.path.join('data', 'metadata', file_name + '.tsv'), sep = '\\t')\n",
    "display(metadata.head(1))\n",
    "metadata = utils.prepare_metadata(id_col = 'sample_accession', label_col = 'sample_title', sep = 'sds', label_list = label_list, df = metadata)\n",
    "\n",
    "display(metadata.head(3))\n",
    "metadata.to_csv(os.path.join('data', 'metadata', file_name + '.csv'), index = False)\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 13,
   "id": "0c2b2034",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>sample_accession</th>\n",
       "      <th>run_accession</th>\n",
       "      <th>fastq_ftp</th>\n",
       "      <th>secondary_study_accession</th>\n",
       "      <th>experiment_accession</th>\n",
       "      <th>study_accession</th>\n",
       "      <th>tax_id</th>\n",
       "      <th>scientific_name</th>\n",
       "      <th>library_layout</th>\n",
       "      <th>library_source</th>\n",
       "      <th>experiment_alias</th>\n",
       "      <th>run_alias</th>\n",
       "      <th>submitted_ftp</th>\n",
       "      <th>sra_ftp</th>\n",
       "      <th>sample_alias</th>\n",
       "      <th>sample_title</th>\n",
       "      <th>bam_ftp</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>SAMN13059457</td>\n",
       "      <td>['SRR10312288']</td>\n",
       "      <td>['ftp.sra.ebi.ac.uk/vol1/fastq/SRR103/088/SRR1...</td>\n",
       "      <td>SRP220375</td>\n",
       "      <td>SRX7023508</td>\n",
       "      <td>PRJNA563926</td>\n",
       "      <td>9606</td>\n",
       "      <td>Homo sapiens</td>\n",
       "      <td>SINGLE</td>\n",
       "      <td>TRANSCRIPTOMIC</td>\n",
       "      <td>GSM4131090</td>\n",
       "      <td>GSM4131090_r1</td>\n",
       "      <td>NaN</td>\n",
       "      <td>ftp.sra.ebi.ac.uk/vol1/srr/SRR103/088/SRR10312288</td>\n",
       "      <td>GSM4131090</td>\n",
       "      <td>Control_CD19_Batch1</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "  sample_accession    run_accession  \\\n",
       "0     SAMN13059457  ['SRR10312288']   \n",
       "\n",
       "                                           fastq_ftp  \\\n",
       "0  ['ftp.sra.ebi.ac.uk/vol1/fastq/SRR103/088/SRR1...   \n",
       "\n",
       "  secondary_study_accession experiment_accession study_accession  tax_id  \\\n",
       "0                 SRP220375           SRX7023508     PRJNA563926    9606   \n",
       "\n",
       "  scientific_name library_layout  library_source experiment_alias  \\\n",
       "0    Homo sapiens         SINGLE  TRANSCRIPTOMIC       GSM4131090   \n",
       "\n",
       "       run_alias  submitted_ftp  \\\n",
       "0  GSM4131090_r1            NaN   \n",
       "\n",
       "                                             sra_ftp sample_alias  \\\n",
       "0  ftp.sra.ebi.ac.uk/vol1/srr/SRR103/088/SRR10312288   GSM4131090   \n",
       "\n",
       "          sample_title  bam_ftp  \n",
       "0  Control_CD19_Batch1      NaN  "
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    },
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>sample_accession</th>\n",
       "      <th>treatment</th>\n",
       "      <th>cell_type</th>\n",
       "      <th>replicate</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>SAMN13059457</td>\n",
       "      <td>Control</td>\n",
       "      <td>CD19</td>\n",
       "      <td>Batch1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>SAMN13059452</td>\n",
       "      <td>JUN</td>\n",
       "      <td>CD19</td>\n",
       "      <td>Batch2</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>SAMN13059450</td>\n",
       "      <td>JUN</td>\n",
       "      <td>HA</td>\n",
       "      <td>Batch3</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "  sample_accession treatment cell_type replicate\n",
       "0     SAMN13059457   Control      CD19    Batch1\n",
       "1     SAMN13059452       JUN      CD19    Batch2\n",
       "2     SAMN13059450       JUN        HA    Batch3"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    }
   ],
   "source": [
    "file_name = 'HS_JUN_O_E_T'\n",
    "label_list = [ 'treatment', 'cell_type', 'replicate']\n",
    "\n",
    "metadata = pd.read_csv(os.path.join('data', 'metadata', file_name + '.tsv'), sep = '\\t')\n",
    "display(metadata.head(1))\n",
    "metadata = utils.prepare_metadata(id_col = 'sample_accession', label_col = 'sample_title', sep = '_', label_list = label_list, df = metadata)\n",
    "\n",
    "display(metadata.head(3))\n",
    "metadata.to_csv(os.path.join('data', 'metadata', file_name + '.csv'), index = False)\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "8c731409",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>study_accession</th>\n",
       "      <th>sample_accession</th>\n",
       "      <th>experiment_accession</th>\n",
       "      <th>run_accession</th>\n",
       "      <th>tax_id</th>\n",
       "      <th>scientific_name</th>\n",
       "      <th>fastq_ftp</th>\n",
       "      <th>submitted_ftp</th>\n",
       "      <th>sra_ftp</th>\n",
       "      <th>sample_title</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>PRJNA667729</td>\n",
       "      <td>SAMN16380772</td>\n",
       "      <td>SRX9251173</td>\n",
       "      <td>SRR12781720</td>\n",
       "      <td>9606</td>\n",
       "      <td>Homo sapiens</td>\n",
       "      <td>ftp.sra.ebi.ac.uk/vol1/fastq/SRR127/020/SRR127...</td>\n",
       "      <td>NaN</td>\n",
       "      <td>ftp.sra.ebi.ac.uk/vol1/srr/SRR127/020/SRR12781720</td>\n",
       "      <td>S01_ControlNormoxiaChIP</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "  study_accession sample_accession experiment_accession run_accession  tax_id  \\\n",
       "0     PRJNA667729     SAMN16380772           SRX9251173   SRR12781720    9606   \n",
       "\n",
       "  scientific_name                                          fastq_ftp  \\\n",
       "0    Homo sapiens  ftp.sra.ebi.ac.uk/vol1/fastq/SRR127/020/SRR127...   \n",
       "\n",
       "   submitted_ftp                                            sra_ftp  \\\n",
       "0            NaN  ftp.sra.ebi.ac.uk/vol1/srr/SRR127/020/SRR12781720   \n",
       "\n",
       "              sample_title  \n",
       "0  S01_ControlNormoxiaChIP  "
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    },
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>sample_accession</th>\n",
       "      <th>antibody</th>\n",
       "      <th>treatment</th>\n",
       "      <th>target</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>SAMN16380772</td>\n",
       "      <td>S01_ControlNormoxiaChIP</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>SAMN16380772</td>\n",
       "      <td>S01_ControlNormoxiaChIP</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>SAMN16380792</td>\n",
       "      <td>S02_ControlHypoxiaChIP</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "  sample_accession                 antibody  treatment  target\n",
       "0     SAMN16380772  S01_ControlNormoxiaChIP        NaN     NaN\n",
       "1     SAMN16380772  S01_ControlNormoxiaChIP        NaN     NaN\n",
       "2     SAMN16380792   S02_ControlHypoxiaChIP        NaN     NaN"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    }
   ],
   "source": [
    "file_name = 'HeLa_chip'\n",
    "label_list = ['antibody', 'treatment', 'target']\n",
    "\n",
    "metadata = pd.read_csv(os.path.join('chip_peak_analysis', 'metadata', file_name + '.tsv'), sep = '\\t')\n",
    "display(metadata.head(1))\n",
    "metadata = utils.prepare_metadata(id_col = 'sample_accession', label_col = 'sample_title', sep = '-', label_list = label_list, df = metadata)\n",
    "\n",
    "display(metadata.head(3))\n",
    "metadata.to_csv(os.path.join('chip_peak_analysis', 'metadata', file_name + '.csv'), index = False)\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 12,
   "id": "88088e42",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>sample_accession</th>\n",
       "      <th>run_accession</th>\n",
       "      <th>fastq_ftp</th>\n",
       "      <th>study_accession</th>\n",
       "      <th>secondary_study_accession</th>\n",
       "      <th>experiment_accession</th>\n",
       "      <th>tax_id</th>\n",
       "      <th>scientific_name</th>\n",
       "      <th>library_layout</th>\n",
       "      <th>library_source</th>\n",
       "      <th>submitted_ftp</th>\n",
       "      <th>sra_ftp</th>\n",
       "      <th>sample_title</th>\n",
       "      <th>bam_ftp</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>SAMN09233681</td>\n",
       "      <td>['SRR7192344']</td>\n",
       "      <td>['ftp.sra.ebi.ac.uk/vol1/fastq/SRR719/004/SRR7...</td>\n",
       "      <td>PRJNA472394</td>\n",
       "      <td>SRP148605</td>\n",
       "      <td>SRX4108915</td>\n",
       "      <td>9606</td>\n",
       "      <td>Homo sapiens</td>\n",
       "      <td>SINGLE</td>\n",
       "      <td>GENOMIC</td>\n",
       "      <td>NaN</td>\n",
       "      <td>ftp.sra.ebi.ac.uk/vol1/srr/SRR719/004/SRR7192344</td>\n",
       "      <td>ET-N-AR</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "  sample_accession   run_accession  \\\n",
       "0     SAMN09233681  ['SRR7192344']   \n",
       "\n",
       "                                           fastq_ftp study_accession  \\\n",
       "0  ['ftp.sra.ebi.ac.uk/vol1/fastq/SRR719/004/SRR7...     PRJNA472394   \n",
       "\n",
       "  secondary_study_accession experiment_accession  tax_id scientific_name  \\\n",
       "0                 SRP148605           SRX4108915    9606    Homo sapiens   \n",
       "\n",
       "  library_layout library_source  submitted_ftp  \\\n",
       "0         SINGLE        GENOMIC            NaN   \n",
       "\n",
       "                                            sra_ftp sample_title  bam_ftp  \n",
       "0  ftp.sra.ebi.ac.uk/vol1/srr/SRR719/004/SRR7192344      ET-N-AR      NaN  "
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    },
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>sample_accession</th>\n",
       "      <th>antibody</th>\n",
       "      <th>treatment</th>\n",
       "      <th>target</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>SAMN09233681</td>\n",
       "      <td>ET</td>\n",
       "      <td>N</td>\n",
       "      <td>AR</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>SAMN09233682</td>\n",
       "      <td>AN</td>\n",
       "      <td>H</td>\n",
       "      <td>AR</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>SAMN09233679</td>\n",
       "      <td>AN</td>\n",
       "      <td>N</td>\n",
       "      <td>HIF</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "  sample_accession antibody treatment target\n",
       "0     SAMN09233681       ET         N     AR\n",
       "1     SAMN09233682       AN         H     AR\n",
       "2     SAMN09233679       AN         N    HIF"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    }
   ],
   "source": [
    "file_name = 'LNCaP_hypx_chip'\n",
    "label_list = ['antibody', 'treatment', 'target']\n",
    "\n",
    "metadata = pd.read_csv(os.path.join('data', 'metadata', file_name + '.tsv'), sep = '\\t')\n",
    "display(metadata.head(1))\n",
    "metadata = utils.prepare_metadata(id_col = 'sample_accession', label_col = 'sample_title', sep = '-', label_list = label_list, df = metadata)\n",
    "\n",
    "display(metadata.head(3))\n",
    "metadata.to_csv(os.path.join('data', 'metadata', file_name + '.csv'), index = False)\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 1,
   "id": "65bc961b",
   "metadata": {},
   "outputs": [
    {
     "ename": "NameError",
     "evalue": "name 'pd' is not defined",
     "output_type": "error",
     "traceback": [
      "\u001b[1;31m---------------------------------------------------------------------------\u001b[0m",
      "\u001b[1;31mNameError\u001b[0m                                 Traceback (most recent call last)",
      "Cell \u001b[1;32mIn[1], line 4\u001b[0m\n\u001b[0;32m      1\u001b[0m file_name \u001b[38;5;241m=\u001b[39m \u001b[38;5;124m'\u001b[39m\u001b[38;5;124mHS_HCT_proseq\u001b[39m\u001b[38;5;124m'\u001b[39m\n\u001b[0;32m      2\u001b[0m label_list \u001b[38;5;241m=\u001b[39m [\u001b[38;5;124m'\u001b[39m\u001b[38;5;124mmanipulation\u001b[39m\u001b[38;5;124m'\u001b[39m, \u001b[38;5;124m'\u001b[39m\u001b[38;5;124mtreatment\u001b[39m\u001b[38;5;124m'\u001b[39m, \u001b[38;5;124m'\u001b[39m\u001b[38;5;124mreplicate\u001b[39m\u001b[38;5;124m'\u001b[39m]\n\u001b[1;32m----> 4\u001b[0m metadata \u001b[38;5;241m=\u001b[39m \u001b[43mpd\u001b[49m\u001b[38;5;241m.\u001b[39mread_csv(os\u001b[38;5;241m.\u001b[39mpath\u001b[38;5;241m.\u001b[39mjoin(\u001b[38;5;124m'\u001b[39m\u001b[38;5;124mdata\u001b[39m\u001b[38;5;124m'\u001b[39m, \u001b[38;5;124m'\u001b[39m\u001b[38;5;124mmetadata\u001b[39m\u001b[38;5;124m'\u001b[39m, file_name \u001b[38;5;241m+\u001b[39m \u001b[38;5;124m'\u001b[39m\u001b[38;5;124m.tsv\u001b[39m\u001b[38;5;124m'\u001b[39m), sep \u001b[38;5;241m=\u001b[39m \u001b[38;5;124m'\u001b[39m\u001b[38;5;130;01m\\t\u001b[39;00m\u001b[38;5;124m'\u001b[39m)\n\u001b[0;32m      5\u001b[0m display(metadata\u001b[38;5;241m.\u001b[39mhead(\u001b[38;5;241m1\u001b[39m))\n\u001b[0;32m      6\u001b[0m metadata \u001b[38;5;241m=\u001b[39m utils\u001b[38;5;241m.\u001b[39mprepare_metadata(id_col \u001b[38;5;241m=\u001b[39m \u001b[38;5;124m'\u001b[39m\u001b[38;5;124mrun_accession\u001b[39m\u001b[38;5;124m'\u001b[39m, label_col \u001b[38;5;241m=\u001b[39m \u001b[38;5;124m'\u001b[39m\u001b[38;5;124msample_title\u001b[39m\u001b[38;5;124m'\u001b[39m, sep \u001b[38;5;241m=\u001b[39m \u001b[38;5;124m'\u001b[39m\u001b[38;5;124m,\u001b[39m\u001b[38;5;124m'\u001b[39m, label_list \u001b[38;5;241m=\u001b[39m label_list, df \u001b[38;5;241m=\u001b[39m metadata)\n",
      "\u001b[1;31mNameError\u001b[0m: name 'pd' is not defined"
     ]
    }
   ],
   "source": [
    "file_name = 'HS_HCT_proseq'\n",
    "label_list = ['manipulation', 'treatment', 'replicate']\n",
    "\n",
    "metadata = pd.read_csv(os.path.join('data', 'metadata', file_name + '.tsv'), sep = '\\t')\n",
    "display(metadata.head(1))\n",
    "metadata = utils.prepare_metadata(id_col = 'run_accession', label_col = 'sample_title', sep = ',', label_list = label_list, df = metadata)\n",
    "\n",
    "display(metadata.head(3))\n",
    "metadata.to_csv(os.path.join('data', 'metadata', file_name + '.csv'), index = False)\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 13,
   "id": "04a38325",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>sample_accession</th>\n",
       "      <th>run_accession</th>\n",
       "      <th>fastq_ftp</th>\n",
       "      <th>study_accession</th>\n",
       "      <th>secondary_study_accession</th>\n",
       "      <th>experiment_accession</th>\n",
       "      <th>tax_id</th>\n",
       "      <th>scientific_name</th>\n",
       "      <th>library_layout</th>\n",
       "      <th>library_source</th>\n",
       "      <th>run_alias</th>\n",
       "      <th>submitted_ftp</th>\n",
       "      <th>sra_ftp</th>\n",
       "      <th>sample_alias</th>\n",
       "      <th>sample_title</th>\n",
       "      <th>bam_ftp</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>SAMN32784734</td>\n",
       "      <td>['SRR23116233']</td>\n",
       "      <td>['ftp.sra.ebi.ac.uk/vol1/fastq/SRR231/033/SRR2...</td>\n",
       "      <td>PRJNA925199</td>\n",
       "      <td>SRP418187</td>\n",
       "      <td>SRX19067221</td>\n",
       "      <td>9606</td>\n",
       "      <td>Homo sapiens</td>\n",
       "      <td>PAIRED</td>\n",
       "      <td>GENOMIC</td>\n",
       "      <td>GSM6940724_r1</td>\n",
       "      <td>NaN</td>\n",
       "      <td>ftp.sra.ebi.ac.uk/vol1/srr/SRR231/033/SRR23116...</td>\n",
       "      <td>GSM6940724</td>\n",
       "      <td>MDAMB436_Norm_2</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "  sample_accession    run_accession  \\\n",
       "0     SAMN32784734  ['SRR23116233']   \n",
       "\n",
       "                                           fastq_ftp study_accession  \\\n",
       "0  ['ftp.sra.ebi.ac.uk/vol1/fastq/SRR231/033/SRR2...     PRJNA925199   \n",
       "\n",
       "  secondary_study_accession experiment_accession  tax_id scientific_name  \\\n",
       "0                 SRP418187          SRX19067221    9606    Homo sapiens   \n",
       "\n",
       "  library_layout library_source      run_alias  submitted_ftp  \\\n",
       "0         PAIRED        GENOMIC  GSM6940724_r1            NaN   \n",
       "\n",
       "                                             sra_ftp sample_alias  \\\n",
       "0  ftp.sra.ebi.ac.uk/vol1/srr/SRR231/033/SRR23116...   GSM6940724   \n",
       "\n",
       "      sample_title  bam_ftp  \n",
       "0  MDAMB436_Norm_2      NaN  "
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    },
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>sample_accession</th>\n",
       "      <th>cell_line</th>\n",
       "      <th>treatment</th>\n",
       "      <th>replicate</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>SAMN32784734</td>\n",
       "      <td>MDAMB436</td>\n",
       "      <td>Norm</td>\n",
       "      <td>2</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>SAMN32784754</td>\n",
       "      <td>MDAMB157</td>\n",
       "      <td>Norm</td>\n",
       "      <td>1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>SAMN32784755</td>\n",
       "      <td>MDAMB468</td>\n",
       "      <td>Norm</td>\n",
       "      <td>1</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "  sample_accession cell_line treatment replicate\n",
       "0     SAMN32784734  MDAMB436      Norm         2\n",
       "1     SAMN32784754  MDAMB157      Norm         1\n",
       "2     SAMN32784755  MDAMB468      Norm         1"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    }
   ],
   "source": [
    "file_name = 'TNBC_cells_hypx_atac'\n",
    "label_list = ['cell_line', 'treatment', 'replicate']\n",
    "\n",
    "metadata = pd.read_csv(os.path.join('data', 'metadata', file_name + '.tsv'), sep = '\\t')\n",
    "display(metadata.head(1))\n",
    "metadata = utils.prepare_metadata(id_col = 'sample_accession', label_col = 'sample_title', sep = '_', label_list = label_list, df = metadata)\n",
    "\n",
    "display(metadata.head(3))\n",
    "metadata.to_csv(os.path.join('data', 'metadata', file_name + '.csv'), index = False)\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 15,
   "id": "cf417408",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>sample_accession</th>\n",
       "      <th>run_accession</th>\n",
       "      <th>fastq_ftp</th>\n",
       "      <th>secondary_study_accession</th>\n",
       "      <th>experiment_accession</th>\n",
       "      <th>study_accession</th>\n",
       "      <th>tax_id</th>\n",
       "      <th>scientific_name</th>\n",
       "      <th>library_layout</th>\n",
       "      <th>library_source</th>\n",
       "      <th>experiment_alias</th>\n",
       "      <th>run_alias</th>\n",
       "      <th>submitted_ftp</th>\n",
       "      <th>sra_ftp</th>\n",
       "      <th>sample_alias</th>\n",
       "      <th>sample_title</th>\n",
       "      <th>bam_ftp</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>SAMN03569159</td>\n",
       "      <td>['SRR1999300']</td>\n",
       "      <td>['ftp.sra.ebi.ac.uk/vol1/fastq/SRR199/000/SRR1...</td>\n",
       "      <td>SRP057708</td>\n",
       "      <td>SRX1011329</td>\n",
       "      <td>PRJNA282396</td>\n",
       "      <td>9606</td>\n",
       "      <td>Homo sapiens</td>\n",
       "      <td>SINGLE</td>\n",
       "      <td>TRANSCRIPTOMIC</td>\n",
       "      <td>GSM1667173</td>\n",
       "      <td>GSM1667173_r1</td>\n",
       "      <td>NaN</td>\n",
       "      <td>ftp.sra.ebi.ac.uk/vol1/srr/SRR199/000/SRR1999300</td>\n",
       "      <td>GSM1667173</td>\n",
       "      <td>shNT, normoxia, replicate 1</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "  sample_accession   run_accession  \\\n",
       "0     SAMN03569159  ['SRR1999300']   \n",
       "\n",
       "                                           fastq_ftp  \\\n",
       "0  ['ftp.sra.ebi.ac.uk/vol1/fastq/SRR199/000/SRR1...   \n",
       "\n",
       "  secondary_study_accession experiment_accession study_accession  tax_id  \\\n",
       "0                 SRP057708           SRX1011329     PRJNA282396    9606   \n",
       "\n",
       "  scientific_name library_layout  library_source experiment_alias  \\\n",
       "0    Homo sapiens         SINGLE  TRANSCRIPTOMIC       GSM1667173   \n",
       "\n",
       "       run_alias  submitted_ftp  \\\n",
       "0  GSM1667173_r1            NaN   \n",
       "\n",
       "                                            sra_ftp sample_alias  \\\n",
       "0  ftp.sra.ebi.ac.uk/vol1/srr/SRR199/000/SRR1999300   GSM1667173   \n",
       "\n",
       "                  sample_title  bam_ftp  \n",
       "0  shNT, normoxia, replicate 1      NaN  "
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    },
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>sample_accession</th>\n",
       "      <th>manipulation</th>\n",
       "      <th>treatment</th>\n",
       "      <th>replicate</th>\n",
       "      <th>cell_line</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>SAMN03569159</td>\n",
       "      <td>shNT</td>\n",
       "      <td>normoxia</td>\n",
       "      <td>replicate 1</td>\n",
       "      <td>HCT-116</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>SAMN03569164</td>\n",
       "      <td>shCDK8</td>\n",
       "      <td>normoxia</td>\n",
       "      <td>replicate 2</td>\n",
       "      <td>HCT-116</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>SAMN03569166</td>\n",
       "      <td>HIF1A-/-</td>\n",
       "      <td>normoxia</td>\n",
       "      <td>replicate 2</td>\n",
       "      <td>HCT-116</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>SAMN03569170</td>\n",
       "      <td>shTIP60</td>\n",
       "      <td>hypoxia</td>\n",
       "      <td>replicate 2</td>\n",
       "      <td>HCT-116</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>SAMN03569171</td>\n",
       "      <td>shCDK8</td>\n",
       "      <td>hypoxia</td>\n",
       "      <td>replicate 1</td>\n",
       "      <td>HCT-116</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>5</th>\n",
       "      <td>SAMN03569173</td>\n",
       "      <td>HIF1A-/-</td>\n",
       "      <td>hypoxia</td>\n",
       "      <td>replicate 1</td>\n",
       "      <td>HCT-116</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>6</th>\n",
       "      <td>SAMN03569174</td>\n",
       "      <td>HIF1A-/-</td>\n",
       "      <td>hypoxia</td>\n",
       "      <td>replicate 2</td>\n",
       "      <td>HCT-116</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>7</th>\n",
       "      <td>SAMN03569160</td>\n",
       "      <td>shNT</td>\n",
       "      <td>normoxia</td>\n",
       "      <td>replicate 2</td>\n",
       "      <td>HCT-116</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>8</th>\n",
       "      <td>SAMN03569161</td>\n",
       "      <td>shTIP60</td>\n",
       "      <td>normoxia</td>\n",
       "      <td>replicate 1</td>\n",
       "      <td>HCT-116</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>9</th>\n",
       "      <td>SAMN03569162</td>\n",
       "      <td>shTIP60</td>\n",
       "      <td>normoxia</td>\n",
       "      <td>replicate 2</td>\n",
       "      <td>HCT-116</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "  sample_accession manipulation  treatment     replicate cell_line\n",
       "0     SAMN03569159         shNT   normoxia   replicate 1   HCT-116\n",
       "1     SAMN03569164       shCDK8   normoxia   replicate 2   HCT-116\n",
       "2     SAMN03569166     HIF1A-/-   normoxia   replicate 2   HCT-116\n",
       "3     SAMN03569170      shTIP60    hypoxia   replicate 2   HCT-116\n",
       "4     SAMN03569171       shCDK8    hypoxia   replicate 1   HCT-116\n",
       "5     SAMN03569173     HIF1A-/-    hypoxia   replicate 1   HCT-116\n",
       "6     SAMN03569174     HIF1A-/-    hypoxia   replicate 2   HCT-116\n",
       "7     SAMN03569160         shNT   normoxia   replicate 2   HCT-116\n",
       "8     SAMN03569161      shTIP60   normoxia   replicate 1   HCT-116\n",
       "9     SAMN03569162      shTIP60   normoxia   replicate 2   HCT-116"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    }
   ],
   "source": [
    "file_name = 'HCT-116_hypx_2'\n",
    "label_list = ['manipulation', 'treatment', 'replicate']\n",
    "\n",
    "metadata = pd.read_csv(os.path.join(PATH, 'data', 'metadata', file_name + '.tsv'), sep = '\\t')\n",
    "display(metadata.head(1))\n",
    "metadata = utils.prepare_metadata(id_col = 'sample_accession', label_col = 'sample_title', sep = ',', label_list = label_list, df = metadata)\n",
    "# Combine replicate and replicate_2\n",
    "#metadata = metadata.drop(['replicate_2'], axis = 1)\n",
    "\n",
    "metadata['cell_line'] = 'HCT-116'\n",
    "\n",
    "display(metadata.head(10))\n",
    "metadata.to_csv(os.path.join(PATH, 'data', 'metadata', file_name + '.csv'), index = False)\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 8,
   "id": "14403c4c",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>sample_accession</th>\n",
       "      <th>run_accession</th>\n",
       "      <th>fastq_ftp</th>\n",
       "      <th>secondary_study_accession</th>\n",
       "      <th>experiment_accession</th>\n",
       "      <th>study_accession</th>\n",
       "      <th>tax_id</th>\n",
       "      <th>scientific_name</th>\n",
       "      <th>library_layout</th>\n",
       "      <th>library_source</th>\n",
       "      <th>experiment_alias</th>\n",
       "      <th>run_alias</th>\n",
       "      <th>submitted_ftp</th>\n",
       "      <th>sra_ftp</th>\n",
       "      <th>sample_alias</th>\n",
       "      <th>sample_title</th>\n",
       "      <th>bam_ftp</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>SAMN10358938</td>\n",
       "      <td>['SRR8144252']</td>\n",
       "      <td>['ftp.sra.ebi.ac.uk/vol1/fastq/SRR814/002/SRR8...</td>\n",
       "      <td>SRP167432</td>\n",
       "      <td>SRX4964912</td>\n",
       "      <td>PRJNA503341</td>\n",
       "      <td>9606</td>\n",
       "      <td>Homo sapiens</td>\n",
       "      <td>SINGLE</td>\n",
       "      <td>TRANSCRIPTOMIC</td>\n",
       "      <td>GSM3453926</td>\n",
       "      <td>GSM3453926_r1</td>\n",
       "      <td>NaN</td>\n",
       "      <td>ftp.sra.ebi.ac.uk/vol1/srr/SRR814/002/SRR81442...</td>\n",
       "      <td>GSM3453926</td>\n",
       "      <td>wt 0h normoxia rep 1</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "  sample_accession   run_accession  \\\n",
       "0     SAMN10358938  ['SRR8144252']   \n",
       "\n",
       "                                           fastq_ftp  \\\n",
       "0  ['ftp.sra.ebi.ac.uk/vol1/fastq/SRR814/002/SRR8...   \n",
       "\n",
       "  secondary_study_accession experiment_accession study_accession  tax_id  \\\n",
       "0                 SRP167432           SRX4964912     PRJNA503341    9606   \n",
       "\n",
       "  scientific_name library_layout  library_source experiment_alias  \\\n",
       "0    Homo sapiens         SINGLE  TRANSCRIPTOMIC       GSM3453926   \n",
       "\n",
       "       run_alias  submitted_ftp  \\\n",
       "0  GSM3453926_r1            NaN   \n",
       "\n",
       "                                             sra_ftp sample_alias  \\\n",
       "0  ftp.sra.ebi.ac.uk/vol1/srr/SRR814/002/SRR81442...   GSM3453926   \n",
       "\n",
       "           sample_title  bam_ftp  \n",
       "0  wt 0h normoxia rep 1      NaN  "
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    },
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>sample_accession</th>\n",
       "      <th>time</th>\n",
       "      <th>treatment</th>\n",
       "      <th>replicate</th>\n",
       "      <th>cell_line</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>SAMN10358938</td>\n",
       "      <td>0h</td>\n",
       "      <td>normoxia</td>\n",
       "      <td>rep_1</td>\n",
       "      <td>MCF7</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>SAMN10358943</td>\n",
       "      <td>24h</td>\n",
       "      <td>hypoxia</td>\n",
       "      <td>rep_2</td>\n",
       "      <td>MCF7</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>SAMN10358955</td>\n",
       "      <td>0h</td>\n",
       "      <td>normoxia</td>\n",
       "      <td>rep_2</td>\n",
       "      <td>MCF7</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>SAMN10358942</td>\n",
       "      <td>24h</td>\n",
       "      <td>hypoxia</td>\n",
       "      <td>rep_3</td>\n",
       "      <td>MCF7</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>SAMN10358953</td>\n",
       "      <td>normoxia</td>\n",
       "      <td>rep</td>\n",
       "      <td>NaN</td>\n",
       "      <td>MCF7</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>5</th>\n",
       "      <td>SAMN10358952</td>\n",
       "      <td>normoxia</td>\n",
       "      <td>rep</td>\n",
       "      <td>NaN</td>\n",
       "      <td>MCF7</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>6</th>\n",
       "      <td>SAMN10358950</td>\n",
       "      <td>3h</td>\n",
       "      <td>hypoxia</td>\n",
       "      <td>rep_1</td>\n",
       "      <td>MCF7</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>7</th>\n",
       "      <td>SAMN10358949</td>\n",
       "      <td>3h</td>\n",
       "      <td>hypoxia</td>\n",
       "      <td>rep_2</td>\n",
       "      <td>MCF7</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>8</th>\n",
       "      <td>SAMN10358948</td>\n",
       "      <td>3h</td>\n",
       "      <td>hypoxia</td>\n",
       "      <td>rep_3</td>\n",
       "      <td>MCF7</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>9</th>\n",
       "      <td>SAMN10358947</td>\n",
       "      <td>3h</td>\n",
       "      <td>hypoxia</td>\n",
       "      <td>rep_1</td>\n",
       "      <td>MCF7</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "  sample_accession      time treatment replicate cell_line\n",
       "0     SAMN10358938        0h  normoxia     rep_1      MCF7\n",
       "1     SAMN10358943       24h   hypoxia     rep_2      MCF7\n",
       "2     SAMN10358955        0h  normoxia     rep_2      MCF7\n",
       "3     SAMN10358942       24h   hypoxia     rep_3      MCF7\n",
       "4     SAMN10358953  normoxia       rep       NaN      MCF7\n",
       "5     SAMN10358952  normoxia       rep       NaN      MCF7\n",
       "6     SAMN10358950        3h   hypoxia     rep_1      MCF7\n",
       "7     SAMN10358949        3h   hypoxia     rep_2      MCF7\n",
       "8     SAMN10358948        3h   hypoxia     rep_3      MCF7\n",
       "9     SAMN10358947        3h   hypoxia     rep_1      MCF7"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    }
   ],
   "source": [
    "file_name = 'MCF7_hypx_2'\n",
    "label_list = ['strain', 'time', 'treatment', 'replicate', 'replicate_2']\n",
    "\n",
    "metadata = pd.read_csv(os.path.join(PATH, 'data', 'metadata', file_name + '.tsv'), sep = '\\t')\n",
    "display(metadata.head(1))\n",
    "metadata = utils.prepare_metadata(id_col = 'sample_accession', label_col = 'sample_title', sep = ' ', label_list = label_list, df = metadata)\n",
    "# Combine replicate and replicate_2\n",
    "metadata['replicate'] = metadata['replicate'] + '_' + metadata['replicate_2']\n",
    "metadata = metadata.drop(['replicate_2', 'strain'], axis = 1)\n",
    "\n",
    "metadata['cell_line'] = 'MCF7'\n",
    "\n",
    "display(metadata.head(10))\n",
    "metadata.to_csv(os.path.join(PATH, 'data', 'metadata', file_name + '.csv'), index = False)\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 14,
   "id": "5851eeed",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>sample_accession</th>\n",
       "      <th>run_accession</th>\n",
       "      <th>fastq_ftp</th>\n",
       "      <th>secondary_study_accession</th>\n",
       "      <th>experiment_accession</th>\n",
       "      <th>study_accession</th>\n",
       "      <th>tax_id</th>\n",
       "      <th>scientific_name</th>\n",
       "      <th>library_layout</th>\n",
       "      <th>library_source</th>\n",
       "      <th>experiment_alias</th>\n",
       "      <th>run_alias</th>\n",
       "      <th>submitted_ftp</th>\n",
       "      <th>sra_ftp</th>\n",
       "      <th>sample_alias</th>\n",
       "      <th>sample_title</th>\n",
       "      <th>bam_ftp</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>SAMN40377576</td>\n",
       "      <td>['SRR28290277']</td>\n",
       "      <td>['ftp.sra.ebi.ac.uk/vol1/fastq/SRR282/077/SRR2...</td>\n",
       "      <td>SRP494524</td>\n",
       "      <td>SRX23899525</td>\n",
       "      <td>PRJNA1086471</td>\n",
       "      <td>9606</td>\n",
       "      <td>Homo sapiens</td>\n",
       "      <td>PAIRED</td>\n",
       "      <td>TRANSCRIPTOMIC</td>\n",
       "      <td>GSM8140396_r1</td>\n",
       "      <td>GSM8140396_r1</td>\n",
       "      <td>NaN</td>\n",
       "      <td>ftp.sra.ebi.ac.uk/vol1/srr/SRR282/077/SRR28290...</td>\n",
       "      <td>GSM8140396</td>\n",
       "      <td>MDA 231 cells, Lactate 3</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "  sample_accession    run_accession  \\\n",
       "0     SAMN40377576  ['SRR28290277']   \n",
       "\n",
       "                                           fastq_ftp  \\\n",
       "0  ['ftp.sra.ebi.ac.uk/vol1/fastq/SRR282/077/SRR2...   \n",
       "\n",
       "  secondary_study_accession experiment_accession study_accession  tax_id  \\\n",
       "0                 SRP494524          SRX23899525    PRJNA1086471    9606   \n",
       "\n",
       "  scientific_name library_layout  library_source experiment_alias  \\\n",
       "0    Homo sapiens         PAIRED  TRANSCRIPTOMIC    GSM8140396_r1   \n",
       "\n",
       "       run_alias  submitted_ftp  \\\n",
       "0  GSM8140396_r1            NaN   \n",
       "\n",
       "                                             sra_ftp sample_alias  \\\n",
       "0  ftp.sra.ebi.ac.uk/vol1/srr/SRR282/077/SRR28290...   GSM8140396   \n",
       "\n",
       "               sample_title  bam_ftp  \n",
       "0  MDA 231 cells, Lactate 3      NaN  "
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    },
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>sample_accession</th>\n",
       "      <th>treatment</th>\n",
       "      <th>replicate</th>\n",
       "      <th>cell_line</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>SAMN40377576</td>\n",
       "      <td>Lactate</td>\n",
       "      <td>3</td>\n",
       "      <td>MDA_231</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>SAMN40377579</td>\n",
       "      <td>Oscillatory</td>\n",
       "      <td>Hypoxia</td>\n",
       "      <td>MDA_231</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>SAMN40377580</td>\n",
       "      <td>Oscillatory</td>\n",
       "      <td>Hypoxia</td>\n",
       "      <td>MDA_231</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>SAMN40377581</td>\n",
       "      <td>Oscillatory</td>\n",
       "      <td>Hypoxia</td>\n",
       "      <td>MDA_231</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>SAMN40377577</td>\n",
       "      <td>Lactate</td>\n",
       "      <td>2</td>\n",
       "      <td>MDA_231</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>5</th>\n",
       "      <td>SAMN40377578</td>\n",
       "      <td>Lactate</td>\n",
       "      <td>1</td>\n",
       "      <td>MDA_231</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>6</th>\n",
       "      <td>SAMN40377583</td>\n",
       "      <td>Hypoxia</td>\n",
       "      <td>2</td>\n",
       "      <td>MDA_231</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>7</th>\n",
       "      <td>SAMN40377584</td>\n",
       "      <td>Hypoxia</td>\n",
       "      <td>1</td>\n",
       "      <td>MDA_231</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>8</th>\n",
       "      <td>SAMN40377585</td>\n",
       "      <td>Control</td>\n",
       "      <td>3</td>\n",
       "      <td>MDA_231</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>9</th>\n",
       "      <td>SAMN40377582</td>\n",
       "      <td>Hypoxia</td>\n",
       "      <td>3</td>\n",
       "      <td>MDA_231</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "  sample_accession    treatment replicate cell_line\n",
       "0     SAMN40377576      Lactate         3   MDA_231\n",
       "1     SAMN40377579  Oscillatory   Hypoxia   MDA_231\n",
       "2     SAMN40377580  Oscillatory   Hypoxia   MDA_231\n",
       "3     SAMN40377581  Oscillatory   Hypoxia   MDA_231\n",
       "4     SAMN40377577      Lactate         2   MDA_231\n",
       "5     SAMN40377578      Lactate         1   MDA_231\n",
       "6     SAMN40377583      Hypoxia         2   MDA_231\n",
       "7     SAMN40377584      Hypoxia         1   MDA_231\n",
       "8     SAMN40377585      Control         3   MDA_231\n",
       "9     SAMN40377582      Hypoxia         3   MDA_231"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    }
   ],
   "source": [
    "file_name = 'MDA_231_hypx_1'\n",
    "label_list = ['cells', 'cells2', 'cells3', 'treatment', 'replicate']\n",
    "\n",
    "metadata = pd.read_csv(os.path.join(PATH, 'data', 'metadata', file_name + '.tsv'), sep = '\\t')\n",
    "display(metadata.head(1))\n",
    "metadata = utils.prepare_metadata(id_col = 'sample_accession', label_col = 'sample_title', sep = ' ', label_list = label_list, df = metadata)\n",
    "metadata['cell_line'] = 'MDA_231'\n",
    "metadata.drop(['cells', 'cells2', 'cells3'], axis = 1, inplace = True)\n",
    "\n",
    "display(metadata.head(10))\n",
    "metadata.to_csv(os.path.join(PATH, 'data', 'metadata', file_name + '.csv'), index = False)\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 8,
   "id": "ed42b321",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>sample_accession</th>\n",
       "      <th>run_accession</th>\n",
       "      <th>fastq_ftp</th>\n",
       "      <th>study_accession</th>\n",
       "      <th>secondary_study_accession</th>\n",
       "      <th>experiment_accession</th>\n",
       "      <th>tax_id</th>\n",
       "      <th>scientific_name</th>\n",
       "      <th>library_layout</th>\n",
       "      <th>library_source</th>\n",
       "      <th>run_alias</th>\n",
       "      <th>submitted_ftp</th>\n",
       "      <th>sra_ftp</th>\n",
       "      <th>sample_alias</th>\n",
       "      <th>sample_title</th>\n",
       "      <th>bam_ftp</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>SAMN30164526</td>\n",
       "      <td>['SRR20853991']</td>\n",
       "      <td>['ftp.sra.ebi.ac.uk/vol1/fastq/SRR208/091/SRR2...</td>\n",
       "      <td>PRJNA866399</td>\n",
       "      <td>SRP390136</td>\n",
       "      <td>SRX16873197</td>\n",
       "      <td>9606</td>\n",
       "      <td>Homo sapiens</td>\n",
       "      <td>PAIRED</td>\n",
       "      <td>GENOMIC</td>\n",
       "      <td>GSM6433321_r1</td>\n",
       "      <td>NaN</td>\n",
       "      <td>ftp.sra.ebi.ac.uk/vol1/srr/SRR208/091/SRR20853991</td>\n",
       "      <td>GSM6433321</td>\n",
       "      <td>ATAC-seq SKOV3 shPLD2 hypoxia, replicate 1</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "  sample_accession    run_accession  \\\n",
       "0     SAMN30164526  ['SRR20853991']   \n",
       "\n",
       "                                           fastq_ftp study_accession  \\\n",
       "0  ['ftp.sra.ebi.ac.uk/vol1/fastq/SRR208/091/SRR2...     PRJNA866399   \n",
       "\n",
       "  secondary_study_accession experiment_accession  tax_id scientific_name  \\\n",
       "0                 SRP390136          SRX16873197    9606    Homo sapiens   \n",
       "\n",
       "  library_layout library_source      run_alias  submitted_ftp  \\\n",
       "0         PAIRED        GENOMIC  GSM6433321_r1            NaN   \n",
       "\n",
       "                                             sra_ftp sample_alias  \\\n",
       "0  ftp.sra.ebi.ac.uk/vol1/srr/SRR208/091/SRR20853991   GSM6433321   \n",
       "\n",
       "                                 sample_title  bam_ftp  \n",
       "0  ATAC-seq SKOV3 shPLD2 hypoxia, replicate 1      NaN  "
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    },
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>sample_accession</th>\n",
       "      <th>method</th>\n",
       "      <th>cell_line</th>\n",
       "      <th>manipulation</th>\n",
       "      <th>treatment</th>\n",
       "      <th>replicate</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>SAMN30164526</td>\n",
       "      <td>ATAC-seq</td>\n",
       "      <td>SKOV3</td>\n",
       "      <td>shPLD2</td>\n",
       "      <td>replicate</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>SAMN30164525</td>\n",
       "      <td>ATAC-seq</td>\n",
       "      <td>SKOV3</td>\n",
       "      <td>shPLD2</td>\n",
       "      <td>replicate</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>SAMN30164530</td>\n",
       "      <td>ATAC-seq</td>\n",
       "      <td>SKOV3</td>\n",
       "      <td>empty</td>\n",
       "      <td>normoxia,</td>\n",
       "      <td>replicate_1</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "  sample_accession    method cell_line manipulation  treatment    replicate\n",
       "0     SAMN30164526  ATAC-seq     SKOV3       shPLD2  replicate          NaN\n",
       "1     SAMN30164525  ATAC-seq     SKOV3       shPLD2  replicate          NaN\n",
       "2     SAMN30164530  ATAC-seq     SKOV3        empty  normoxia,  replicate_1"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    }
   ],
   "source": [
    "file_name = 'SKOV3_hypx_atac'\n",
    "label_list = ['method', 'cell_line', 'manipulation', 'x', 'treatment', 'replicate1', 'replicate2']\n",
    "\n",
    "metadata = pd.read_csv(os.path.join('data', 'metadata', file_name + '.tsv'), sep = '\\t')\n",
    "display(metadata.head(1))\n",
    "metadata = utils.prepare_metadata(id_col = 'sample_accession', label_col = 'sample_title', sep = ' ', label_list = label_list, df = metadata)\n",
    "\n",
    "# Drop x\n",
    "metadata = metadata.drop(['x'], axis = 1)\n",
    "# Combine replicate1 and replicate2\n",
    "metadata['replicate'] = metadata['replicate1'] + '_' + metadata['replicate2']\n",
    "metadata = metadata.drop(['replicate1', 'replicate2'], axis = 1)\n",
    "display(metadata.head(3))\n",
    "metadata.to_csv(os.path.join('data', 'metadata', file_name + '.csv'), index = False)\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 7,
   "id": "429f3aa3",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>sample_accession</th>\n",
       "      <th>run_accession</th>\n",
       "      <th>fastq_ftp</th>\n",
       "      <th>study_accession</th>\n",
       "      <th>secondary_study_accession</th>\n",
       "      <th>experiment_accession</th>\n",
       "      <th>tax_id</th>\n",
       "      <th>scientific_name</th>\n",
       "      <th>library_layout</th>\n",
       "      <th>library_source</th>\n",
       "      <th>run_alias</th>\n",
       "      <th>submitted_ftp</th>\n",
       "      <th>sra_ftp</th>\n",
       "      <th>sample_alias</th>\n",
       "      <th>sample_title</th>\n",
       "      <th>bam_ftp</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>SAMN08683145</td>\n",
       "      <td>['SRR6822802']</td>\n",
       "      <td>['ftp.sra.ebi.ac.uk/vol1/fastq/SRR682/002/SRR6...</td>\n",
       "      <td>PRJNA437670</td>\n",
       "      <td>SRP134389</td>\n",
       "      <td>SRX3779341</td>\n",
       "      <td>9606</td>\n",
       "      <td>Homo sapiens</td>\n",
       "      <td>PAIRED</td>\n",
       "      <td>TRANSCRIPTOMIC</td>\n",
       "      <td>GSM3036637_r1</td>\n",
       "      <td>NaN</td>\n",
       "      <td>ftp.sra.ebi.ac.uk/vol1/srr/SRR682/002/SRR6822802</td>\n",
       "      <td>GSM3036637</td>\n",
       "      <td>BT549_20percentO2</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "  sample_accession   run_accession  \\\n",
       "0     SAMN08683145  ['SRR6822802']   \n",
       "\n",
       "                                           fastq_ftp study_accession  \\\n",
       "0  ['ftp.sra.ebi.ac.uk/vol1/fastq/SRR682/002/SRR6...     PRJNA437670   \n",
       "\n",
       "  secondary_study_accession experiment_accession  tax_id scientific_name  \\\n",
       "0                 SRP134389           SRX3779341    9606    Homo sapiens   \n",
       "\n",
       "  library_layout  library_source      run_alias  submitted_ftp  \\\n",
       "0         PAIRED  TRANSCRIPTOMIC  GSM3036637_r1            NaN   \n",
       "\n",
       "                                            sra_ftp sample_alias  \\\n",
       "0  ftp.sra.ebi.ac.uk/vol1/srr/SRR682/002/SRR6822802   GSM3036637   \n",
       "\n",
       "        sample_title  bam_ftp  \n",
       "0  BT549_20percentO2      NaN  "
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    },
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>sample_accession</th>\n",
       "      <th>cell_line</th>\n",
       "      <th>treatment</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>SAMN08683145</td>\n",
       "      <td>BT549</td>\n",
       "      <td>20percentO2</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>SAMN08683143</td>\n",
       "      <td>CAMA1</td>\n",
       "      <td>20percentO2</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>SAMN08683140</td>\n",
       "      <td>HBL100</td>\n",
       "      <td>1percentO2</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "  sample_accession cell_line    treatment\n",
       "0     SAMN08683145     BT549  20percentO2\n",
       "1     SAMN08683143     CAMA1  20percentO2\n",
       "2     SAMN08683140    HBL100   1percentO2"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    }
   ],
   "source": [
    "file_name = 'bc_mix_hypx'\n",
    "label_list = ['cell_line', 'treatment']\n",
    "\n",
    "metadata = pd.read_csv(os.path.join('data', 'metadata', file_name + '.tsv'), sep = '\\t')\n",
    "display(metadata.head(1))\n",
    "metadata = utils.prepare_metadata(id_col = 'sample_accession', label_col = 'sample_title', sep = '_', label_list = label_list, df = metadata)\n",
    "\n",
    "display(metadata.head(3))\n",
    "metadata.to_csv(os.path.join('data', 'metadata', file_name + '.csv'), index = False)\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 7,
   "id": "87036c77",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>sample_accession</th>\n",
       "      <th>run_accession</th>\n",
       "      <th>fastq_ftp</th>\n",
       "      <th>study_accession</th>\n",
       "      <th>secondary_study_accession</th>\n",
       "      <th>experiment_accession</th>\n",
       "      <th>tax_id</th>\n",
       "      <th>scientific_name</th>\n",
       "      <th>library_layout</th>\n",
       "      <th>library_source</th>\n",
       "      <th>run_alias</th>\n",
       "      <th>submitted_ftp</th>\n",
       "      <th>sra_ftp</th>\n",
       "      <th>sample_title</th>\n",
       "      <th>bam_ftp</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>SRR10743072</td>\n",
       "      <td>['SRR10743072']</td>\n",
       "      <td>['ftp.sra.ebi.ac.uk/vol1/fastq/SRR107/072/SRR1...</td>\n",
       "      <td>PRJNA596881</td>\n",
       "      <td>SRP238299</td>\n",
       "      <td>SRX7418487</td>\n",
       "      <td>9606</td>\n",
       "      <td>Homo sapiens</td>\n",
       "      <td>PAIRED</td>\n",
       "      <td>GENOMIC</td>\n",
       "      <td>Nor-Ctrl-2-CGTACTAG_S2_R1_001.fastq.gz</td>\n",
       "      <td>NaN</td>\n",
       "      <td>ftp.sra.ebi.ac.uk/vol1/srr/SRR107/072/SRR10743072</td>\n",
       "      <td>Human sample from Homo sapiens</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "  sample_accession    run_accession  \\\n",
       "0      SRR10743072  ['SRR10743072']   \n",
       "\n",
       "                                           fastq_ftp study_accession  \\\n",
       "0  ['ftp.sra.ebi.ac.uk/vol1/fastq/SRR107/072/SRR1...     PRJNA596881   \n",
       "\n",
       "  secondary_study_accession experiment_accession  tax_id scientific_name  \\\n",
       "0                 SRP238299           SRX7418487    9606    Homo sapiens   \n",
       "\n",
       "  library_layout library_source                               run_alias  \\\n",
       "0         PAIRED        GENOMIC  Nor-Ctrl-2-CGTACTAG_S2_R1_001.fastq.gz   \n",
       "\n",
       "   submitted_ftp                                            sra_ftp  \\\n",
       "0            NaN  ftp.sra.ebi.ac.uk/vol1/srr/SRR107/072/SRR10743072   \n",
       "\n",
       "                     sample_title  bam_ftp  \n",
       "0  Human sample from Homo sapiens      NaN  "
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    },
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>sample_accession</th>\n",
       "      <th>treatment</th>\n",
       "      <th>manipulation</th>\n",
       "      <th>replicate</th>\n",
       "      <th>d</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>SRR10743072</td>\n",
       "      <td>Nor</td>\n",
       "      <td>Ctrl</td>\n",
       "      <td>2</td>\n",
       "      <td>CGTACTAG_S2_R1_001.fastq.gz</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>SRR10743068</td>\n",
       "      <td>Hyp</td>\n",
       "      <td>Ac</td>\n",
       "      <td>2</td>\n",
       "      <td>TAGGCATG_S6_R1_001.fastq.gz</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>SRR10743070</td>\n",
       "      <td>Hyp</td>\n",
       "      <td>Ctrl</td>\n",
       "      <td>2</td>\n",
       "      <td>CCGTTTGT_S4_R1_001.fastq.gz</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "  sample_accession treatment manipulation replicate  \\\n",
       "0      SRR10743072       Nor         Ctrl         2   \n",
       "1      SRR10743068       Hyp           Ac         2   \n",
       "2      SRR10743070       Hyp         Ctrl         2   \n",
       "\n",
       "                             d  \n",
       "0  CGTACTAG_S2_R1_001.fastq.gz  \n",
       "1  TAGGCATG_S6_R1_001.fastq.gz  \n",
       "2  CCGTTTGT_S4_R1_001.fastq.gz  "
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    }
   ],
   "source": [
    "file_name = 'CHP134_hypx_atac'\n",
    "label_list = ['treatment', 'manipulation', 'replicate', 'd']\n",
    "\n",
    "metadata = pd.read_csv(os.path.join('data', 'metadata', file_name + '.tsv'), sep = '\\t')\n",
    "display(metadata.head(1))\n",
    "metadata = utils.prepare_metadata(id_col = 'sample_accession', label_col = 'run_alias', sep = '-', label_list = label_list, df = metadata)\n",
    "\n",
    "display(metadata.head(3))\n",
    "metadata.to_csv(os.path.join('data', 'metadata', file_name + '.csv'), index = False)\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 13,
   "id": "f3254826",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>sample_accession</th>\n",
       "      <th>run_accession</th>\n",
       "      <th>fastq_ftp</th>\n",
       "      <th>study_accession</th>\n",
       "      <th>experiment_accession</th>\n",
       "      <th>tax_id</th>\n",
       "      <th>scientific_name</th>\n",
       "      <th>library_layout</th>\n",
       "      <th>library_source</th>\n",
       "      <th>submitted_ftp</th>\n",
       "      <th>sra_ftp</th>\n",
       "      <th>sample_title</th>\n",
       "      <th>bam_ftp</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>SAMN03078837</td>\n",
       "      <td>['SRR1585478', 'SRR1585479']</td>\n",
       "      <td>['ftp.sra.ebi.ac.uk/vol1/fastq/SRR158/008/SRR1...</td>\n",
       "      <td>PRJNA262005</td>\n",
       "      <td>SRX709826</td>\n",
       "      <td>9606</td>\n",
       "      <td>Homo sapiens</td>\n",
       "      <td>PAIRED</td>\n",
       "      <td>GENOMIC</td>\n",
       "      <td>NaN</td>\n",
       "      <td>ftp.sra.ebi.ac.uk/vol1/srr/SRR158/008/SRR1585478</td>\n",
       "      <td>H3K27me3 normoxia</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "  sample_accession                 run_accession  \\\n",
       "0     SAMN03078837  ['SRR1585478', 'SRR1585479']   \n",
       "\n",
       "                                           fastq_ftp study_accession  \\\n",
       "0  ['ftp.sra.ebi.ac.uk/vol1/fastq/SRR158/008/SRR1...     PRJNA262005   \n",
       "\n",
       "  experiment_accession  tax_id scientific_name library_layout library_source  \\\n",
       "0            SRX709826    9606    Homo sapiens         PAIRED        GENOMIC   \n",
       "\n",
       "   submitted_ftp                                           sra_ftp  \\\n",
       "0            NaN  ftp.sra.ebi.ac.uk/vol1/srr/SRR158/008/SRR1585478   \n",
       "\n",
       "        sample_title  bam_ftp  \n",
       "0  H3K27me3 normoxia      NaN  "
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    },
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>sample_accession</th>\n",
       "      <th>target</th>\n",
       "      <th>treatment</th>\n",
       "      <th>time</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>SAMN03078837</td>\n",
       "      <td>H3K27me3</td>\n",
       "      <td>normoxia</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>SAMN03078838</td>\n",
       "      <td>H3K27me3</td>\n",
       "      <td>hypoxia</td>\n",
       "      <td>8</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>SAMN03078839</td>\n",
       "      <td>H3K27me3</td>\n",
       "      <td>hypoxia</td>\n",
       "      <td>24</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "  sample_accession    target treatment time\n",
       "0     SAMN03078837  H3K27me3  normoxia    0\n",
       "1     SAMN03078838  H3K27me3   hypoxia    8\n",
       "2     SAMN03078839  H3K27me3   hypoxia   24"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    }
   ],
   "source": [
    "file_name = 'MCF7_hypx_chip1'\n",
    "label_list = ['target', 'treatment', 'time', 'remove']\n",
    "\n",
    "metadata = pd.read_csv(os.path.join('data', 'metadata', file_name + '.tsv'), sep = '\\t')\n",
    "display(metadata.head(1))\n",
    "metadata = utils.prepare_metadata(id_col = 'sample_accession', label_col = 'sample_title', sep = ' ', label_list = label_list, df = metadata)\n",
    "metadata['time'] = metadata['time'].fillna('0')\n",
    "metadata = metadata.drop(['remove'], axis = 1)\n",
    "display(metadata.head(3))\n",
    "metadata.to_csv(os.path.join('data', 'metadata', file_name + '.csv'), index = False)\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 9,
   "id": "c583a2cb",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>sample_accession</th>\n",
       "      <th>run_accession</th>\n",
       "      <th>fastq_ftp</th>\n",
       "      <th>study_accession</th>\n",
       "      <th>experiment_accession</th>\n",
       "      <th>tax_id</th>\n",
       "      <th>scientific_name</th>\n",
       "      <th>library_layout</th>\n",
       "      <th>library_source</th>\n",
       "      <th>submitted_ftp</th>\n",
       "      <th>sra_ftp</th>\n",
       "      <th>sample_title</th>\n",
       "      <th>bam_ftp</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>SAMN03876576</td>\n",
       "      <td>['SRR2105961', 'SRR2105962']</td>\n",
       "      <td>['ftp.sra.ebi.ac.uk/vol1/fastq/SRR210/001/SRR2...</td>\n",
       "      <td>PRJNA290150</td>\n",
       "      <td>SRX1099916</td>\n",
       "      <td>9606</td>\n",
       "      <td>Homo sapiens</td>\n",
       "      <td>PAIRED</td>\n",
       "      <td>GENOMIC</td>\n",
       "      <td>NaN</td>\n",
       "      <td>ftp.sra.ebi.ac.uk/vol1/srr/SRR210/001/SRR2105961</td>\n",
       "      <td>H3K27me3 normoxia</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "  sample_accession                 run_accession  \\\n",
       "0     SAMN03876576  ['SRR2105961', 'SRR2105962']   \n",
       "\n",
       "                                           fastq_ftp study_accession  \\\n",
       "0  ['ftp.sra.ebi.ac.uk/vol1/fastq/SRR210/001/SRR2...     PRJNA290150   \n",
       "\n",
       "  experiment_accession  tax_id scientific_name library_layout library_source  \\\n",
       "0           SRX1099916    9606    Homo sapiens         PAIRED        GENOMIC   \n",
       "\n",
       "   submitted_ftp                                           sra_ftp  \\\n",
       "0            NaN  ftp.sra.ebi.ac.uk/vol1/srr/SRR210/001/SRR2105961   \n",
       "\n",
       "        sample_title  bam_ftp  \n",
       "0  H3K27me3 normoxia      NaN  "
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    },
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>sample_accession</th>\n",
       "      <th>target</th>\n",
       "      <th>treatment</th>\n",
       "      <th>time</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>SAMN03876576</td>\n",
       "      <td>H3K27me3</td>\n",
       "      <td>normoxia</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>SAMN03876579</td>\n",
       "      <td>H3K27me3</td>\n",
       "      <td>reoxygenation</td>\n",
       "      <td>8</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>SAMN03876577</td>\n",
       "      <td>H3K27me3</td>\n",
       "      <td>hypoxia</td>\n",
       "      <td>8</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "  sample_accession    target      treatment time\n",
       "0     SAMN03876576  H3K27me3       normoxia    0\n",
       "1     SAMN03876579  H3K27me3  reoxygenation    8\n",
       "2     SAMN03876577  H3K27me3        hypoxia    8"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    }
   ],
   "source": [
    "file_name = 'MCF7_hypx_chip2'\n",
    "label_list = ['target', 'treatment', 'time', 'remove']\n",
    "\n",
    "metadata = pd.read_csv(os.path.join('data', 'metadata', file_name + '.tsv'), sep = '\\t')\n",
    "display(metadata.head(1))\n",
    "metadata = utils.prepare_metadata(id_col = 'sample_accession', label_col = 'sample_title', sep = ' ', label_list = label_list, df = metadata)\n",
    "metadata['time'] = metadata['time'].fillna(0)\n",
    "metadata = metadata.drop(['remove'], axis = 1)\n",
    "display(metadata.head(3))\n",
    "metadata.to_csv(os.path.join('data', 'metadata', file_name + '.csv'), index = False)\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 23,
   "id": "f809f313",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>sample_accession</th>\n",
       "      <th>run_accession</th>\n",
       "      <th>fastq_ftp</th>\n",
       "      <th>study_accession</th>\n",
       "      <th>experiment_accession</th>\n",
       "      <th>tax_id</th>\n",
       "      <th>scientific_name</th>\n",
       "      <th>library_layout</th>\n",
       "      <th>library_source</th>\n",
       "      <th>submitted_ftp</th>\n",
       "      <th>sra_ftp</th>\n",
       "      <th>sample_title</th>\n",
       "      <th>bam_ftp</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>SAMN02189823</td>\n",
       "      <td>['SRR873384']</td>\n",
       "      <td>['ftp.sra.ebi.ac.uk/vol1/fastq/SRR873/SRR87338...</td>\n",
       "      <td>PRJNA206504</td>\n",
       "      <td>SRX290633</td>\n",
       "      <td>9606</td>\n",
       "      <td>Homo sapiens</td>\n",
       "      <td>SINGLE</td>\n",
       "      <td>TRANSCRIPTOMIC</td>\n",
       "      <td>NaN</td>\n",
       "      <td>ftp.sra.ebi.ac.uk/vol1/srr/SRR873/SRR873384</td>\n",
       "      <td>MCF7_Hypoxia_16h_microRNAseq_rep1</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "  sample_accession  run_accession  \\\n",
       "0     SAMN02189823  ['SRR873384']   \n",
       "\n",
       "                                           fastq_ftp study_accession  \\\n",
       "0  ['ftp.sra.ebi.ac.uk/vol1/fastq/SRR873/SRR87338...     PRJNA206504   \n",
       "\n",
       "  experiment_accession  tax_id scientific_name library_layout  library_source  \\\n",
       "0            SRX290633    9606    Homo sapiens         SINGLE  TRANSCRIPTOMIC   \n",
       "\n",
       "   submitted_ftp                                      sra_ftp  \\\n",
       "0            NaN  ftp.sra.ebi.ac.uk/vol1/srr/SRR873/SRR873384   \n",
       "\n",
       "                        sample_title  bam_ftp  \n",
       "0  MCF7_Hypoxia_16h_microRNAseq_rep1      NaN  "
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    },
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>sample_accession</th>\n",
       "      <th>cell_line</th>\n",
       "      <th>treatment</th>\n",
       "      <th>time</th>\n",
       "      <th>method</th>\n",
       "      <th>replicate</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>SAMN02189823</td>\n",
       "      <td>MCF7</td>\n",
       "      <td>Hypoxia</td>\n",
       "      <td>16h</td>\n",
       "      <td>microRNAseq</td>\n",
       "      <td>rep1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>SAMN02189824</td>\n",
       "      <td>MCF7</td>\n",
       "      <td>Hypoxia</td>\n",
       "      <td>16h</td>\n",
       "      <td>microRNAseq</td>\n",
       "      <td>rep2</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>SAMN02189827</td>\n",
       "      <td>MCF7</td>\n",
       "      <td>Hypoxia</td>\n",
       "      <td>48h</td>\n",
       "      <td>microRNAseq</td>\n",
       "      <td>rep1</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "  sample_accession cell_line treatment time       method replicate\n",
       "0     SAMN02189823      MCF7   Hypoxia  16h  microRNAseq      rep1\n",
       "1     SAMN02189824      MCF7   Hypoxia  16h  microRNAseq      rep2\n",
       "2     SAMN02189827      MCF7   Hypoxia  48h  microRNAseq      rep1"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    }
   ],
   "source": [
    "file_name = 'MCF7_hypx_mirna'\n",
    "label_list = ['cell_line', 'treatment', 'time', 'method', 'replicate']\n",
    "\n",
    "metadata = pd.read_csv(os.path.join('data', 'metadata', file_name + '.tsv'), sep = '\\t')\n",
    "display(metadata.head(1))\n",
    "metadata = utils.prepare_metadata(id_col = 'sample_accession', label_col = 'sample_title', sep = '_', label_list = label_list, df = metadata)\n",
    "display(metadata.head(3))\n",
    "metadata.to_csv(os.path.join('data', 'metadata', file_name + '.csv'), index = False)\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 8,
   "id": "181fed81",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>Sample</th>\n",
       "      <th>Sample_Name</th>\n",
       "      <th>Treatment</th>\n",
       "      <th>cell_line</th>\n",
       "      <th>replicate</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>MAC</td>\n",
       "      <td>DM12</td>\n",
       "      <td>Control</td>\n",
       "      <td>MCF7</td>\n",
       "      <td>1</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "  Sample Sample_Name Treatment cell_line  replicate\n",
       "0    MAC        DM12   Control      MCF7          1"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    },
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>Sample_Name</th>\n",
       "      <th>Treatment</th>\n",
       "      <th>cell_line</th>\n",
       "      <th>replicate</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "   Sample_Name  Treatment  cell_line  replicate\n",
       "0          NaN        NaN        NaN        NaN\n",
       "1          NaN        NaN        NaN        NaN\n",
       "2          NaN        NaN        NaN        NaN"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    }
   ],
   "source": [
    "file_name = 'RNA-seq_MCF7_HCT116'\n",
    "label_list = ['Sample', 'Sample_Name', 'Treatment', 'cell_line', 'replicate']\n",
    "\n",
    "metadata = pd.read_csv(os.path.join('data', 'metadata', file_name + '.csv'), sep = ',')\n",
    "display(metadata.head(1))\n",
    "metadata = utils.prepare_metadata(id_col = 'Sample_Name', label_col = 'Sample', sep = '_', label_list = label_list, df = metadata)\n",
    "display(metadata.head(3))\n",
    "metadata.to_csv(os.path.join('data', 'metadata', file_name + '.csv'), index = False)\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 9,
   "id": "8252a799",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>scientific_name</th>\n",
       "      <th>fastq_ftp</th>\n",
       "      <th>run_accession</th>\n",
       "      <th>sample_title</th>\n",
       "      <th>library_strategy</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>Homo_sapiens</td>\n",
       "      <td>AC_S1_L001_R1_001;AC_S1_L001_R2_001</td>\n",
       "      <td>AC_S1_L001_001</td>\n",
       "      <td>AC_S1</td>\n",
       "      <td>RNA-seq</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "  scientific_name                            fastq_ftp   run_accession   \n",
       "0    Homo_sapiens  AC_S1_L001_R1_001;AC_S1_L001_R2_001  AC_S1_L001_001  \\\n",
       "\n",
       "  sample_title library_strategy  \n",
       "0        AC_S1          RNA-seq  "
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    },
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>run_accession</th>\n",
       "      <th>treatment</th>\n",
       "      <th>replicate</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>AC_S1_L001_001</td>\n",
       "      <td>AC</td>\n",
       "      <td>S1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>AH_S2_L001_001</td>\n",
       "      <td>AH</td>\n",
       "      <td>S2</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>BC_S3_L001_001</td>\n",
       "      <td>BC</td>\n",
       "      <td>S3</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "    run_accession treatment replicate\n",
       "0  AC_S1_L001_001        AC        S1\n",
       "1  AH_S2_L001_001        AH        S2\n",
       "2  BC_S3_L001_001        BC        S3"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    }
   ],
   "source": [
    "file_name = 'rna-seq_19-09-23_all'\n",
    "label_list = ['treatment', 'replicate']\n",
    "\n",
    "metadata = pd.read_csv(os.path.join('data', 'metadata', file_name + '.tsv'), sep = '\\t')\n",
    "display(metadata.head(1))\n",
    "metadata = utils.prepare_metadata(id_col = 'run_accession', label_col = 'sample_title', sep = '_', label_list = label_list, df = metadata)\n",
    "display(metadata.head(3))\n",
    "metadata.to_csv(os.path.join('data', 'metadata', file_name + '.csv'), index = False)\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 18,
   "id": "bafed3cd",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>sample_accession</th>\n",
       "      <th>run_accession</th>\n",
       "      <th>fastq_ftp</th>\n",
       "      <th>experiment_accession</th>\n",
       "      <th>study_accession</th>\n",
       "      <th>tax_id</th>\n",
       "      <th>scientific_name</th>\n",
       "      <th>submitted_ftp</th>\n",
       "      <th>sra_ftp</th>\n",
       "      <th>sample_title</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>SAMN04271156</td>\n",
       "      <td>['SRR2919698', 'SRR2919697']</td>\n",
       "      <td>['ftp.sra.ebi.ac.uk/vol1/fastq/SRR291/008/SRR2...</td>\n",
       "      <td>SRX1433535</td>\n",
       "      <td>PRJNA302280</td>\n",
       "      <td>10090</td>\n",
       "      <td>Mus musculus</td>\n",
       "      <td>NaN</td>\n",
       "      <td>ftp.sra.ebi.ac.uk/vol1/srr/SRR291/008/SRR2919698</td>\n",
       "      <td>in vivo, KO 2</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "  sample_accession                 run_accession   \n",
       "0     SAMN04271156  ['SRR2919698', 'SRR2919697']  \\\n",
       "\n",
       "                                           fastq_ftp experiment_accession   \n",
       "0  ['ftp.sra.ebi.ac.uk/vol1/fastq/SRR291/008/SRR2...           SRX1433535  \\\n",
       "\n",
       "  study_accession  tax_id scientific_name  submitted_ftp   \n",
       "0     PRJNA302280   10090    Mus musculus            NaN  \\\n",
       "\n",
       "                                            sra_ftp   sample_title  \n",
       "0  ftp.sra.ebi.ac.uk/vol1/srr/SRR291/008/SRR2919698  in vivo, KO 2  "
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    },
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>sample_accession</th>\n",
       "      <th>cell_type</th>\n",
       "      <th>vector</th>\n",
       "      <th>treatment</th>\n",
       "      <th>replicate</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>30</th>\n",
       "      <td>SAMN04271182</td>\n",
       "      <td>KO</td>\n",
       "      <td>Vehicle</td>\n",
       "      <td>Hypoxia</td>\n",
       "      <td>2</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>11</th>\n",
       "      <td>SAMN04271168</td>\n",
       "      <td>WT</td>\n",
       "      <td>POVPC</td>\n",
       "      <td>Normoxia</td>\n",
       "      <td>3</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>26</th>\n",
       "      <td>SAMN04271177</td>\n",
       "      <td>WT</td>\n",
       "      <td>Vehicle</td>\n",
       "      <td>Hypoxia</td>\n",
       "      <td>3</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "   sample_accession cell_type   vector treatment replicate\n",
       "30     SAMN04271182        KO  Vehicle   Hypoxia         2\n",
       "11     SAMN04271168        WT    POVPC  Normoxia         3\n",
       "26     SAMN04271177        WT  Vehicle   Hypoxia         3"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    }
   ],
   "source": [
    "file_name = 'MM_mouse_heart_hypx'\n",
    "label_list = ['filler1', 'filler2', 'cell_type', 'vector', 'treatment', 'replicate']\n",
    "\n",
    "metadata = pd.read_csv(os.path.join(BASE, 'data', 'metadata', file_name + '.tsv'), sep = '\\t')\n",
    "display(metadata.head(1))\n",
    "metadata = utils.prepare_metadata(id_col = 'sample_accession', label_col = 'sample_title', sep = ' ', label_list = label_list, df = metadata)\n",
    "metadata = metadata.drop(['filler1', 'filler2'], axis = 1)\n",
    "metadata = metadata.dropna()\n",
    "display(metadata.sample(3))\n",
    "metadata.to_csv(os.path.join(BASE, 'data', 'metadata', file_name + '.csv'), index = False)\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 34,
   "id": "13ce09c8",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>sample_accession</th>\n",
       "      <th>run_accession</th>\n",
       "      <th>fastq_ftp</th>\n",
       "      <th>experiment_accession</th>\n",
       "      <th>study_accession</th>\n",
       "      <th>tax_id</th>\n",
       "      <th>scientific_name</th>\n",
       "      <th>submitted_ftp</th>\n",
       "      <th>sra_ftp</th>\n",
       "      <th>sample_title</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>SAMN08285281</td>\n",
       "      <td>['SRR6435406']</td>\n",
       "      <td>['ftp.sra.ebi.ac.uk/vol1/fastq/SRR643/006/SRR6...</td>\n",
       "      <td>SRX3526959</td>\n",
       "      <td>PRJNA428243</td>\n",
       "      <td>9606</td>\n",
       "      <td>Homo sapiens</td>\n",
       "      <td>NaN</td>\n",
       "      <td>ftp.sra.ebi.ac.uk/vol1/srr/SRR643/006/SRR6435406</td>\n",
       "      <td>D737N</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "  sample_accession   run_accession   \n",
       "0     SAMN08285281  ['SRR6435406']  \\\n",
       "\n",
       "                                           fastq_ftp experiment_accession   \n",
       "0  ['ftp.sra.ebi.ac.uk/vol1/fastq/SRR643/006/SRR6...           SRX3526959  \\\n",
       "\n",
       "  study_accession  tax_id scientific_name  submitted_ftp   \n",
       "0     PRJNA428243    9606    Homo sapiens            NaN  \\\n",
       "\n",
       "                                            sra_ftp sample_title  \n",
       "0  ftp.sra.ebi.ac.uk/vol1/srr/SRR643/006/SRR6435406        D737N  "
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    },
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>sample_accession</th>\n",
       "      <th>title</th>\n",
       "      <th>treatment</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>SAMN08285281</td>\n",
       "      <td>D737N</td>\n",
       "      <td>N</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>SAMN08285280</td>\n",
       "      <td>D740H</td>\n",
       "      <td>H</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>SAMN08285279</td>\n",
       "      <td>D740N</td>\n",
       "      <td>N</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "  sample_accession  title treatment\n",
       "0     SAMN08285281  D737N         N\n",
       "1     SAMN08285280  D740H         H\n",
       "2     SAMN08285279  D740N         N"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    }
   ],
   "source": [
    "file_name = 'MM_monkey_heart_hypx'\n",
    "label_list = ['title', 'test', ]\n",
    "\n",
    "metadata = pd.read_csv(os.path.join(BASE, 'data', 'metadata', file_name + '.tsv'), sep = '\\t')\n",
    "display(metadata.head(1))\n",
    "metadata = utils.prepare_metadata(id_col = 'sample_accession', label_col = 'sample_title', sep = ' ', label_list = label_list, df = metadata)\n",
    "metadata['treatment'] = metadata['title'].str.split('').str[-2]\n",
    "metadata = metadata.drop(['test'], axis = 1)\n",
    "display(metadata.head(3))\n",
    "metadata.to_csv(os.path.join(BASE, 'data', 'metadata', file_name + '.csv'), index = False)\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 3,
   "id": "5b8e3d78",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>Run</th>\n",
       "      <th>Assay Type</th>\n",
       "      <th>Individual</th>\n",
       "      <th>stress</th>\n",
       "      <th>Time</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>SRR11149147</td>\n",
       "      <td>ATAC-seq</td>\n",
       "      <td>Tibetan-1</td>\n",
       "      <td>hypoxia</td>\n",
       "      <td>5d</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>SRR11149148</td>\n",
       "      <td>ATAC-seq</td>\n",
       "      <td>Tibetan-2</td>\n",
       "      <td>normoxia</td>\n",
       "      <td>0h</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>SRR11149149</td>\n",
       "      <td>ATAC-seq</td>\n",
       "      <td>Tibetan-2</td>\n",
       "      <td>hypoxia</td>\n",
       "      <td>6h</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>SRR11149151</td>\n",
       "      <td>ATAC-seq</td>\n",
       "      <td>Tibetan-2</td>\n",
       "      <td>hypoxia</td>\n",
       "      <td>3d</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>SRR11149154</td>\n",
       "      <td>ATAC-seq</td>\n",
       "      <td>Tibetan-3</td>\n",
       "      <td>hypoxia</td>\n",
       "      <td>6h</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>5</th>\n",
       "      <td>SRR11149155</td>\n",
       "      <td>ATAC-seq</td>\n",
       "      <td>Tibetan-3</td>\n",
       "      <td>hypoxia</td>\n",
       "      <td>1d</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "           Run Assay Type Individual    stress Time\n",
       "0  SRR11149147   ATAC-seq  Tibetan-1   hypoxia   5d\n",
       "1  SRR11149148   ATAC-seq  Tibetan-2  normoxia   0h\n",
       "2  SRR11149149   ATAC-seq  Tibetan-2   hypoxia   6h\n",
       "3  SRR11149151   ATAC-seq  Tibetan-2   hypoxia   3d\n",
       "4  SRR11149154   ATAC-seq  Tibetan-3   hypoxia   6h\n",
       "5  SRR11149155   ATAC-seq  Tibetan-3   hypoxia   1d"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    }
   ],
   "source": [
    "# Prep metadata for SRA Run Selector metadata files\n",
    "\n",
    "name = 'HUVEC_rna_hypx_2'\n",
    "metadata = pd.read_csv(os.path.join('data', 'metadata', name + '.txt'), sep = ',')\n",
    "metadata = metadata[['Run', 'Assay Type', 'Individual','stress', 'Time']]\n",
    "display(metadata.head(6))\n",
    "\n",
    "metadata.to_csv(os.path.join('data', 'metadata', name + '.csv'), index = False)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 30,
   "id": "750d7c0a",
   "metadata": {},
   "outputs": [],
   "source": [
    "name = 'CM_turtle_hypx'\n",
    "metadata = pd.read_csv(os.path.join(BASE, 'data', 'metadata', name + '.txt'), sep = ',')\n",
    "metadata = metadata[['Run', 'CONDITION', 'Tissue', 'source_name']]\n",
    "metadata.columns = ['run_accession', 'condition', 'tissue', 'time']\n",
    "metadata['treatment'] = metadata['condition'].str.split(' ').str[2]\n",
    "metadata['time'] = metadata['time'].str.split(' ').str[0].str.replace('hs', 'hrs')\n",
    "metadata.to_csv(os.path.join(BASE, 'data', 'metadata', name + '.csv'), index = False)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 31,
   "id": "465c137e",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>organism</th>\n",
       "      <th>run_accession</th>\n",
       "      <th>treatment</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>Danio rerio</td>\n",
       "      <td>SAMN01831289</td>\n",
       "      <td>Normoxia</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>Danio rerio</td>\n",
       "      <td>SAMN01831290</td>\n",
       "      <td>Normoxia</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>Danio rerio</td>\n",
       "      <td>SAMN01831290</td>\n",
       "      <td>Normoxia</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>Danio rerio</td>\n",
       "      <td>SAMN01831291</td>\n",
       "      <td>Normoxia</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>Danio rerio</td>\n",
       "      <td>SAMN01831291</td>\n",
       "      <td>Normoxia</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>5</th>\n",
       "      <td>Danio rerio</td>\n",
       "      <td>SAMN01831292</td>\n",
       "      <td>cold</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "      organism run_accession treatment\n",
       "0  Danio rerio  SAMN01831289  Normoxia\n",
       "1  Danio rerio  SAMN01831290  Normoxia\n",
       "2  Danio rerio  SAMN01831290  Normoxia\n",
       "3  Danio rerio  SAMN01831291  Normoxia\n",
       "4  Danio rerio  SAMN01831291  Normoxia\n",
       "5  Danio rerio  SAMN01831292      cold"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    }
   ],
   "source": [
    "# Prep metadata for SRA Run Selector metadata files\n",
    "\n",
    "name = 'DR_zebrafish_hypx'\n",
    "metadata = pd.read_csv(os.path.join(BASE, 'data', 'metadata', name + '.txt'), sep = ',')\n",
    "metadata = metadata[[ 'Organism',  'BioSample', 'Sample Name']]\n",
    "metadata.columns = [ 'organism',  'run_accession', 'sample_name']\n",
    "# Split sample_name into treatment and replicate based on digit\n",
    "metadata[['treatment', 'replicate']] = metadata['sample_name'].str.split('\\d{1}$', expand = True, regex = True)\n",
    "metadata = metadata.drop(['sample_name', 'replicate'], axis = 1)\n",
    "# Replace ck with Normoxia and hypo with Hypoxia\n",
    "metadata['treatment'] = metadata['treatment'].str.replace('ck', 'Normoxia').str.replace('hypo', 'Hypoxia')\n",
    "display(metadata.head(6))\n",
    "metadata.to_csv(os.path.join(BASE, 'data', 'metadata', name + '.csv'), index = False)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 59,
   "id": "9770bfc1",
   "metadata": {},
   "outputs": [
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "C:\\Users\\noam1\\AppData\\Local\\Temp/ipykernel_4416/3970143689.py:6: FutureWarning: The default value of regex will change from True to False in a future version.\n",
      "  metadata['replicate'] = metadata['replicate'].str.replace('\\D', '')\n",
      "C:\\Users\\noam1\\AppData\\Local\\Temp/ipykernel_4416/3970143689.py:7: FutureWarning: The default value of regex will change from True to False in a future version.\n",
      "  metadata['time'] = metadata['time'].str.replace('\\D', '')\n"
     ]
    },
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>organism</th>\n",
       "      <th>run_accession</th>\n",
       "      <th>replicate</th>\n",
       "      <th>treatment</th>\n",
       "      <th>time</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>Gallus gallus</td>\n",
       "      <td>SAMN10449641</td>\n",
       "      <td>2</td>\n",
       "      <td>Hypoxia</td>\n",
       "      <td>12</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>Gallus gallus</td>\n",
       "      <td>SAMN10449640</td>\n",
       "      <td>1</td>\n",
       "      <td>Hypoxia</td>\n",
       "      <td>12</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>Gallus gallus</td>\n",
       "      <td>SAMN10449637</td>\n",
       "      <td>1</td>\n",
       "      <td>Control</td>\n",
       "      <td>12</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>Gallus gallus</td>\n",
       "      <td>SAMN10449636</td>\n",
       "      <td>3</td>\n",
       "      <td>Hypoxia</td>\n",
       "      <td>8</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>Gallus gallus</td>\n",
       "      <td>SAMN10449639</td>\n",
       "      <td>3</td>\n",
       "      <td>Control</td>\n",
       "      <td>12</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>5</th>\n",
       "      <td>Gallus gallus</td>\n",
       "      <td>SAMN10449638</td>\n",
       "      <td>2</td>\n",
       "      <td>Control</td>\n",
       "      <td>12</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "        organism run_accession replicate treatment time\n",
       "0  Gallus gallus  SAMN10449641         2   Hypoxia   12\n",
       "1  Gallus gallus  SAMN10449640         1   Hypoxia   12\n",
       "2  Gallus gallus  SAMN10449637         1   Control   12\n",
       "3  Gallus gallus  SAMN10449636         3   Hypoxia    8\n",
       "4  Gallus gallus  SAMN10449639         3   Control   12\n",
       "5  Gallus gallus  SAMN10449638         2   Control   12"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    }
   ],
   "source": [
    "name = 'GG_chicken_embryo_hypoxia'\n",
    "metadata = pd.read_csv(os.path.join(BASE, 'data', 'metadata', name + '.txt'), sep = ',')\n",
    "metadata = metadata[[ 'Organism',  'BioSample', 'REPLICATE', 'Treatment', 'Isolate']]\n",
    "metadata.columns = [ 'organism',  'run_accession', 'replicate', 'treatment', 'time']\n",
    "# Remove all non-numeric characters from replicate and time\n",
    "metadata['replicate'] = metadata['replicate'].str.replace('\\D', '')\n",
    "metadata['time'] = metadata['time'].str.replace('\\D', '')\n",
    "\n",
    "display(metadata.head(6))\n",
    "metadata.to_csv(os.path.join(BASE, 'data', 'metadata', name + '.csv'), index = False)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 4,
   "id": "a4b9537e",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>study_accession</th>\n",
       "      <th>sample_accession</th>\n",
       "      <th>experiment_accession</th>\n",
       "      <th>run_accession</th>\n",
       "      <th>tax_id</th>\n",
       "      <th>scientific_name</th>\n",
       "      <th>fastq_ftp</th>\n",
       "      <th>submitted_ftp</th>\n",
       "      <th>sra_ftp</th>\n",
       "      <th>sample_title</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>PRJNA338259</td>\n",
       "      <td>SAMN12766909</td>\n",
       "      <td>SRX6853603</td>\n",
       "      <td>SRR10125020</td>\n",
       "      <td>9606</td>\n",
       "      <td>Homo sapiens</td>\n",
       "      <td>ftp.sra.ebi.ac.uk/vol1/fastq/SRR101/020/SRR101...</td>\n",
       "      <td>NaN</td>\n",
       "      <td>ftp.sra.ebi.ac.uk/vol1/srr/SRR101/020/SRR10125020</td>\n",
       "      <td>SK-MEL-28_RNAseq_Normoxia_r1</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "  study_accession sample_accession experiment_accession run_accession  tax_id  \\\n",
       "0     PRJNA338259     SAMN12766909           SRX6853603   SRR10125020    9606   \n",
       "\n",
       "  scientific_name                                          fastq_ftp  \\\n",
       "0    Homo sapiens  ftp.sra.ebi.ac.uk/vol1/fastq/SRR101/020/SRR101...   \n",
       "\n",
       "   submitted_ftp                                            sra_ftp  \\\n",
       "0            NaN  ftp.sra.ebi.ac.uk/vol1/srr/SRR101/020/SRR10125020   \n",
       "\n",
       "                   sample_title  \n",
       "0  SK-MEL-28_RNAseq_Normoxia_r1  "
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    },
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>sample_accession</th>\n",
       "      <th>cell_type</th>\n",
       "      <th>platform</th>\n",
       "      <th>treatment</th>\n",
       "      <th>replicate</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>SAMN12766909</td>\n",
       "      <td>SK-MEL-28</td>\n",
       "      <td>RNAseq</td>\n",
       "      <td>Normoxia</td>\n",
       "      <td>r1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>SAMN12766909</td>\n",
       "      <td>SK-MEL-28</td>\n",
       "      <td>RNAseq</td>\n",
       "      <td>Normoxia</td>\n",
       "      <td>r1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>SAMN12766908</td>\n",
       "      <td>SK-MEL-28</td>\n",
       "      <td>RNAseq</td>\n",
       "      <td>Hypoxia</td>\n",
       "      <td>r1</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "  sample_accession  cell_type platform treatment replicate\n",
       "0     SAMN12766909  SK-MEL-28   RNAseq  Normoxia        r1\n",
       "1     SAMN12766909  SK-MEL-28   RNAseq  Normoxia        r1\n",
       "2     SAMN12766908  SK-MEL-28   RNAseq   Hypoxia        r1"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    }
   ],
   "source": [
    "\n",
    "metadata = pd.read_csv(os.path.join(PATH, 'data', 'metadata', 'diff_cell_lines_hypx_2.tsv'), sep = '\\t')\n",
    "display(metadata.head(1))\n",
    "metadata = utils.prepare_metadata(id_col = 'sample_accession', label_col = 'sample_title', sep = '_', label_list = ['cell_type', 'platform', 'treatment', 'replicate'], df = metadata)\n",
    "display(metadata.head(3))\n",
    "metadata.to_csv(os.path.join(PATH, 'data', 'metadata', 'diff_cell_lines_hypx_2.csv'), index = False)\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 7,
   "id": "e377b0ff",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>sample_accession</th>\n",
       "      <th>run_accession</th>\n",
       "      <th>fastq_ftp</th>\n",
       "      <th>study_accession</th>\n",
       "      <th>experiment_accession</th>\n",
       "      <th>tax_id</th>\n",
       "      <th>scientific_name</th>\n",
       "      <th>submitted_ftp</th>\n",
       "      <th>sra_ftp</th>\n",
       "      <th>sample_alias</th>\n",
       "      <th>sample_title</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>SAMN12698301</td>\n",
       "      <td>['SRR10071159']</td>\n",
       "      <td>['ftp.sra.ebi.ac.uk/vol1/fastq/SRR100/059/SRR1...</td>\n",
       "      <td>PRJNA564026</td>\n",
       "      <td>SRX6804742</td>\n",
       "      <td>9606</td>\n",
       "      <td>Homo sapiens</td>\n",
       "      <td>NaN</td>\n",
       "      <td>ftp.sra.ebi.ac.uk/vol1/srr/SRR100/059/SRR10071159</td>\n",
       "      <td>GSM4061744</td>\n",
       "      <td>0h rep1</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "  sample_accession    run_accession  \\\n",
       "0     SAMN12698301  ['SRR10071159']   \n",
       "\n",
       "                                           fastq_ftp study_accession  \\\n",
       "0  ['ftp.sra.ebi.ac.uk/vol1/fastq/SRR100/059/SRR1...     PRJNA564026   \n",
       "\n",
       "  experiment_accession  tax_id scientific_name  submitted_ftp  \\\n",
       "0           SRX6804742    9606    Homo sapiens            NaN   \n",
       "\n",
       "                                             sra_ftp sample_alias sample_title  \n",
       "0  ftp.sra.ebi.ac.uk/vol1/srr/SRR100/059/SRR10071159   GSM4061744      0h rep1  "
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    },
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>sample_accession</th>\n",
       "      <th>time</th>\n",
       "      <th>replicate</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>SAMN12698301</td>\n",
       "      <td>0h</td>\n",
       "      <td>rep1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>SAMN12698297</td>\n",
       "      <td>24h</td>\n",
       "      <td>rep1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>SAMN12698296</td>\n",
       "      <td>24h</td>\n",
       "      <td>rep2</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "  sample_accession time replicate\n",
       "0     SAMN12698301   0h      rep1\n",
       "1     SAMN12698297  24h      rep1\n",
       "2     SAMN12698296  24h      rep2"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    }
   ],
   "source": [
    "\n",
    "metadata = pd.read_csv(os.path.join(PATH, 'data', 'metadata', 'HMVEC_hypx.tsv'), sep = '\\t')\n",
    "display(metadata.head(1))\n",
    "metadata = utils.prepare_metadata(id_col = 'sample_accession', label_col = 'sample_title', sep = ' ', label_list = ['time', 'replicate'], df = metadata)\n",
    "\n",
    "display(metadata.head(3))\n",
    "metadata.to_csv(os.path.join(PATH, 'data', 'metadata', 'HMVEC_hypx.csv'), index = False)\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 9,
   "id": "37731517",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>study_accession</th>\n",
       "      <th>sample_accession</th>\n",
       "      <th>experiment_accession</th>\n",
       "      <th>run_accession</th>\n",
       "      <th>tax_id</th>\n",
       "      <th>scientific_name</th>\n",
       "      <th>library_layout</th>\n",
       "      <th>library_source</th>\n",
       "      <th>fastq_ftp</th>\n",
       "      <th>submitted_ftp</th>\n",
       "      <th>sra_ftp</th>\n",
       "      <th>sample_title</th>\n",
       "      <th>bam_ftp</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>PRJNA685592</td>\n",
       "      <td>SAMN17084531</td>\n",
       "      <td>SRX9686685</td>\n",
       "      <td>SRR13255727</td>\n",
       "      <td>9606</td>\n",
       "      <td>Homo sapiens</td>\n",
       "      <td>PAIRED</td>\n",
       "      <td>TRANSCRIPTOMIC</td>\n",
       "      <td>ftp.sra.ebi.ac.uk/vol1/fastq/SRR132/027/SRR132...</td>\n",
       "      <td>NaN</td>\n",
       "      <td>ftp.sra.ebi.ac.uk/vol1/srr/SRR132/027/SRR13255727</td>\n",
       "      <td>SMC 0hr replicate 1</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "  study_accession sample_accession experiment_accession run_accession  tax_id  \\\n",
       "0     PRJNA685592     SAMN17084531           SRX9686685   SRR13255727    9606   \n",
       "\n",
       "  scientific_name library_layout  library_source  \\\n",
       "0    Homo sapiens         PAIRED  TRANSCRIPTOMIC   \n",
       "\n",
       "                                           fastq_ftp  submitted_ftp  \\\n",
       "0  ftp.sra.ebi.ac.uk/vol1/fastq/SRR132/027/SRR132...            NaN   \n",
       "\n",
       "                                             sra_ftp         sample_title  \\\n",
       "0  ftp.sra.ebi.ac.uk/vol1/srr/SRR132/027/SRR13255727  SMC 0hr replicate 1   \n",
       "\n",
       "   bam_ftp  \n",
       "0      NaN  "
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    },
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>sample_accession</th>\n",
       "      <th>cell</th>\n",
       "      <th>time</th>\n",
       "      <th>replicate</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>SAMN17084531</td>\n",
       "      <td>SMC</td>\n",
       "      <td>0hr</td>\n",
       "      <td>replicate1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>SAMN17084530</td>\n",
       "      <td>SMC</td>\n",
       "      <td>0hr</td>\n",
       "      <td>replicate2</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>SAMN17084527</td>\n",
       "      <td>SMC</td>\n",
       "      <td>24hr</td>\n",
       "      <td>replicate2</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "  sample_accession cell  time   replicate\n",
       "0     SAMN17084531  SMC   0hr  replicate1\n",
       "1     SAMN17084530  SMC   0hr  replicate2\n",
       "2     SAMN17084527  SMC  24hr  replicate2"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    }
   ],
   "source": [
    "\n",
    "metadata = pd.read_csv(os.path.join(PATH, 'data', 'metadata', 'SMC_hypx.tsv'), sep = '\\t')\n",
    "display(metadata.head(1))\n",
    "metadata = utils.prepare_metadata(id_col = 'sample_accession', label_col = 'sample_title', sep = ' ', label_list = ['cell', 'time', 'replicate', 'add'], df = metadata)\n",
    "metadata['replicate'] = metadata['replicate'] + metadata['add']\n",
    "metadata = metadata.drop(['add'], axis = 1)\n",
    "display(metadata.head(3))\n",
    "metadata.to_csv(os.path.join(PATH, 'data', 'metadata', 'SMC_hypx.csv'), index = False)\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 16,
   "id": "ff2cba33",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>study_accession</th>\n",
       "      <th>sample_accession</th>\n",
       "      <th>experiment_accession</th>\n",
       "      <th>run_accession</th>\n",
       "      <th>tax_id</th>\n",
       "      <th>scientific_name</th>\n",
       "      <th>fastq_ftp</th>\n",
       "      <th>submitted_ftp</th>\n",
       "      <th>sra_ftp</th>\n",
       "      <th>sample_title</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>PRJNA563926</td>\n",
       "      <td>SAMN13059457</td>\n",
       "      <td>SRX7023508</td>\n",
       "      <td>SRR10312288</td>\n",
       "      <td>9606</td>\n",
       "      <td>Homo sapiens</td>\n",
       "      <td>ftp.sra.ebi.ac.uk/vol1/fastq/SRR103/088/SRR103...</td>\n",
       "      <td>NaN</td>\n",
       "      <td>ftp.sra.ebi.ac.uk/vol1/srr/SRR103/088/SRR10312288</td>\n",
       "      <td>C_CD19_28Z_Batch1</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "  study_accession sample_accession experiment_accession run_accession  tax_id  \\\n",
       "0     PRJNA563926     SAMN13059457           SRX7023508   SRR10312288    9606   \n",
       "\n",
       "  scientific_name                                          fastq_ftp  \\\n",
       "0    Homo sapiens  ftp.sra.ebi.ac.uk/vol1/fastq/SRR103/088/SRR103...   \n",
       "\n",
       "   submitted_ftp                                            sra_ftp  \\\n",
       "0            NaN  ftp.sra.ebi.ac.uk/vol1/srr/SRR103/088/SRR10312288   \n",
       "\n",
       "        sample_title  \n",
       "0  C_CD19_28Z_Batch1  "
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "0      [C, CD19, 28Z, Batch1]\n",
      "1    [JUN, CD19, 28Z, Batch1]\n",
      "Name: sample_title, dtype: object\n"
     ]
    },
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>run_accession</th>\n",
       "      <th>treatment</th>\n",
       "      <th>celltype</th>\n",
       "      <th>marker</th>\n",
       "      <th>replicate</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>SRR10312288</td>\n",
       "      <td>C</td>\n",
       "      <td>CD19</td>\n",
       "      <td>28Z</td>\n",
       "      <td>Batch1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>SRR10312289</td>\n",
       "      <td>JUN</td>\n",
       "      <td>CD19</td>\n",
       "      <td>28Z</td>\n",
       "      <td>Batch1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>SRR10312290</td>\n",
       "      <td>C</td>\n",
       "      <td>HA</td>\n",
       "      <td>28Z</td>\n",
       "      <td>Batch1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>SRR10312291</td>\n",
       "      <td>JUN</td>\n",
       "      <td>HA</td>\n",
       "      <td>28Z</td>\n",
       "      <td>Batch1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>SRR10312292</td>\n",
       "      <td>C</td>\n",
       "      <td>CD19</td>\n",
       "      <td>28Z</td>\n",
       "      <td>Batch2</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "  run_accession treatment celltype marker replicate\n",
       "0   SRR10312288         C     CD19    28Z    Batch1\n",
       "1   SRR10312289       JUN     CD19    28Z    Batch1\n",
       "2   SRR10312290         C       HA    28Z    Batch1\n",
       "3   SRR10312291       JUN       HA    28Z    Batch1\n",
       "4   SRR10312292         C     CD19    28Z    Batch2"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    }
   ],
   "source": [
    "\n",
    "metadata = pd.read_csv(os.path.join(PATH, 'data', 'metadata', 'T_cells_jun_oe.tsv'), sep = '\\t')\n",
    "display(metadata.head(1))\n",
    "metadata = utils.prepare_metadata(id_col = 'run_accession', label_col = 'sample_title', sep = '_', label_list = ['treatment', 'celltype', 'marker', 'replicate'], df = metadata)\n",
    "\n",
    "display(metadata.head(5))\n",
    "metadata.to_csv(os.path.join(PATH, 'data', 'metadata', 'T_cells_jun_oe.csv'), index = False)\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 7,
   "id": "dc06c32a",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>study_accession</th>\n",
       "      <th>sample_accession</th>\n",
       "      <th>secondary_sample_accession</th>\n",
       "      <th>experiment_accession</th>\n",
       "      <th>run_accession</th>\n",
       "      <th>tax_id</th>\n",
       "      <th>scientific_name</th>\n",
       "      <th>library_name</th>\n",
       "      <th>library_layout</th>\n",
       "      <th>library_strategy</th>\n",
       "      <th>library_source</th>\n",
       "      <th>library_selection</th>\n",
       "      <th>experiment_title</th>\n",
       "      <th>study_title</th>\n",
       "      <th>experiment_alias</th>\n",
       "      <th>fastq_ftp</th>\n",
       "      <th>submitted_ftp</th>\n",
       "      <th>sample_alias</th>\n",
       "      <th>sample_title</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>PRJNA494075</td>\n",
       "      <td>SAMN10148127</td>\n",
       "      <td>SRS3860176</td>\n",
       "      <td>SRX4779520</td>\n",
       "      <td>SRR7945355</td>\n",
       "      <td>9606</td>\n",
       "      <td>Homo sapiens</td>\n",
       "      <td>NaN</td>\n",
       "      <td>PAIRED</td>\n",
       "      <td>RNA-Seq</td>\n",
       "      <td>TRANSCRIPTOMIC</td>\n",
       "      <td>cDNA</td>\n",
       "      <td>NextSeq 500 sequencing; GSM3408111: DMSO_1; Ho...</td>\n",
       "      <td>Gene expression analysis in response to hypoxi...</td>\n",
       "      <td>GSM3408111</td>\n",
       "      <td>ftp.sra.ebi.ac.uk/vol1/fastq/SRR794/005/SRR794...</td>\n",
       "      <td>NaN</td>\n",
       "      <td>GSM3408111</td>\n",
       "      <td>DMSO_1</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "  study_accession sample_accession secondary_sample_accession  \\\n",
       "0     PRJNA494075     SAMN10148127                 SRS3860176   \n",
       "\n",
       "  experiment_accession run_accession  tax_id scientific_name  library_name  \\\n",
       "0           SRX4779520    SRR7945355    9606    Homo sapiens           NaN   \n",
       "\n",
       "  library_layout library_strategy  library_source library_selection  \\\n",
       "0         PAIRED          RNA-Seq  TRANSCRIPTOMIC              cDNA   \n",
       "\n",
       "                                    experiment_title  \\\n",
       "0  NextSeq 500 sequencing; GSM3408111: DMSO_1; Ho...   \n",
       "\n",
       "                                         study_title experiment_alias  \\\n",
       "0  Gene expression analysis in response to hypoxi...       GSM3408111   \n",
       "\n",
       "                                           fastq_ftp  submitted_ftp  \\\n",
       "0  ftp.sra.ebi.ac.uk/vol1/fastq/SRR794/005/SRR794...            NaN   \n",
       "\n",
       "  sample_alias sample_title  \n",
       "0   GSM3408111       DMSO_1  "
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    },
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>run_accession</th>\n",
       "      <th>treatment</th>\n",
       "      <th>replicate</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>SRR7945355</td>\n",
       "      <td>DMSO</td>\n",
       "      <td>1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>SRR7945356</td>\n",
       "      <td>Hpx</td>\n",
       "      <td>1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>SRR7945357</td>\n",
       "      <td>VH032</td>\n",
       "      <td>1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>SRR7945358</td>\n",
       "      <td>IOX2</td>\n",
       "      <td>1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>SRR7945359</td>\n",
       "      <td>DMSO</td>\n",
       "      <td>2</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "  run_accession treatment replicate\n",
       "0    SRR7945355      DMSO         1\n",
       "1    SRR7945356       Hpx         1\n",
       "2    SRR7945357     VH032         1\n",
       "3    SRR7945358      IOX2         1\n",
       "4    SRR7945359      DMSO         2"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    }
   ],
   "source": [
    "\n",
    "metadata = pd.read_csv(os.path.join(PATH, 'data', 'metadata', 'HeLa_hypx_rna_2.tsv'), sep = '\\t')\n",
    "display(metadata.head(1))\n",
    "metadata = utils.prepare_metadata(id_col = 'run_accession', label_col = 'sample_title', sep = '_', label_list = ['treatment', 'replicate'], df = metadata)\n",
    "\n",
    "display(metadata.head(5))\n",
    "metadata.to_csv(os.path.join(PATH, 'data', 'metadata', 'HeLa_hypx_rna_2.csv'), index = False)\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 3,
   "id": "0b01ba47",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>study_accession</th>\n",
       "      <th>sample_accession</th>\n",
       "      <th>experiment_accession</th>\n",
       "      <th>run_accession</th>\n",
       "      <th>tax_id</th>\n",
       "      <th>scientific_name</th>\n",
       "      <th>fastq_ftp</th>\n",
       "      <th>submitted_ftp</th>\n",
       "      <th>sra_ftp</th>\n",
       "      <th>sample_title</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>PRJNA606047</td>\n",
       "      <td>SAMN14083844</td>\n",
       "      <td>SRX7708145</td>\n",
       "      <td>SRR11067864</td>\n",
       "      <td>9606</td>\n",
       "      <td>Homo sapiens</td>\n",
       "      <td>ftp.sra.ebi.ac.uk/vol1/fastq/SRR110/064/SRR110...</td>\n",
       "      <td>NaN</td>\n",
       "      <td>ftp.sra.ebi.ac.uk/vol1/srr/SRR110/064/SRR11067864</td>\n",
       "      <td>RKO, normoxia, replicate 1</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "  study_accession sample_accession experiment_accession run_accession  tax_id  \\\n",
       "0     PRJNA606047     SAMN14083844           SRX7708145   SRR11067864    9606   \n",
       "\n",
       "  scientific_name                                          fastq_ftp  \\\n",
       "0    Homo sapiens  ftp.sra.ebi.ac.uk/vol1/fastq/SRR110/064/SRR110...   \n",
       "\n",
       "   submitted_ftp                                            sra_ftp  \\\n",
       "0            NaN  ftp.sra.ebi.ac.uk/vol1/srr/SRR110/064/SRR11067864   \n",
       "\n",
       "                 sample_title  \n",
       "0  RKO, normoxia, replicate 1  "
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    },
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>run_accession</th>\n",
       "      <th>cell_type</th>\n",
       "      <th>treatment</th>\n",
       "      <th>replicate</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>SRR11067864</td>\n",
       "      <td>RKO</td>\n",
       "      <td>normoxia</td>\n",
       "      <td>replicate 1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>SRR11067865</td>\n",
       "      <td>RKO</td>\n",
       "      <td>hypoxia</td>\n",
       "      <td>replicate 1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>SRR11067866</td>\n",
       "      <td>RKO</td>\n",
       "      <td>normoxia</td>\n",
       "      <td>replicate 2</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "  run_accession cell_type  treatment     replicate\n",
       "0   SRR11067864       RKO   normoxia   replicate 1\n",
       "1   SRR11067865       RKO    hypoxia   replicate 1\n",
       "2   SRR11067866       RKO   normoxia   replicate 2"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    }
   ],
   "source": [
    "metadata = pd.read_csv(os.path.join(PATH, 'data', 'metadata', 'diff_cell_lines_hypx_3.tsv'), sep = '\\t')\n",
    "display(metadata.head(1))\n",
    "metadata = utils.prepare_metadata(id_col = 'run_accession', label_col = 'sample_title', sep = ',', label_list = ['cell_type', 'treatment', 'replicate'], df = metadata)\n",
    "display(metadata.head(3))\n",
    "metadata.to_csv(os.path.join(PATH, 'data', 'metadata', 'diff_cell_lines_hypx_3.csv'), index = False)\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 7,
   "id": "e8edde0a",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>study_accession</th>\n",
       "      <th>sample_accession</th>\n",
       "      <th>experiment_accession</th>\n",
       "      <th>run_accession</th>\n",
       "      <th>tax_id</th>\n",
       "      <th>scientific_name</th>\n",
       "      <th>fastq_ftp</th>\n",
       "      <th>submitted_ftp</th>\n",
       "      <th>sra_ftp</th>\n",
       "      <th>sample_title</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>PRJNA773654</td>\n",
       "      <td>SAMN22503470</td>\n",
       "      <td>SRX12734408</td>\n",
       "      <td>SRR16531798</td>\n",
       "      <td>9606</td>\n",
       "      <td>Homo sapiens</td>\n",
       "      <td>ftp.sra.ebi.ac.uk/vol1/fastq/SRR165/098/SRR165...</td>\n",
       "      <td>NaN</td>\n",
       "      <td>ftp.sra.ebi.ac.uk/vol1/srr/SRR165/098/SRR16531798</td>\n",
       "      <td>HeLa_Control_rep1</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "  study_accession sample_accession experiment_accession run_accession  tax_id  \\\n",
       "0     PRJNA773654     SAMN22503470          SRX12734408   SRR16531798    9606   \n",
       "\n",
       "  scientific_name                                          fastq_ftp  \\\n",
       "0    Homo sapiens  ftp.sra.ebi.ac.uk/vol1/fastq/SRR165/098/SRR165...   \n",
       "\n",
       "   submitted_ftp                                            sra_ftp  \\\n",
       "0            NaN  ftp.sra.ebi.ac.uk/vol1/srr/SRR165/098/SRR16531798   \n",
       "\n",
       "        sample_title  \n",
       "0  HeLa_Control_rep1  "
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    },
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>run_accession</th>\n",
       "      <th>cell_type</th>\n",
       "      <th>treatment</th>\n",
       "      <th>replicate</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>5</th>\n",
       "      <td>SRR16531803</td>\n",
       "      <td>HeLa</td>\n",
       "      <td>1h</td>\n",
       "      <td>Hpx</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>SRR16531802</td>\n",
       "      <td>HeLa</td>\n",
       "      <td>1h</td>\n",
       "      <td>Hpx</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>9</th>\n",
       "      <td>SRR16531807</td>\n",
       "      <td>A549</td>\n",
       "      <td>Control</td>\n",
       "      <td>rep1</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "  run_accession cell_type treatment replicate\n",
       "5   SRR16531803      HeLa        1h       Hpx\n",
       "4   SRR16531802      HeLa        1h       Hpx\n",
       "9   SRR16531807      A549   Control      rep1"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    }
   ],
   "source": [
    "metadata = pd.read_csv(os.path.join(PATH, 'data', 'metadata', 'HeLa_hypx_3.tsv'), sep = '\\t')\n",
    "display(metadata.head(1))\n",
    "metadata = utils.prepare_metadata(id_col = 'run_accession', label_col = 'sample_title', sep = '_', label_list = ['cell_type', 'treatment', 'replicate'], df = metadata)\n",
    "display(metadata.sample(3))\n",
    "metadata.to_csv(os.path.join(PATH, 'data', 'metadata', 'HeLa_hypx_3.csv'), index = False)\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 6,
   "id": "e728ead4",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>study_accession</th>\n",
       "      <th>sample_accession</th>\n",
       "      <th>experiment_accession</th>\n",
       "      <th>run_accession</th>\n",
       "      <th>tax_id</th>\n",
       "      <th>scientific_name</th>\n",
       "      <th>fastq_ftp</th>\n",
       "      <th>submitted_ftp</th>\n",
       "      <th>sra_ftp</th>\n",
       "      <th>sample_title</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>PRJNA773425</td>\n",
       "      <td>SAMN22486709</td>\n",
       "      <td>SRX12727825</td>\n",
       "      <td>SRR16525202</td>\n",
       "      <td>9606</td>\n",
       "      <td>Homo sapiens</td>\n",
       "      <td>ftp.sra.ebi.ac.uk/vol1/fastq/SRR165/002/SRR165...</td>\n",
       "      <td>NaN</td>\n",
       "      <td>ftp.sra.ebi.ac.uk/vol1/srr/SRR165/002/SRR16525202</td>\n",
       "      <td>0h_Hpx_Control_rep1</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "  study_accession sample_accession experiment_accession run_accession  tax_id  \\\n",
       "0     PRJNA773425     SAMN22486709          SRX12727825   SRR16525202    9606   \n",
       "\n",
       "  scientific_name                                          fastq_ftp  \\\n",
       "0    Homo sapiens  ftp.sra.ebi.ac.uk/vol1/fastq/SRR165/002/SRR165...   \n",
       "\n",
       "   submitted_ftp                                            sra_ftp  \\\n",
       "0            NaN  ftp.sra.ebi.ac.uk/vol1/srr/SRR165/002/SRR16525202   \n",
       "\n",
       "          sample_title  \n",
       "0  0h_Hpx_Control_rep1  "
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    },
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>run_accession</th>\n",
       "      <th>hypx_time</th>\n",
       "      <th>treatment</th>\n",
       "      <th>title</th>\n",
       "      <th>replicate</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>SRR16525202</td>\n",
       "      <td>0h</td>\n",
       "      <td>Hpx</td>\n",
       "      <td>Control</td>\n",
       "      <td>rep1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>SRR16525203</td>\n",
       "      <td>0h</td>\n",
       "      <td>Hpx</td>\n",
       "      <td>Control</td>\n",
       "      <td>rep2</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>8</th>\n",
       "      <td>SRR16525210</td>\n",
       "      <td>24h</td>\n",
       "      <td>Hpx</td>\n",
       "      <td>Control</td>\n",
       "      <td>rep1</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "  run_accession hypx_time treatment    title replicate\n",
       "0   SRR16525202        0h       Hpx  Control      rep1\n",
       "1   SRR16525203        0h       Hpx  Control      rep2\n",
       "8   SRR16525210       24h       Hpx  Control      rep1"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    }
   ],
   "source": [
    "metadata = pd.read_csv(os.path.join(PATH, 'data', 'metadata', 'HeLa_hypx_atac.tsv'), sep = '\\t')\n",
    "display(metadata.head(1))\n",
    "metadata = utils.prepare_metadata(id_col = 'run_accession', label_col = 'sample_title', sep = '_', label_list = ['hypx_time', 'treatment', 'title', 'replicate'], df = metadata)\n",
    "display(metadata.sample(3))\n",
    "metadata.to_csv(os.path.join(PATH, 'data', 'metadata', 'HeLa_hypx_atac.csv'), index = False)\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 4,
   "id": "a08aa5a7",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>sample_accession</th>\n",
       "      <th>run_accession</th>\n",
       "      <th>fastq_ftp</th>\n",
       "      <th>study_accession</th>\n",
       "      <th>experiment_accession</th>\n",
       "      <th>tax_id</th>\n",
       "      <th>scientific_name</th>\n",
       "      <th>submitted_ftp</th>\n",
       "      <th>sra_ftp</th>\n",
       "      <th>sample_title</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>SAMN12139323</td>\n",
       "      <td>['SRR9602206']</td>\n",
       "      <td>['ftp.sra.ebi.ac.uk/vol1/fastq/SRR960/006/SRR9...</td>\n",
       "      <td>PRJNA551124</td>\n",
       "      <td>SRX6367491</td>\n",
       "      <td>9606</td>\n",
       "      <td>Homo sapiens</td>\n",
       "      <td>NaN</td>\n",
       "      <td>ftp.sra.ebi.ac.uk/vol1/srr/SRR960/006/SRR9602206</td>\n",
       "      <td>Hypoxia_rep1_tech1</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "  sample_accession   run_accession  \\\n",
       "0     SAMN12139323  ['SRR9602206']   \n",
       "\n",
       "                                           fastq_ftp study_accession  \\\n",
       "0  ['ftp.sra.ebi.ac.uk/vol1/fastq/SRR960/006/SRR9...     PRJNA551124   \n",
       "\n",
       "  experiment_accession  tax_id scientific_name  submitted_ftp  \\\n",
       "0           SRX6367491    9606    Homo sapiens            NaN   \n",
       "\n",
       "                                            sra_ftp        sample_title  \n",
       "0  ftp.sra.ebi.ac.uk/vol1/srr/SRR960/006/SRR9602206  Hypoxia_rep1_tech1  "
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    },
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>sample_accession</th>\n",
       "      <th>treatment</th>\n",
       "      <th>replicate</th>\n",
       "      <th>tech_rep</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>8</th>\n",
       "      <td>SAMN12139326</td>\n",
       "      <td>Normoxia</td>\n",
       "      <td>rep2</td>\n",
       "      <td>tech2</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>SAMN12139321</td>\n",
       "      <td>Hypoxia</td>\n",
       "      <td>rep2</td>\n",
       "      <td>tech1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>SAMN12139323</td>\n",
       "      <td>Hypoxia</td>\n",
       "      <td>rep1</td>\n",
       "      <td>tech1</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "  sample_accession treatment replicate tech_rep\n",
       "8     SAMN12139326  Normoxia      rep2    tech2\n",
       "2     SAMN12139321   Hypoxia      rep2    tech1\n",
       "0     SAMN12139323   Hypoxia      rep1    tech1"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    }
   ],
   "source": [
    "file_name = 'MCF7_hypx_atac'\n",
    "metadata = pd.read_csv(os.path.join(PATH, 'data', 'metadata', file_name + '.tsv'), sep = '\\t')\n",
    "display(metadata.head(1))\n",
    "metadata = utils.prepare_metadata(id_col = 'sample_accession', label_col = 'sample_title', sep = '_', label_list = ['treatment', 'replicate', 'tech_rep'], df = metadata)\n",
    "\n",
    "display(metadata.sample(3))\n",
    "metadata.to_csv(os.path.join(PATH, 'data', 'metadata', file_name + '.csv'), index = False)\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 3,
   "id": "b669d765",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>sample_accession</th>\n",
       "      <th>run_accession</th>\n",
       "      <th>fastq_ftp</th>\n",
       "      <th>experiment_accession</th>\n",
       "      <th>study_accession</th>\n",
       "      <th>tax_id</th>\n",
       "      <th>scientific_name</th>\n",
       "      <th>submitted_ftp</th>\n",
       "      <th>sra_ftp</th>\n",
       "      <th>sample_title</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>SAMN07946226</td>\n",
       "      <td>['SRR6233823', 'SRR6233822', 'SRR6233824', 'SR...</td>\n",
       "      <td>['ftp.sra.ebi.ac.uk/vol1/fastq/SRR623/003/SRR6...</td>\n",
       "      <td>SRX3342251</td>\n",
       "      <td>PRJNA416257</td>\n",
       "      <td>9606</td>\n",
       "      <td>Homo sapiens</td>\n",
       "      <td>NaN</td>\n",
       "      <td>ftp.sra.ebi.ac.uk/vol1/srr/SRR623/003/SRR6233823</td>\n",
       "      <td>PC3-HIF1A-ChIP-Seq-siCtrl</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "  sample_accession                                      run_accession  \\\n",
       "0     SAMN07946226  ['SRR6233823', 'SRR6233822', 'SRR6233824', 'SR...   \n",
       "\n",
       "                                           fastq_ftp experiment_accession  \\\n",
       "0  ['ftp.sra.ebi.ac.uk/vol1/fastq/SRR623/003/SRR6...           SRX3342251   \n",
       "\n",
       "  study_accession  tax_id scientific_name  submitted_ftp  \\\n",
       "0     PRJNA416257    9606    Homo sapiens            NaN   \n",
       "\n",
       "                                            sra_ftp               sample_title  \n",
       "0  ftp.sra.ebi.ac.uk/vol1/srr/SRR623/003/SRR6233823  PC3-HIF1A-ChIP-Seq-siCtrl  "
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    },
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>sample_accession</th>\n",
       "      <th>cell_type</th>\n",
       "      <th>expr</th>\n",
       "      <th>manip</th>\n",
       "      <th>treatment</th>\n",
       "      <th>replicate</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>23</th>\n",
       "      <td>SAMN09221637</td>\n",
       "      <td>Flag-OC2</td>\n",
       "      <td>ChIP-Seq</td>\n",
       "      <td>in</td>\n",
       "      <td>PC3</td>\n",
       "      <td>Normoxia</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>17</th>\n",
       "      <td>SAMN09221619</td>\n",
       "      <td>PC3</td>\n",
       "      <td>RNA-Seq</td>\n",
       "      <td>siOC2</td>\n",
       "      <td>Hypoxia</td>\n",
       "      <td>siRNA2</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>26</th>\n",
       "      <td>SAMN09221623</td>\n",
       "      <td>PC3</td>\n",
       "      <td>RNA-Seq</td>\n",
       "      <td>siOC2</td>\n",
       "      <td>Normoxia</td>\n",
       "      <td>siRNA2</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "   sample_accession cell_type      expr  manip treatment replicate\n",
       "23     SAMN09221637  Flag-OC2  ChIP-Seq     in       PC3  Normoxia\n",
       "17     SAMN09221619       PC3   RNA-Seq  siOC2   Hypoxia    siRNA2\n",
       "26     SAMN09221623       PC3   RNA-Seq  siOC2  Normoxia    siRNA2"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    }
   ],
   "source": [
    "file_name = 'LNCaP_PC3_chip_rna_hypx'\n",
    "metadata = pd.read_csv(os.path.join(PATH, 'data', 'metadata', file_name + '.tsv'), sep = '\\t')\n",
    "display(metadata.head(1))\n",
    "metadata = utils.prepare_metadata(id_col = 'sample_accession', label_col = 'sample_title', sep = '_', label_list = ['cell_type', 'expr', 'manip', 'treatment', 'replicate'], df = metadata)\n",
    "display(metadata.sample(3))\n",
    "metadata.to_csv(os.path.join(PATH, 'data', 'metadata', file_name + '.csv'), index = False)\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 31,
   "id": "b4106daf",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>study_accession</th>\n",
       "      <th>sample_accession</th>\n",
       "      <th>experiment_accession</th>\n",
       "      <th>run_accession</th>\n",
       "      <th>tax_id</th>\n",
       "      <th>scientific_name</th>\n",
       "      <th>fastq_ftp</th>\n",
       "      <th>submitted_ftp</th>\n",
       "      <th>sra_ftp</th>\n",
       "      <th>sample_title</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>PRJNA586858</td>\n",
       "      <td>SAMN13169410</td>\n",
       "      <td>SRX7080695</td>\n",
       "      <td>SRR10379822</td>\n",
       "      <td>9606</td>\n",
       "      <td>Homo sapiens</td>\n",
       "      <td>ftp.sra.ebi.ac.uk/vol1/fastq/SRR103/022/SRR103...</td>\n",
       "      <td>NaN</td>\n",
       "      <td>ftp.sra.ebi.ac.uk/vol1/srr/SRR103/022/SRR10379822</td>\n",
       "      <td>1_Normoxia_siControl_A</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "  study_accession sample_accession experiment_accession run_accession  tax_id  \\\n",
       "0     PRJNA586858     SAMN13169410           SRX7080695   SRR10379822    9606   \n",
       "\n",
       "  scientific_name                                          fastq_ftp  \\\n",
       "0    Homo sapiens  ftp.sra.ebi.ac.uk/vol1/fastq/SRR103/022/SRR103...   \n",
       "\n",
       "   submitted_ftp                                            sra_ftp  \\\n",
       "0            NaN  ftp.sra.ebi.ac.uk/vol1/srr/SRR103/022/SRR10379822   \n",
       "\n",
       "             sample_title  \n",
       "0  1_Normoxia_siControl_A  "
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    },
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>run_accession</th>\n",
       "      <th>indiv</th>\n",
       "      <th>treatment</th>\n",
       "      <th>Manipulation</th>\n",
       "      <th>replicate</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>SRR10379826</td>\n",
       "      <td>2</td>\n",
       "      <td>Normoxia</td>\n",
       "      <td>siControl</td>\n",
       "      <td>B</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>11</th>\n",
       "      <td>SRR10379833</td>\n",
       "      <td>4</td>\n",
       "      <td>Normoxia</td>\n",
       "      <td>siARNT</td>\n",
       "      <td>A</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>27</th>\n",
       "      <td>SRR10379849</td>\n",
       "      <td>10</td>\n",
       "      <td>Hypoxia</td>\n",
       "      <td>siARNT</td>\n",
       "      <td>A</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "   run_accession indiv treatment Manipulation replicate\n",
       "4    SRR10379826     2  Normoxia    siControl         B\n",
       "11   SRR10379833     4  Normoxia       siARNT         A\n",
       "27   SRR10379849    10   Hypoxia       siARNT         A"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    }
   ],
   "source": [
    "file_name = 'PC_hypx'\n",
    "metadata = pd.read_csv(os.path.join(PATH, 'data', 'metadata', file_name + '.tsv'), sep = '\\t')\n",
    "display(metadata.head(1))\n",
    "metadata = utils.prepare_metadata(id_col = 'run_accession', label_col = 'sample_title', sep = '_', label_list = ['indiv', 'treatment', 'Manipulation', 'replicate'], df = metadata)\n",
    "display(metadata.sample(3))\n",
    "metadata.to_csv(os.path.join(PATH, 'data', 'metadata', file_name + '.csv'), index = False)\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 9,
   "id": "a2332abc",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>run_accession</th>\n",
       "      <th>sample_accession</th>\n",
       "      <th>experiment_accession</th>\n",
       "      <th>study_accession</th>\n",
       "      <th>tax_id</th>\n",
       "      <th>scientific_name</th>\n",
       "      <th>fastq_ftp</th>\n",
       "      <th>submitted_ftp</th>\n",
       "      <th>sra_ftp</th>\n",
       "      <th>sample_title</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>SRR10014304</td>\n",
       "      <td>SAMN12623142</td>\n",
       "      <td>SRX6752537</td>\n",
       "      <td>PRJNA561635</td>\n",
       "      <td>9606</td>\n",
       "      <td>Homo sapiens</td>\n",
       "      <td>ftp.sra.ebi.ac.uk/vol1/fastq/SRR100/004/SRR100...</td>\n",
       "      <td>NaN</td>\n",
       "      <td>ftp.sra.ebi.ac.uk/vol1/srr/SRR100/004/SRR10014304</td>\n",
       "      <td>HUVEC_1pct_oxygen_7hrs_EPAS1v2_siRNA_rep2</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "  run_accession sample_accession experiment_accession study_accession  tax_id  \\\n",
       "0   SRR10014304     SAMN12623142           SRX6752537     PRJNA561635    9606   \n",
       "\n",
       "  scientific_name                                          fastq_ftp  \\\n",
       "0    Homo sapiens  ftp.sra.ebi.ac.uk/vol1/fastq/SRR100/004/SRR100...   \n",
       "\n",
       "   submitted_ftp                                            sra_ftp  \\\n",
       "0            NaN  ftp.sra.ebi.ac.uk/vol1/srr/SRR100/004/SRR10014304   \n",
       "\n",
       "                                sample_title  \n",
       "0  HUVEC_1pct_oxygen_7hrs_EPAS1v2_siRNA_rep2  "
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    },
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>sample_accession</th>\n",
       "      <th>cell_type</th>\n",
       "      <th>ox_pct</th>\n",
       "      <th>filler</th>\n",
       "      <th>time</th>\n",
       "      <th>replicate1</th>\n",
       "      <th>replicate2</th>\n",
       "      <th>replicate3</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>45</th>\n",
       "      <td>SAMN12623095</td>\n",
       "      <td>HUVEC</td>\n",
       "      <td>0pct</td>\n",
       "      <td>oxygen</td>\n",
       "      <td>12hrs</td>\n",
       "      <td>rep1</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>35</th>\n",
       "      <td>SAMN12623103</td>\n",
       "      <td>HUVEC</td>\n",
       "      <td>1pct</td>\n",
       "      <td>oxygen</td>\n",
       "      <td>1hrs</td>\n",
       "      <td>rep1</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>50</th>\n",
       "      <td>SAMN12623160</td>\n",
       "      <td>HUVEC</td>\n",
       "      <td>21pct</td>\n",
       "      <td>oxygen</td>\n",
       "      <td>7hrs</td>\n",
       "      <td>HIF1Av2</td>\n",
       "      <td>siRNA</td>\n",
       "      <td>rep2</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "   sample_accession cell_type ox_pct  filler   time replicate1 replicate2  \\\n",
       "45     SAMN12623095     HUVEC   0pct  oxygen  12hrs       rep1        NaN   \n",
       "35     SAMN12623103     HUVEC   1pct  oxygen   1hrs       rep1        NaN   \n",
       "50     SAMN12623160     HUVEC  21pct  oxygen   7hrs    HIF1Av2      siRNA   \n",
       "\n",
       "   replicate3  \n",
       "45        NaN  \n",
       "35        NaN  \n",
       "50       rep2  "
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    }
   ],
   "source": [
    "file_name = 'HUVEC_hypx_rna'\n",
    "metadata = pd.read_csv(os.path.join(PATH, 'data', 'metadata', file_name + '.tsv'), sep = '\\t')\n",
    "display(metadata.head(1))\n",
    "metadata = utils.prepare_metadata(id_col = 'sample_accession', label_col = 'sample_title', sep = '_', label_list = ['cell_type', 'ox_pct', 'filler','time','replicate1', 'replicate2', 'replicate3'], df = metadata)\n",
    "display(metadata.sample(3))\n",
    "metadata.to_csv(os.path.join(PATH, 'data', 'metadata', file_name + '.csv'), index = False)\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 4,
   "id": "a7ca7ec6",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>scientific_name</th>\n",
       "      <th>fastq_ftp</th>\n",
       "      <th>run_accession</th>\n",
       "      <th>sample_title</th>\n",
       "      <th>library_strategy</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>Homo_sapiens</td>\n",
       "      <td>EV1_R1;EV1_R2</td>\n",
       "      <td>JunD_EV1</td>\n",
       "      <td>JunD_EV1</td>\n",
       "      <td>RNA-seq</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "  scientific_name      fastq_ftp run_accession sample_title library_strategy\n",
       "0    Homo_sapiens  EV1_R1;EV1_R2      JunD_EV1     JunD_EV1          RNA-seq"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    },
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>run_accession</th>\n",
       "      <th>cell_type</th>\n",
       "      <th>treatment</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>JunD_MNLS2</td>\n",
       "      <td>JunD</td>\n",
       "      <td>MNLS2</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>JunD_EV1</td>\n",
       "      <td>JunD</td>\n",
       "      <td>EV1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>JunD_MLS1</td>\n",
       "      <td>JunD</td>\n",
       "      <td>MLS1</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "  run_accession cell_type treatment\n",
       "3    JunD_MNLS2      JunD     MNLS2\n",
       "0      JunD_EV1      JunD       EV1\n",
       "1     JunD_MLS1      JunD      MLS1"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    }
   ],
   "source": [
    "file_name = 'jund_oe'\n",
    "metadata = pd.read_csv(os.path.join(PATH, 'data', 'metadata', file_name + '.tsv'), sep = '\\t')\n",
    "display(metadata.head(1))\n",
    "metadata = utils.prepare_metadata(id_col = 'run_accession', label_col = 'sample_title', sep = '_', label_list = ['cell_type', 'treatment'], df = metadata)\n",
    "display(metadata.sample(3))\n",
    "metadata.to_csv(os.path.join(PATH, 'data', 'metadata', file_name + '.csv'), index = False)\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 33,
   "id": "95a680da",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>sample_accession</th>\n",
       "      <th>run_accession</th>\n",
       "      <th>fastq_ftp</th>\n",
       "      <th>study_accession</th>\n",
       "      <th>experiment_accession</th>\n",
       "      <th>tax_id</th>\n",
       "      <th>scientific_name</th>\n",
       "      <th>submitted_ftp</th>\n",
       "      <th>sra_ftp</th>\n",
       "      <th>sample_title</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>SAMN10181169</td>\n",
       "      <td>['SRR7968970', 'SRR7968971']</td>\n",
       "      <td>['ftp.sra.ebi.ac.uk/vol1/fastq/SRR796/000/SRR7...</td>\n",
       "      <td>PRJNA494832</td>\n",
       "      <td>SRX4802365</td>\n",
       "      <td>9606</td>\n",
       "      <td>Homo sapiens</td>\n",
       "      <td>NaN</td>\n",
       "      <td>ftp.sra.ebi.ac.uk/vol1/srr/SRR796/000/SRR7968970</td>\n",
       "      <td>HKC_8hrs_Normoxia_21_rep1</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "  sample_accession                 run_accession  \\\n",
       "0     SAMN10181169  ['SRR7968970', 'SRR7968971']   \n",
       "\n",
       "                                           fastq_ftp study_accession  \\\n",
       "0  ['ftp.sra.ebi.ac.uk/vol1/fastq/SRR796/000/SRR7...     PRJNA494832   \n",
       "\n",
       "  experiment_accession  tax_id scientific_name  submitted_ftp  \\\n",
       "0           SRX4802365    9606    Homo sapiens            NaN   \n",
       "\n",
       "                                            sra_ftp               sample_title  \n",
       "0  ftp.sra.ebi.ac.uk/vol1/srr/SRR796/000/SRR7968970  HKC_8hrs_Normoxia_21_rep1  "
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    },
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>sample_accession</th>\n",
       "      <th>cell_type</th>\n",
       "      <th>time</th>\n",
       "      <th>treatment</th>\n",
       "      <th>ox_conc</th>\n",
       "      <th>replicate</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>SAMN10181211</td>\n",
       "      <td>HKC</td>\n",
       "      <td>8hrs</td>\n",
       "      <td>Hypoxia</td>\n",
       "      <td>0.1</td>\n",
       "      <td>rep2</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>13</th>\n",
       "      <td>SAMN10181190</td>\n",
       "      <td>HepG2</td>\n",
       "      <td>0hrs</td>\n",
       "      <td>Normoxia</td>\n",
       "      <td>21</td>\n",
       "      <td>rep2</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>8</th>\n",
       "      <td>SAMN10181207</td>\n",
       "      <td>RCC4</td>\n",
       "      <td>VHL</td>\n",
       "      <td>Nx</td>\n",
       "      <td>1</td>\n",
       "      <td>Replacement</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "   sample_accession cell_type  time treatment ox_conc    replicate\n",
       "4      SAMN10181211       HKC  8hrs   Hypoxia     0.1         rep2\n",
       "13     SAMN10181190     HepG2  0hrs  Normoxia      21         rep2\n",
       "8      SAMN10181207      RCC4   VHL        Nx       1  Replacement"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    }
   ],
   "source": [
    "file_name = 'diff_cell_lines_hypx_4'\n",
    "metadata = pd.read_csv(os.path.join(PATH, 'data', 'metadata', file_name + '.tsv'), sep = '\\t')\n",
    "display(metadata.head(1))\n",
    "metadata = utils.prepare_metadata(id_col = 'sample_accession', label_col = 'sample_title', sep = '_', label_list = ['cell_type', 'time', 'treatment', 'ox_conc', 'replicate'], df = metadata)\n",
    "display(metadata.sample(3))\n",
    "metadata[metadata['cell_type'] == 'HKC'].to_csv(os.path.join(PATH, 'data', 'metadata', 'HKC_hypx.csv'), index = False)\n",
    "metadata[metadata['cell_type'] == 'HepG2'].to_csv(os.path.join(PATH, 'data', 'metadata', 'HepG2_hypx.csv'), index = False)\n",
    "metadata.to_csv(os.path.join(PATH, 'data', 'metadata', file_name + '.csv'), index = False)\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 7,
   "id": "ff5b9030",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>study_accession</th>\n",
       "      <th>sample_accession</th>\n",
       "      <th>experiment_accession</th>\n",
       "      <th>run_accession</th>\n",
       "      <th>tax_id</th>\n",
       "      <th>scientific_name</th>\n",
       "      <th>fastq_ftp</th>\n",
       "      <th>submitted_ftp</th>\n",
       "      <th>sra_ftp</th>\n",
       "      <th>sample_title</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>PRJNA608195</td>\n",
       "      <td>SAMN14167476</td>\n",
       "      <td>SRX7785223</td>\n",
       "      <td>SRR11149143</td>\n",
       "      <td>9606</td>\n",
       "      <td>Homo sapiens</td>\n",
       "      <td>ftp.sra.ebi.ac.uk/vol1/fastq/SRR111/043/SRR111...</td>\n",
       "      <td>NaN</td>\n",
       "      <td>ftp.sra.ebi.ac.uk/vol1/srr/SRR111/043/SRR11149143</td>\n",
       "      <td>HUVECs, A-1-1_ATAC-seq</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "  study_accession sample_accession experiment_accession run_accession  tax_id  \\\n",
       "0     PRJNA608195     SAMN14167476           SRX7785223   SRR11149143    9606   \n",
       "\n",
       "  scientific_name                                          fastq_ftp  \\\n",
       "0    Homo sapiens  ftp.sra.ebi.ac.uk/vol1/fastq/SRR111/043/SRR111...   \n",
       "\n",
       "   submitted_ftp                                            sra_ftp  \\\n",
       "0            NaN  ftp.sra.ebi.ac.uk/vol1/srr/SRR111/043/SRR11149143   \n",
       "\n",
       "             sample_title  \n",
       "0  HUVECs, A-1-1_ATAC-seq  "
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    },
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>run_accession</th>\n",
       "      <th>cell_type</th>\n",
       "      <th>method</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>68</th>\n",
       "      <td>SRR11149211</td>\n",
       "      <td>HUVECs, A-4-4</td>\n",
       "      <td>RNA-seq</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>103</th>\n",
       "      <td>SRR11149246</td>\n",
       "      <td>HUVECs, A-3-4</td>\n",
       "      <td>Hi-C</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>78</th>\n",
       "      <td>SRR11149221</td>\n",
       "      <td>HUVECs, W-1-4</td>\n",
       "      <td>RNA-seq</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "    run_accession      cell_type   method\n",
       "68    SRR11149211  HUVECs, A-4-4  RNA-seq\n",
       "103   SRR11149246  HUVECs, A-3-4     Hi-C\n",
       "78    SRR11149221  HUVECs, W-1-4  RNA-seq"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    }
   ],
   "source": [
    "metadata = pd.read_csv(os.path.join(PATH, 'data', 'metadata', 'HUVEC_hypx.tsv'), sep = '\\t')\n",
    "display(metadata.head(1))\n",
    "metadata = utils.prepare_metadata(id_col = 'run_accession', label_col = 'sample_title', sep = '_', label_list = ['cell_type', 'method'], df = metadata)\n",
    "display(metadata.sample(3))\n",
    "metadata.to_csv(os.path.join(PATH, 'data', 'metadata', 'HUVEC_hypx.csv'), index = False)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 3,
   "id": "72246fbc",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>study_accession</th>\n",
       "      <th>sample_accession</th>\n",
       "      <th>experiment_accession</th>\n",
       "      <th>run_accession</th>\n",
       "      <th>tax_id</th>\n",
       "      <th>scientific_name</th>\n",
       "      <th>fastq_ftp</th>\n",
       "      <th>submitted_ftp</th>\n",
       "      <th>sra_ftp</th>\n",
       "      <th>sample_title</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>PRJNA667729</td>\n",
       "      <td>SAMN16380772</td>\n",
       "      <td>SRX9251173</td>\n",
       "      <td>SRR12781720</td>\n",
       "      <td>9606</td>\n",
       "      <td>Homo sapiens</td>\n",
       "      <td>ftp.sra.ebi.ac.uk/vol1/fastq/SRR127/020/SRR127...</td>\n",
       "      <td>NaN</td>\n",
       "      <td>ftp.sra.ebi.ac.uk/vol1/srr/SRR127/020/SRR12781720</td>\n",
       "      <td>S01_ControlNormoxiaChIP</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "  study_accession sample_accession experiment_accession run_accession  tax_id  \\\n",
       "0     PRJNA667729     SAMN16380772           SRX9251173   SRR12781720    9606   \n",
       "\n",
       "  scientific_name                                          fastq_ftp  \\\n",
       "0    Homo sapiens  ftp.sra.ebi.ac.uk/vol1/fastq/SRR127/020/SRR127...   \n",
       "\n",
       "   submitted_ftp                                            sra_ftp  \\\n",
       "0            NaN  ftp.sra.ebi.ac.uk/vol1/srr/SRR127/020/SRR12781720   \n",
       "\n",
       "              sample_title  \n",
       "0  S01_ControlNormoxiaChIP  "
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    },
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>run_accession</th>\n",
       "      <th>replicate</th>\n",
       "      <th>target</th>\n",
       "      <th>treatment</th>\n",
       "      <th>experiment</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>30</th>\n",
       "      <td>SRR12781750</td>\n",
       "      <td>S16</td>\n",
       "      <td>HIF1B</td>\n",
       "      <td>Hypoxia</td>\n",
       "      <td>Input</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>21</th>\n",
       "      <td>SRR12781741</td>\n",
       "      <td>S11</td>\n",
       "      <td>SET1B</td>\n",
       "      <td>Normoxia</td>\n",
       "      <td>ChIP2</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>34</th>\n",
       "      <td>SRR12781754</td>\n",
       "      <td>S18</td>\n",
       "      <td>SET1B</td>\n",
       "      <td>Hypoxia</td>\n",
       "      <td>Input</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "   run_accession replicate target treatment experiment\n",
       "30   SRR12781750       S16  HIF1B   Hypoxia      Input\n",
       "21   SRR12781741       S11  SET1B  Normoxia      ChIP2\n",
       "34   SRR12781754       S18  SET1B   Hypoxia      Input"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    }
   ],
   "source": [
    "import re\n",
    "from importlib import reload\n",
    "reload(utils)\n",
    "metadata = pd.read_csv(os.path.join(PATH, 'data', 'metadata', 'HeLa_chip.tsv'), sep = '\\t')\n",
    "display(metadata.head(1))\n",
    "metadata = utils.prepare_metadata(id_col = 'run_accession', label_col = 'sample_title', sep = '_|(Normoxia|Hypoxia)', label_list = ['replicate', 'x', 'target','treatment','experiment'], df = metadata)\n",
    "metadata = metadata.drop(columns = ['x'])\n",
    "display(metadata.sample(3))\n",
    "metadata.to_csv(os.path.join(PATH, 'data', 'metadata', 'HeLa_chip.csv'), index = False)\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 3,
   "id": "b0626e87",
   "metadata": {},
   "outputs": [
    {
     "ename": "FileNotFoundError",
     "evalue": "[Errno 2] No such file or directory: 'c:\\\\Users\\\\noam1\\\\Desktop\\\\Hypoxia\\\\data\\\\metadata\\\\rotem_atac_seq.xls'",
     "output_type": "error",
     "traceback": [
      "\u001b[1;31m---------------------------------------------------------------------------\u001b[0m",
      "\u001b[1;31mFileNotFoundError\u001b[0m                         Traceback (most recent call last)",
      "\u001b[1;32mc:\\Users\\noam1\\Desktop\\Hypoxia\\prep_metadata_and_counts.ipynb Cell 33\u001b[0m line \u001b[0;36m4\n\u001b[0;32m      <a href='vscode-notebook-cell:/c%3A/Users/noam1/Desktop/Hypoxia/prep_metadata_and_counts.ipynb#X44sZmlsZQ%3D%3D?line=1'>2</a>\u001b[0m \u001b[39mfrom\u001b[39;00m \u001b[39mimportlib\u001b[39;00m \u001b[39mimport\u001b[39;00m reload\n\u001b[0;32m      <a href='vscode-notebook-cell:/c%3A/Users/noam1/Desktop/Hypoxia/prep_metadata_and_counts.ipynb#X44sZmlsZQ%3D%3D?line=2'>3</a>\u001b[0m reload(utils)\n\u001b[1;32m----> <a href='vscode-notebook-cell:/c%3A/Users/noam1/Desktop/Hypoxia/prep_metadata_and_counts.ipynb#X44sZmlsZQ%3D%3D?line=3'>4</a>\u001b[0m metadata \u001b[39m=\u001b[39m pd\u001b[39m.\u001b[39;49mread_excel(os\u001b[39m.\u001b[39;49mpath\u001b[39m.\u001b[39;49mjoin(PATH, \u001b[39m'\u001b[39;49m\u001b[39mdata\u001b[39;49m\u001b[39m'\u001b[39;49m, \u001b[39m'\u001b[39;49m\u001b[39mmetadata\u001b[39;49m\u001b[39m'\u001b[39;49m, \u001b[39m'\u001b[39;49m\u001b[39mrotem_atac_seq.xls\u001b[39;49m\u001b[39m'\u001b[39;49m), sheet_name\u001b[39m=\u001b[39;49m \u001b[39m'\u001b[39;49m\u001b[39mSheet1\u001b[39;49m\u001b[39m'\u001b[39;49m)\n",
      "File \u001b[1;32m~\\AppData\\Roaming\\Python\\Python311\\site-packages\\pandas\\io\\excel\\_base.py:478\u001b[0m, in \u001b[0;36mread_excel\u001b[1;34m(io, sheet_name, header, names, index_col, usecols, dtype, engine, converters, true_values, false_values, skiprows, nrows, na_values, keep_default_na, na_filter, verbose, parse_dates, date_parser, date_format, thousands, decimal, comment, skipfooter, storage_options, dtype_backend)\u001b[0m\n\u001b[0;32m    476\u001b[0m \u001b[39mif\u001b[39;00m \u001b[39mnot\u001b[39;00m \u001b[39misinstance\u001b[39m(io, ExcelFile):\n\u001b[0;32m    477\u001b[0m     should_close \u001b[39m=\u001b[39m \u001b[39mTrue\u001b[39;00m\n\u001b[1;32m--> 478\u001b[0m     io \u001b[39m=\u001b[39m ExcelFile(io, storage_options\u001b[39m=\u001b[39;49mstorage_options, engine\u001b[39m=\u001b[39;49mengine)\n\u001b[0;32m    479\u001b[0m \u001b[39melif\u001b[39;00m engine \u001b[39mand\u001b[39;00m engine \u001b[39m!=\u001b[39m io\u001b[39m.\u001b[39mengine:\n\u001b[0;32m    480\u001b[0m     \u001b[39mraise\u001b[39;00m \u001b[39mValueError\u001b[39;00m(\n\u001b[0;32m    481\u001b[0m         \u001b[39m\"\u001b[39m\u001b[39mEngine should not be specified when passing \u001b[39m\u001b[39m\"\u001b[39m\n\u001b[0;32m    482\u001b[0m         \u001b[39m\"\u001b[39m\u001b[39man ExcelFile - ExcelFile already has the engine set\u001b[39m\u001b[39m\"\u001b[39m\n\u001b[0;32m    483\u001b[0m     )\n",
      "File \u001b[1;32m~\\AppData\\Roaming\\Python\\Python311\\site-packages\\pandas\\io\\excel\\_base.py:1496\u001b[0m, in \u001b[0;36mExcelFile.__init__\u001b[1;34m(self, path_or_buffer, engine, storage_options)\u001b[0m\n\u001b[0;32m   1494\u001b[0m     ext \u001b[39m=\u001b[39m \u001b[39m\"\u001b[39m\u001b[39mxls\u001b[39m\u001b[39m\"\u001b[39m\n\u001b[0;32m   1495\u001b[0m \u001b[39melse\u001b[39;00m:\n\u001b[1;32m-> 1496\u001b[0m     ext \u001b[39m=\u001b[39m inspect_excel_format(\n\u001b[0;32m   1497\u001b[0m         content_or_path\u001b[39m=\u001b[39;49mpath_or_buffer, storage_options\u001b[39m=\u001b[39;49mstorage_options\n\u001b[0;32m   1498\u001b[0m     )\n\u001b[0;32m   1499\u001b[0m     \u001b[39mif\u001b[39;00m ext \u001b[39mis\u001b[39;00m \u001b[39mNone\u001b[39;00m:\n\u001b[0;32m   1500\u001b[0m         \u001b[39mraise\u001b[39;00m \u001b[39mValueError\u001b[39;00m(\n\u001b[0;32m   1501\u001b[0m             \u001b[39m\"\u001b[39m\u001b[39mExcel file format cannot be determined, you must specify \u001b[39m\u001b[39m\"\u001b[39m\n\u001b[0;32m   1502\u001b[0m             \u001b[39m\"\u001b[39m\u001b[39man engine manually.\u001b[39m\u001b[39m\"\u001b[39m\n\u001b[0;32m   1503\u001b[0m         )\n",
      "File \u001b[1;32m~\\AppData\\Roaming\\Python\\Python311\\site-packages\\pandas\\io\\excel\\_base.py:1371\u001b[0m, in \u001b[0;36minspect_excel_format\u001b[1;34m(content_or_path, storage_options)\u001b[0m\n\u001b[0;32m   1368\u001b[0m \u001b[39mif\u001b[39;00m \u001b[39misinstance\u001b[39m(content_or_path, \u001b[39mbytes\u001b[39m):\n\u001b[0;32m   1369\u001b[0m     content_or_path \u001b[39m=\u001b[39m BytesIO(content_or_path)\n\u001b[1;32m-> 1371\u001b[0m \u001b[39mwith\u001b[39;00m get_handle(\n\u001b[0;32m   1372\u001b[0m     content_or_path, \u001b[39m\"\u001b[39;49m\u001b[39mrb\u001b[39;49m\u001b[39m\"\u001b[39;49m, storage_options\u001b[39m=\u001b[39;49mstorage_options, is_text\u001b[39m=\u001b[39;49m\u001b[39mFalse\u001b[39;49;00m\n\u001b[0;32m   1373\u001b[0m ) \u001b[39mas\u001b[39;00m handle:\n\u001b[0;32m   1374\u001b[0m     stream \u001b[39m=\u001b[39m handle\u001b[39m.\u001b[39mhandle\n\u001b[0;32m   1375\u001b[0m     stream\u001b[39m.\u001b[39mseek(\u001b[39m0\u001b[39m)\n",
      "File \u001b[1;32m~\\AppData\\Roaming\\Python\\Python311\\site-packages\\pandas\\io\\common.py:868\u001b[0m, in \u001b[0;36mget_handle\u001b[1;34m(path_or_buf, mode, encoding, compression, memory_map, is_text, errors, storage_options)\u001b[0m\n\u001b[0;32m    859\u001b[0m         handle \u001b[39m=\u001b[39m \u001b[39mopen\u001b[39m(\n\u001b[0;32m    860\u001b[0m             handle,\n\u001b[0;32m    861\u001b[0m             ioargs\u001b[39m.\u001b[39mmode,\n\u001b[1;32m   (...)\u001b[0m\n\u001b[0;32m    864\u001b[0m             newline\u001b[39m=\u001b[39m\u001b[39m\"\u001b[39m\u001b[39m\"\u001b[39m,\n\u001b[0;32m    865\u001b[0m         )\n\u001b[0;32m    866\u001b[0m     \u001b[39melse\u001b[39;00m:\n\u001b[0;32m    867\u001b[0m         \u001b[39m# Binary mode\u001b[39;00m\n\u001b[1;32m--> 868\u001b[0m         handle \u001b[39m=\u001b[39m \u001b[39mopen\u001b[39m(handle, ioargs\u001b[39m.\u001b[39mmode)\n\u001b[0;32m    869\u001b[0m     handles\u001b[39m.\u001b[39mappend(handle)\n\u001b[0;32m    871\u001b[0m \u001b[39m# Convert BytesIO or file objects passed with an encoding\u001b[39;00m\n",
      "\u001b[1;31mFileNotFoundError\u001b[0m: [Errno 2] No such file or directory: 'c:\\\\Users\\\\noam1\\\\Desktop\\\\Hypoxia\\\\data\\\\metadata\\\\rotem_atac_seq.xls'"
     ]
    }
   ],
   "source": [
    "import re\n",
    "from importlib import reload\n",
    "reload(utils)\n",
    "metadata = pd.read_excel(os.path.join(PATH, 'data', 'metadata', 'rotem_atac_seq.xlsx'), sheet_name= 'Sheet1')\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 27,
   "id": "040e0cd7",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>run_accession</th>\n",
       "      <th>cell_type</th>\n",
       "      <th>type</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>SRR11072716</td>\n",
       "      <td>HCT116</td>\n",
       "      <td>input</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>SRR11072717</td>\n",
       "      <td>RKO</td>\n",
       "      <td>input</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>SRR11072718</td>\n",
       "      <td>A549</td>\n",
       "      <td>input</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "  run_accession cell_type    type\n",
       "0   SRR11072716    HCT116   input\n",
       "1   SRR11072717       RKO   input\n",
       "2   SRR11072718      A549   input"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    },
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>run_accession</th>\n",
       "      <th>cell_type</th>\n",
       "      <th>treatment</th>\n",
       "      <th>replicate</th>\n",
       "      <th>type</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>SRR11072720</td>\n",
       "      <td>HCT116</td>\n",
       "      <td>normoxia</td>\n",
       "      <td>replicate 1</td>\n",
       "      <td>HIF1A ChIP</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>SRR11072721</td>\n",
       "      <td>HCT116</td>\n",
       "      <td>normoxia</td>\n",
       "      <td>replicate 2</td>\n",
       "      <td>HIF1A ChIP</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>SRR11072722</td>\n",
       "      <td>HCT116</td>\n",
       "      <td>hypoxia</td>\n",
       "      <td>replicate 1</td>\n",
       "      <td>HIF1A ChIP</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "  run_accession cell_type  treatment     replicate         type\n",
       "0   SRR11072720    HCT116   normoxia   replicate 1   HIF1A ChIP\n",
       "1   SRR11072721    HCT116   normoxia   replicate 2   HIF1A ChIP\n",
       "2   SRR11072722    HCT116    hypoxia   replicate 1   HIF1A ChIP"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    }
   ],
   "source": [
    "from importlib import reload\n",
    "reload(utils)\n",
    "metadata = pd.read_csv(os.path.join(PATH, 'data', 'metadata', 'chip_seq_diff_cell_lines_hypx_3.tsv'), sep = '\\t')\n",
    "metadata_1 = metadata.head(4)\n",
    "metadata_2 = metadata.iloc[4:, :].reset_index(drop = True)\n",
    "\n",
    "metadata_1 = utils.prepare_metadata(id_col = 'run_accession', label_col = 'sample_title', sep = ',', label_list = ['cell_type', 'type'], df = metadata_1)\n",
    "metadata_2 = utils.prepare_metadata(id_col = 'run_accession', label_col = 'sample_title', sep = ',', label_list = ['cell_type', 'treatment', 'replicate','type'], df = metadata_2)\n",
    "display(metadata_1.head(3))\n",
    "display(metadata_2.head(3))\n",
    "metadata = pd.concat([metadata_1, metadata_2], axis = 0)\n",
    "metadata.to_csv(os.path.join(PATH, 'data', 'metadata', 'chip_seq_diff_cell_lines_hypx_3.csv'), index = False)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 8,
   "id": "a4443a09",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>study_accession</th>\n",
       "      <th>sample_accession</th>\n",
       "      <th>experiment_accession</th>\n",
       "      <th>run_accession</th>\n",
       "      <th>tax_id</th>\n",
       "      <th>scientific_name</th>\n",
       "      <th>fastq_ftp</th>\n",
       "      <th>submitted_ftp</th>\n",
       "      <th>sra_ftp</th>\n",
       "      <th>sample_title</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>PRJNA822831</td>\n",
       "      <td>SAMN27279372</td>\n",
       "      <td>SRX14719198</td>\n",
       "      <td>SRR18590497</td>\n",
       "      <td>10090</td>\n",
       "      <td>Mus musculus</td>\n",
       "      <td>ftp.sra.ebi.ac.uk/vol1/fastq/SRR185/097/SRR185...</td>\n",
       "      <td>NaN</td>\n",
       "      <td>ftp.sra.ebi.ac.uk/vol1/srr/SRR185/097/SRR18590497</td>\n",
       "      <td>Normoxia sample 4 [N4_RNA-seq]</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "  study_accession sample_accession experiment_accession run_accession  tax_id  \\\n",
       "0     PRJNA822831     SAMN27279372          SRX14719198   SRR18590497   10090   \n",
       "\n",
       "  scientific_name                                          fastq_ftp  \\\n",
       "0    Mus musculus  ftp.sra.ebi.ac.uk/vol1/fastq/SRR185/097/SRR185...   \n",
       "\n",
       "   submitted_ftp                                            sra_ftp  \\\n",
       "0            NaN  ftp.sra.ebi.ac.uk/vol1/srr/SRR185/097/SRR18590497   \n",
       "\n",
       "                     sample_title  \n",
       "0  Normoxia sample 4 [N4_RNA-seq]  "
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    },
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>run_accession</th>\n",
       "      <th>treatment</th>\n",
       "      <th>replicate</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>SRR18590499</td>\n",
       "      <td>Normoxia</td>\n",
       "      <td>2</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>SRR18590497</td>\n",
       "      <td>Normoxia</td>\n",
       "      <td>4</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>7</th>\n",
       "      <td>SRR18590504</td>\n",
       "      <td>Hypoxia</td>\n",
       "      <td>1</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "  run_accession treatment replicate\n",
       "2   SRR18590499  Normoxia         2\n",
       "0   SRR18590497  Normoxia         4\n",
       "7   SRR18590504   Hypoxia         1"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    }
   ],
   "source": [
    "# Mice hypoxia\n",
    "metadata = pd.read_csv(os.path.join(PATH, 'data', 'metadata', 'MM_hypx.tsv'), sep = '\\t')\n",
    "display(metadata.head(1))\n",
    "metadata = utils.prepare_metadata(id_col = 'run_accession', label_col = 'sample_title', sep = ' ', label_list = ['treatment', 'remove', 'replicate', 'tet'], df = metadata)\n",
    "metadata = metadata.drop(columns = ['remove', 'tet'])\n",
    "display(metadata.sample(3))\n",
    "metadata.to_csv(os.path.join(PATH, 'data', 'metadata', 'MM_hypx.csv'), index = False)\n",
    "\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 21,
   "id": "b37884b5",
   "metadata": {},
   "outputs": [],
   "source": [
    "# Fix peaks_counts by converting a [chr, start, end, id, counts] to [chr, start, end, counts] as required by DiffBind (https://bioconductor.org/packages/release/bioc/manuals/DiffBind/man/DiffBind.pdf#page=38&zoom=100,134,300)\n",
    "\n",
    "from glob import glob\n",
    "PATH = os.getcwd()\n",
    "peaks_path = os.path.join(PATH, 'test_diffbind', 'peaks_count')\n",
    "peak_files = glob(os.path.join(peaks_path, '*.txt'))\n",
    "for peak_file in peak_files[1:]:\n",
    "    peaks_cur = pd.read_csv(peak_file, sep = '\\t', names = ['chr', 'start', 'end', 'peak_id', 'counts'], header = None, usecols = range(5))\n",
    "    peaks_cur = peaks_cur.drop(columns = ['peak_id'])\n",
    "    peaks_cur.to_csv(peak_file, sep = '\\t', index = False, header = False)\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 10,
   "id": "935969a0",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>run_accession</th>\n",
       "      <th>sample_accession</th>\n",
       "      <th>experiment_accession</th>\n",
       "      <th>study_accession</th>\n",
       "      <th>tax_id</th>\n",
       "      <th>scientific_name</th>\n",
       "      <th>fastq_ftp</th>\n",
       "      <th>submitted_ftp</th>\n",
       "      <th>sra_ftp</th>\n",
       "      <th>sample_title</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>SRR5253457</td>\n",
       "      <td>SAMN06330258</td>\n",
       "      <td>SRX2558999</td>\n",
       "      <td>PRJNA374680</td>\n",
       "      <td>9606</td>\n",
       "      <td>Homo sapiens</td>\n",
       "      <td>ftp.sra.ebi.ac.uk/vol1/fastq/SRR525/007/SRR525...</td>\n",
       "      <td>NaN</td>\n",
       "      <td>ftp.sra.ebi.ac.uk/vol1/srr/SRR525/007/SRR5253457</td>\n",
       "      <td>HUVEC GRO-seq normoxia replicate 1</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "  run_accession sample_accession experiment_accession study_accession  tax_id   \n",
       "0    SRR5253457     SAMN06330258           SRX2558999     PRJNA374680    9606  \\\n",
       "\n",
       "  scientific_name                                          fastq_ftp   \n",
       "0    Homo sapiens  ftp.sra.ebi.ac.uk/vol1/fastq/SRR525/007/SRR525...  \\\n",
       "\n",
       "   submitted_ftp                                           sra_ftp   \n",
       "0            NaN  ftp.sra.ebi.ac.uk/vol1/srr/SRR525/007/SRR5253457  \\\n",
       "\n",
       "                         sample_title  \n",
       "0  HUVEC GRO-seq normoxia replicate 1  "
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    },
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>sample_accession</th>\n",
       "      <th>cell_type</th>\n",
       "      <th>method</th>\n",
       "      <th>treatment</th>\n",
       "      <th>replicate</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>20</th>\n",
       "      <td>SAMN06330278</td>\n",
       "      <td>HUVEC</td>\n",
       "      <td>GRO-seq</td>\n",
       "      <td>normoxia</td>\n",
       "      <td>replicate_2</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>SAMN06330277</td>\n",
       "      <td>HUVEC</td>\n",
       "      <td>GRO-seq</td>\n",
       "      <td>normoxia</td>\n",
       "      <td>replicate_3</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>11</th>\n",
       "      <td>SAMN06330258</td>\n",
       "      <td>HUVEC</td>\n",
       "      <td>GRO-seq</td>\n",
       "      <td>normoxia</td>\n",
       "      <td>replicate_1</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "   sample_accession cell_type   method treatment    replicate\n",
       "20     SAMN06330278     HUVEC  GRO-seq  normoxia  replicate_2\n",
       "3      SAMN06330277     HUVEC  GRO-seq  normoxia  replicate_3\n",
       "11     SAMN06330258     HUVEC  GRO-seq  normoxia  replicate_1"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    }
   ],
   "source": [
    "huvec1_metadata = pd.read_csv(os.path.join(PATH, 'data', 'HUVEC_hypx_gro_countdata', 'HUVEC_hypx_gro.tsv'), sep = '\\t')\n",
    "display(huvec1_metadata.head(1))\n",
    "huvec1_metadata = utils.prepare_metadata(id_col = 'sample_accession', label_col = 'sample_title', sep = ' ', label_list = ['cell_type', 'method', 'treatment','replicate1','replicate2'], df = huvec1_metadata)\n",
    "# unite replicate1 and replicate2\n",
    "huvec1_metadata['replicate'] = huvec1_metadata['replicate1'] + '_' + huvec1_metadata['replicate2']\n",
    "huvec1_metadata = huvec1_metadata.drop(columns = ['replicate1', 'replicate2'])\n",
    "display(huvec1_metadata.sample(3))\n",
    "huvec1_metadata.to_csv(os.path.join(PATH, 'data', 'metadata', 'HUVEC_hypx_gro.csv'), index = False)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 2,
   "id": "083314fb",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>run_accession</th>\n",
       "      <th>sample_accession</th>\n",
       "      <th>experiment_accession</th>\n",
       "      <th>study_accession</th>\n",
       "      <th>tax_id</th>\n",
       "      <th>scientific_name</th>\n",
       "      <th>fastq_ftp</th>\n",
       "      <th>submitted_ftp</th>\n",
       "      <th>sra_ftp</th>\n",
       "      <th>sample_title</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>SRR10063441</td>\n",
       "      <td>SAMN12687155</td>\n",
       "      <td>SRX6797318</td>\n",
       "      <td>PRJNA563721</td>\n",
       "      <td>9606</td>\n",
       "      <td>Homo sapiens</td>\n",
       "      <td>ftp.sra.ebi.ac.uk/vol1/fastq/SRR100/041/SRR100...</td>\n",
       "      <td>NaN</td>\n",
       "      <td>ftp.sra.ebi.ac.uk/vol1/srr/SRR100/041/SRR10063441</td>\n",
       "      <td>HAEC.miRNA.control.rep1</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "  run_accession sample_accession experiment_accession study_accession  tax_id   \n",
       "0   SRR10063441     SAMN12687155           SRX6797318     PRJNA563721    9606  \\\n",
       "\n",
       "  scientific_name                                          fastq_ftp   \n",
       "0    Homo sapiens  ftp.sra.ebi.ac.uk/vol1/fastq/SRR100/041/SRR100...  \\\n",
       "\n",
       "   submitted_ftp                                            sra_ftp   \n",
       "0            NaN  ftp.sra.ebi.ac.uk/vol1/srr/SRR100/041/SRR10063441  \\\n",
       "\n",
       "              sample_title  \n",
       "0  HAEC.miRNA.control.rep1  "
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    },
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>sample_accession</th>\n",
       "      <th>cell_type</th>\n",
       "      <th>method</th>\n",
       "      <th>treatment</th>\n",
       "      <th>replicate</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>67</th>\n",
       "      <td>SAMN12687141</td>\n",
       "      <td>HUVEC</td>\n",
       "      <td>miRNA</td>\n",
       "      <td>hypoxia_7h</td>\n",
       "      <td>rep3</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>80</th>\n",
       "      <td>SAMN12687172</td>\n",
       "      <td>CD14+</td>\n",
       "      <td>GRO</td>\n",
       "      <td>control</td>\n",
       "      <td>rep1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>SAMN12687143</td>\n",
       "      <td>HUVEC</td>\n",
       "      <td>miRNA</td>\n",
       "      <td>hypoxia_7h</td>\n",
       "      <td>rep1</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "   sample_accession cell_type method   treatment replicate\n",
       "67     SAMN12687141     HUVEC  miRNA  hypoxia_7h      rep3\n",
       "80     SAMN12687172     CD14+    GRO     control      rep1\n",
       "4      SAMN12687143     HUVEC  miRNA  hypoxia_7h      rep1"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    }
   ],
   "source": [
    "huvec2_metadata = pd.read_csv(os.path.join(PATH, 'data', 'metadata', 'HS_hypx_huvec_haec_gro.tsv'), sep = '\\t')\n",
    "display(huvec2_metadata.head(1))\n",
    "huvec2_metadata = utils.prepare_metadata(id_col = 'sample_accession', label_col = 'sample_title', sep = '\\.', label_list = ['cell_type', 'method', 'treatment','replicate'], df = huvec2_metadata)\n",
    "# unite replicate1 and replicate2\n",
    "display(huvec2_metadata.sample(3))\n",
    "huvec2_metadata.to_csv(os.path.join(PATH, 'data', 'metadata', 'HS_hypx_huvec_haec_gro.csv'), index = False)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 39,
   "id": "f1fa56ad",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>symbol</th>\n",
       "      <th>SAMN06330278</th>\n",
       "      <th>SAMN06330276</th>\n",
       "      <th>SAMN06330275</th>\n",
       "      <th>SAMN06330274</th>\n",
       "      <th>SAMN06330258</th>\n",
       "      <th>SAMN06330277</th>\n",
       "      <th>SAMN06330273</th>\n",
       "      <th>SAMN06330272</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>gene-TRNS2</th>\n",
       "      <td>TRNS2</td>\n",
       "      <td>544</td>\n",
       "      <td>44</td>\n",
       "      <td>82</td>\n",
       "      <td>409</td>\n",
       "      <td>109</td>\n",
       "      <td>63</td>\n",
       "      <td>17</td>\n",
       "      <td>10</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>gene-TRNT</th>\n",
       "      <td>TRNT</td>\n",
       "      <td>235</td>\n",
       "      <td>35</td>\n",
       "      <td>112</td>\n",
       "      <td>429</td>\n",
       "      <td>178</td>\n",
       "      <td>48</td>\n",
       "      <td>95</td>\n",
       "      <td>65</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>gene-TRNV</th>\n",
       "      <td>TRNV</td>\n",
       "      <td>8377</td>\n",
       "      <td>55</td>\n",
       "      <td>1189</td>\n",
       "      <td>3371</td>\n",
       "      <td>2059</td>\n",
       "      <td>176</td>\n",
       "      <td>107</td>\n",
       "      <td>99</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>gene-TRNW</th>\n",
       "      <td>TRNW</td>\n",
       "      <td>272</td>\n",
       "      <td>160</td>\n",
       "      <td>320</td>\n",
       "      <td>360</td>\n",
       "      <td>486</td>\n",
       "      <td>227</td>\n",
       "      <td>90</td>\n",
       "      <td>74</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>gene-TRNY</th>\n",
       "      <td>TRNY</td>\n",
       "      <td>238</td>\n",
       "      <td>63</td>\n",
       "      <td>144</td>\n",
       "      <td>215</td>\n",
       "      <td>280</td>\n",
       "      <td>99</td>\n",
       "      <td>23</td>\n",
       "      <td>17</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "           symbol  SAMN06330278  SAMN06330276  SAMN06330275  SAMN06330274   \n",
       "gene-TRNS2  TRNS2           544            44            82           409  \\\n",
       "gene-TRNT    TRNT           235            35           112           429   \n",
       "gene-TRNV    TRNV          8377            55          1189          3371   \n",
       "gene-TRNW    TRNW           272           160           320           360   \n",
       "gene-TRNY    TRNY           238            63           144           215   \n",
       "\n",
       "            SAMN06330258  SAMN06330277  SAMN06330273  SAMN06330272  \n",
       "gene-TRNS2           109            63            17            10  \n",
       "gene-TRNT            178            48            95            65  \n",
       "gene-TRNV           2059           176           107            99  \n",
       "gene-TRNW            486           227            90            74  \n",
       "gene-TRNY            280            99            23            17  "
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    }
   ],
   "source": [
    "# Unite mt count files and countData file\n",
    "from glob import glob\n",
    "huvec1_countdata = pd.read_csv(os.path.join(PATH, 'data', 'HUVEC_hypx_gro_countdata', 'HUVEC_hypx_gro_countData.csv')).rename(columns = {'geneName' : 'symbol'})\n",
    "huvec1_countdata.columns = [i.split('_')[0] for i in huvec1_countdata.columns]\n",
    "mt_counts = utils.prepare_counts(os.path.join(PATH, 'data', 'HUVEC_hypx_gro_countdata', 'countdata'), keep_symbol = True)\n",
    "huvec1_countdata = pd.concat([huvec1_countdata, mt_counts], axis = 0)\n",
    "display(huvec1_countdata.tail(5))\n",
    "\n",
    "huvec1_countdata.to_csv(os.path.join(PATH, 'data', 'count_tables', 'HUVEC_hypx_gro.csv'), index = False)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 41,
   "id": "9ef745da",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>symbol</th>\n",
       "      <th>SAMN12687223</th>\n",
       "      <th>SAMN12687237</th>\n",
       "      <th>SAMN12687203</th>\n",
       "      <th>SAMN12687201</th>\n",
       "      <th>SAMN12687230</th>\n",
       "      <th>SAMN12687225</th>\n",
       "      <th>SAMN12687224</th>\n",
       "      <th>SAMN12687170</th>\n",
       "      <th>SAMN12687189</th>\n",
       "      <th>SAMN12687232</th>\n",
       "      <th>SAMN12687231</th>\n",
       "      <th>SAMN12687222</th>\n",
       "      <th>SAMN12687221</th>\n",
       "      <th>SAMN12687213</th>\n",
       "      <th>SAMN12687172</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>gene-TRNN</th>\n",
       "      <td>TRNN</td>\n",
       "      <td>74</td>\n",
       "      <td>596</td>\n",
       "      <td>5803</td>\n",
       "      <td>1293</td>\n",
       "      <td>87</td>\n",
       "      <td>253</td>\n",
       "      <td>85</td>\n",
       "      <td>138</td>\n",
       "      <td>120</td>\n",
       "      <td>96</td>\n",
       "      <td>209</td>\n",
       "      <td>62</td>\n",
       "      <td>23</td>\n",
       "      <td>496</td>\n",
       "      <td>67.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>gene-TRNP</th>\n",
       "      <td>TRNP</td>\n",
       "      <td>1304</td>\n",
       "      <td>1058</td>\n",
       "      <td>30740</td>\n",
       "      <td>7554</td>\n",
       "      <td>340</td>\n",
       "      <td>247</td>\n",
       "      <td>205</td>\n",
       "      <td>408</td>\n",
       "      <td>139</td>\n",
       "      <td>154</td>\n",
       "      <td>1125</td>\n",
       "      <td>751</td>\n",
       "      <td>323</td>\n",
       "      <td>847</td>\n",
       "      <td>167.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>gene-TRNQ</th>\n",
       "      <td>TRNQ</td>\n",
       "      <td>205</td>\n",
       "      <td>265</td>\n",
       "      <td>4262</td>\n",
       "      <td>2780</td>\n",
       "      <td>57</td>\n",
       "      <td>110</td>\n",
       "      <td>71</td>\n",
       "      <td>128</td>\n",
       "      <td>89</td>\n",
       "      <td>25</td>\n",
       "      <td>101</td>\n",
       "      <td>37</td>\n",
       "      <td>23</td>\n",
       "      <td>145</td>\n",
       "      <td>81.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>gene-TRNR</th>\n",
       "      <td>TRNR</td>\n",
       "      <td>116</td>\n",
       "      <td>233</td>\n",
       "      <td>4760</td>\n",
       "      <td>616</td>\n",
       "      <td>42</td>\n",
       "      <td>137</td>\n",
       "      <td>112</td>\n",
       "      <td>117</td>\n",
       "      <td>80</td>\n",
       "      <td>48</td>\n",
       "      <td>147</td>\n",
       "      <td>40</td>\n",
       "      <td>17</td>\n",
       "      <td>233</td>\n",
       "      <td>274.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>gene-TRNS1</th>\n",
       "      <td>TRNS1</td>\n",
       "      <td>197</td>\n",
       "      <td>689</td>\n",
       "      <td>10674</td>\n",
       "      <td>49999</td>\n",
       "      <td>175</td>\n",
       "      <td>245</td>\n",
       "      <td>206</td>\n",
       "      <td>407</td>\n",
       "      <td>155</td>\n",
       "      <td>83</td>\n",
       "      <td>220</td>\n",
       "      <td>110</td>\n",
       "      <td>110</td>\n",
       "      <td>345</td>\n",
       "      <td>183.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>gene-TRNS2</th>\n",
       "      <td>TRNS2</td>\n",
       "      <td>58</td>\n",
       "      <td>212</td>\n",
       "      <td>6447</td>\n",
       "      <td>2402</td>\n",
       "      <td>29</td>\n",
       "      <td>63</td>\n",
       "      <td>44</td>\n",
       "      <td>37</td>\n",
       "      <td>43</td>\n",
       "      <td>29</td>\n",
       "      <td>75</td>\n",
       "      <td>17</td>\n",
       "      <td>10</td>\n",
       "      <td>104</td>\n",
       "      <td>83.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>gene-TRNT</th>\n",
       "      <td>TRNT</td>\n",
       "      <td>142</td>\n",
       "      <td>150</td>\n",
       "      <td>518</td>\n",
       "      <td>539</td>\n",
       "      <td>37</td>\n",
       "      <td>48</td>\n",
       "      <td>35</td>\n",
       "      <td>35</td>\n",
       "      <td>26</td>\n",
       "      <td>31</td>\n",
       "      <td>203</td>\n",
       "      <td>95</td>\n",
       "      <td>65</td>\n",
       "      <td>172</td>\n",
       "      <td>22.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>gene-TRNV</th>\n",
       "      <td>TRNV</td>\n",
       "      <td>771</td>\n",
       "      <td>521</td>\n",
       "      <td>12959</td>\n",
       "      <td>3810</td>\n",
       "      <td>87</td>\n",
       "      <td>176</td>\n",
       "      <td>55</td>\n",
       "      <td>191</td>\n",
       "      <td>146</td>\n",
       "      <td>68</td>\n",
       "      <td>289</td>\n",
       "      <td>107</td>\n",
       "      <td>99</td>\n",
       "      <td>247</td>\n",
       "      <td>349.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>gene-TRNW</th>\n",
       "      <td>TRNW</td>\n",
       "      <td>205</td>\n",
       "      <td>306</td>\n",
       "      <td>3006</td>\n",
       "      <td>448</td>\n",
       "      <td>82</td>\n",
       "      <td>227</td>\n",
       "      <td>160</td>\n",
       "      <td>200</td>\n",
       "      <td>126</td>\n",
       "      <td>31</td>\n",
       "      <td>206</td>\n",
       "      <td>90</td>\n",
       "      <td>74</td>\n",
       "      <td>198</td>\n",
       "      <td>76.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>gene-TRNY</th>\n",
       "      <td>TRNY</td>\n",
       "      <td>52</td>\n",
       "      <td>220</td>\n",
       "      <td>5346</td>\n",
       "      <td>6339</td>\n",
       "      <td>42</td>\n",
       "      <td>99</td>\n",
       "      <td>63</td>\n",
       "      <td>103</td>\n",
       "      <td>73</td>\n",
       "      <td>45</td>\n",
       "      <td>67</td>\n",
       "      <td>23</td>\n",
       "      <td>17</td>\n",
       "      <td>82</td>\n",
       "      <td>36.0</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "           symbol  SAMN12687223  SAMN12687237  SAMN12687203  SAMN12687201   \n",
       "gene-TRNN    TRNN            74           596          5803          1293  \\\n",
       "gene-TRNP    TRNP          1304          1058         30740          7554   \n",
       "gene-TRNQ    TRNQ           205           265          4262          2780   \n",
       "gene-TRNR    TRNR           116           233          4760           616   \n",
       "gene-TRNS1  TRNS1           197           689         10674         49999   \n",
       "gene-TRNS2  TRNS2            58           212          6447          2402   \n",
       "gene-TRNT    TRNT           142           150           518           539   \n",
       "gene-TRNV    TRNV           771           521         12959          3810   \n",
       "gene-TRNW    TRNW           205           306          3006           448   \n",
       "gene-TRNY    TRNY            52           220          5346          6339   \n",
       "\n",
       "            SAMN12687230  SAMN12687225  SAMN12687224  SAMN12687170   \n",
       "gene-TRNN             87           253            85           138  \\\n",
       "gene-TRNP            340           247           205           408   \n",
       "gene-TRNQ             57           110            71           128   \n",
       "gene-TRNR             42           137           112           117   \n",
       "gene-TRNS1           175           245           206           407   \n",
       "gene-TRNS2            29            63            44            37   \n",
       "gene-TRNT             37            48            35            35   \n",
       "gene-TRNV             87           176            55           191   \n",
       "gene-TRNW             82           227           160           200   \n",
       "gene-TRNY             42            99            63           103   \n",
       "\n",
       "            SAMN12687189  SAMN12687232  SAMN12687231  SAMN12687222   \n",
       "gene-TRNN            120            96           209            62  \\\n",
       "gene-TRNP            139           154          1125           751   \n",
       "gene-TRNQ             89            25           101            37   \n",
       "gene-TRNR             80            48           147            40   \n",
       "gene-TRNS1           155            83           220           110   \n",
       "gene-TRNS2            43            29            75            17   \n",
       "gene-TRNT             26            31           203            95   \n",
       "gene-TRNV            146            68           289           107   \n",
       "gene-TRNW            126            31           206            90   \n",
       "gene-TRNY             73            45            67            23   \n",
       "\n",
       "            SAMN12687221  SAMN12687213  SAMN12687172  \n",
       "gene-TRNN             23           496          67.0  \n",
       "gene-TRNP            323           847         167.0  \n",
       "gene-TRNQ             23           145          81.0  \n",
       "gene-TRNR             17           233         274.0  \n",
       "gene-TRNS1           110           345         183.0  \n",
       "gene-TRNS2            10           104          83.0  \n",
       "gene-TRNT             65           172          22.0  \n",
       "gene-TRNV             99           247         349.0  \n",
       "gene-TRNW             74           198          76.0  \n",
       "gene-TRNY             17            82          36.0  "
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    }
   ],
   "source": [
    "# Unite mt count files and countData file\n",
    "from glob import glob\n",
    "huvec2_countdata = pd.read_csv(os.path.join(PATH, 'data', 'HS_hypx_huvec_haec_gro_countdata', 'HS_hypx_huvec_haec_gro_countData.csv')).rename(columns = {'geneName' : 'symbol'})\n",
    "huvec2_countdata.columns = [i.split('_')[0] for i in huvec2_countdata.columns]\n",
    "mt_counts = utils.prepare_counts(os.path.join(PATH, 'data', 'HS_hypx_huvec_haec_gro_countdata', 'countdata'), keep_symbol = True)\n",
    "huvec2_countdata = pd.concat([huvec2_countdata, mt_counts], axis = 0)\n",
    "display(huvec2_countdata.tail(10))\n",
    "\n",
    "huvec2_countdata.to_csv(os.path.join(PATH, 'data', 'count_tables', 'HS_hypx_huvec_haec_gro.csv'), index = False)"
   ]
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python 3",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.11.9"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 5
}
